{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# Pathway Analysis\n",
    "\n",
    "Pathway Analysis refers to the set of tools and techniques used to assess data collected from high-throughput technologies, with the aim of discovering meaningful group of genes that are altered in case samples when compared to a control. BY measuring gene expression changes in a pathway, Pathway Analysis methods attempt to derive insight from lists of genes obtained from Differential Expression analyses. Pathway Enrichment Analysis helps identify Biological pathways enriched in a genelist more than would be expected by chance.\n",
    "In this module, Pathway Analysis methods including Over Representation Analysis (ORA), Fast Gene Set Enrichment Analysis (FGSEA) and GSA are explained.\n",
    "\n",
    "In ORA, for each pathway of interest, several statistical tests are used to test the pathways for over-representation in the gene list compared to what would be expected by chance, and these consider the number of genes detected in the experiment and their relative ranking as well as the  number of genes annotated to a pathway of interest. ORA obtains the statistical significance of the overlap between the number of similar genes shared by an input gene set and each annotated gene set via statistical tests and uses a p-value cutoff to identify annotated gene sets with significant overlaps to the input gene set.\n",
    "Gene Set Enrichment Analysis (GSEA) avoids the need for an ad hoc cutoff (e.g. expression fold change) in identifying the input gene set by ranking all genes in the genome by differential expression level and then checks if any annotated gene set is scored unexpectedly high or low.\n",
    "\n",
    "\n",
    "\n",
    "## Learning Objectives:\n",
    "1. Introduction to enrichment analysis,\n",
    "2. Performing Enrichment analysis using different methods,\n",
    "3. Visualizing the outputs,\n",
    "4. Saving results."
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Enrichment analysis\n",
    " Gene Set Enrichment Analysis methods help gain insight into obtained gene lists by identifying pathways that are enriched in a gene list more than would be expected by chance. The three major steps involved the process include; definition of a gene list from omics data, determination of statistically enriched pathways, and visualization and interpretation of the results."
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Loading and mapping DE genes"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "outputs": [
    {
     "data": {
      "text/html": "<table class=\"dataframe\">\n<caption>A data.frame: 6 × 7</caption>\n<thead>\n\t<tr><th></th><th scope=col>PROBEID</th><th scope=col>logFC</th><th scope=col>AveExpr</th><th scope=col>t</th><th scope=col>P.Value</th><th scope=col>adj.P.Val</th><th scope=col>B</th></tr>\n\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n</thead>\n<tbody>\n\t<tr><th scope=row>222178_s_at</th><td>222178_s_at</td><td>-0.4473780</td><td>0.3653633</td><td>-73.01106</td><td>4.019112e-55</td><td>2.197450e-50</td><td>81.05538</td></tr>\n\t<tr><th scope=row>224687_at</th><td>224687_at  </td><td>-4.1431631</td><td>2.2572347</td><td>-48.92036</td><td>5.489424e-46</td><td>1.500671e-41</td><td>74.18354</td></tr>\n\t<tr><th scope=row>207488_at</th><td>207488_at  </td><td>-0.4017530</td><td>0.6243144</td><td>-39.29933</td><td>4.735684e-41</td><td>8.630785e-37</td><td>68.83493</td></tr>\n\t<tr><th scope=row>239226_at</th><td>239226_at  </td><td> 0.4483302</td><td>1.3245501</td><td> 28.36499</td><td>7.626558e-34</td><td>1.042455e-29</td><td>58.84908</td></tr>\n\t<tr><th scope=row>234109_x_at</th><td>234109_x_at</td><td>-0.2289726</td><td>0.7508256</td><td>-27.67010</td><td>2.635337e-33</td><td>2.881741e-29</td><td>58.00438</td></tr>\n\t<tr><th scope=row>212833_at</th><td>212833_at  </td><td>-2.5976592</td><td>1.7308944</td><td>-24.01172</td><td>2.914912e-30</td><td>2.656213e-26</td><td>52.99567</td></tr>\n</tbody>\n</table>\n",
      "text/markdown": "\nA data.frame: 6 × 7\n\n| <!--/--> | PROBEID &lt;chr&gt; | logFC &lt;dbl&gt; | AveExpr &lt;dbl&gt; | t &lt;dbl&gt; | P.Value &lt;dbl&gt; | adj.P.Val &lt;dbl&gt; | B &lt;dbl&gt; |\n|---|---|---|---|---|---|---|---|\n| 222178_s_at | 222178_s_at | -0.4473780 | 0.3653633 | -73.01106 | 4.019112e-55 | 2.197450e-50 | 81.05538 |\n| 224687_at | 224687_at   | -4.1431631 | 2.2572347 | -48.92036 | 5.489424e-46 | 1.500671e-41 | 74.18354 |\n| 207488_at | 207488_at   | -0.4017530 | 0.6243144 | -39.29933 | 4.735684e-41 | 8.630785e-37 | 68.83493 |\n| 239226_at | 239226_at   |  0.4483302 | 1.3245501 |  28.36499 | 7.626558e-34 | 1.042455e-29 | 58.84908 |\n| 234109_x_at | 234109_x_at | -0.2289726 | 0.7508256 | -27.67010 | 2.635337e-33 | 2.881741e-29 | 58.00438 |\n| 212833_at | 212833_at   | -2.5976592 | 1.7308944 | -24.01172 | 2.914912e-30 | 2.656213e-26 | 52.99567 |\n\n",
      "text/latex": "A data.frame: 6 × 7\n\\begin{tabular}{r|lllllll}\n  & PROBEID & logFC & AveExpr & t & P.Value & adj.P.Val & B\\\\\n  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n\\hline\n\t222178\\_s\\_at & 222178\\_s\\_at & -0.4473780 & 0.3653633 & -73.01106 & 4.019112e-55 & 2.197450e-50 & 81.05538\\\\\n\t224687\\_at & 224687\\_at   & -4.1431631 & 2.2572347 & -48.92036 & 5.489424e-46 & 1.500671e-41 & 74.18354\\\\\n\t207488\\_at & 207488\\_at   & -0.4017530 & 0.6243144 & -39.29933 & 4.735684e-41 & 8.630785e-37 & 68.83493\\\\\n\t239226\\_at & 239226\\_at   &  0.4483302 & 1.3245501 &  28.36499 & 7.626558e-34 & 1.042455e-29 & 58.84908\\\\\n\t234109\\_x\\_at & 234109\\_x\\_at & -0.2289726 & 0.7508256 & -27.67010 & 2.635337e-33 & 2.881741e-29 & 58.00438\\\\\n\t212833\\_at & 212833\\_at   & -2.5976592 & 1.7308944 & -24.01172 & 2.914912e-30 & 2.656213e-26 & 52.99567\\\\\n\\end{tabular}\n",
      "text/plain": "            PROBEID     logFC      AveExpr   t         P.Value     \n222178_s_at 222178_s_at -0.4473780 0.3653633 -73.01106 4.019112e-55\n224687_at   224687_at   -4.1431631 2.2572347 -48.92036 5.489424e-46\n207488_at   207488_at   -0.4017530 0.6243144 -39.29933 4.735684e-41\n239226_at   239226_at    0.4483302 1.3245501  28.36499 7.626558e-34\n234109_x_at 234109_x_at -0.2289726 0.7508256 -27.67010 2.635337e-33\n212833_at   212833_at   -2.5976592 1.7308944 -24.01172 2.914912e-30\n            adj.P.Val    B       \n222178_s_at 2.197450e-50 81.05538\n224687_at   1.500671e-41 74.18354\n207488_at   8.630785e-37 68.83493\n239226_at   1.042455e-29 58.84908\n234109_x_at 2.881741e-29 58.00438\n212833_at   2.656213e-26 52.99567"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": "<style>\n.list-inline {list-style: none; margin:0; padding: 0}\n.list-inline>li {display: inline-block}\n.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n</style>\n<ol class=list-inline><li>54675</li><li>7</li></ol>\n",
      "text/markdown": "1. 54675\n2. 7\n\n\n",
      "text/latex": "\\begin{enumerate*}\n\\item 54675\n\\item 7\n\\end{enumerate*}\n",
      "text/plain": "[1] 54675     7"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data <- readRDS(\"./data/DE_genes.rds\")\n",
    "rownames(data) <- data$PROBEID\n",
    "head(data)\n",
    "dim(data)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# Loading needed libraries"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "outputs": [],
   "source": [
    "suppressMessages({\n",
    "  library(hgu133plus2.db)\n",
    "  library(AnnotationDbi)\n",
    "})"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "outputs": [],
   "source": [
    "# We will then obtain the probe IDs and map them to the Gene name and Gene Symbol"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "probeIDs = data$PROBEID"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "outputs": [],
   "source": [
    "suppressMessages({\n",
    "annotLookup <- AnnotationDbi::select(hgu133plus2.db, keys = probeIDs, columns = c('PROBEID','GENENAME','SYMBOL'))\n",
    "})"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "outputs": [
    {
     "data": {
      "text/html": "<table class=\"dataframe\">\n<caption>A data.frame: 6 × 3</caption>\n<thead>\n\t<tr><th></th><th scope=col>PROBEID</th><th scope=col>GENENAME</th><th scope=col>SYMBOL</th></tr>\n\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n</thead>\n<tbody>\n\t<tr><th scope=row>1</th><td>222178_s_at</td><td>NA                                        </td><td>NA      </td></tr>\n\t<tr><th scope=row>2</th><td>224687_at  </td><td>ankyrin repeat and IBR domain containing 1</td><td>ANKIB1  </td></tr>\n\t<tr><th scope=row>3</th><td>207488_at  </td><td>NA                                        </td><td>NA      </td></tr>\n\t<tr><th scope=row>4</th><td>239226_at  </td><td>NA                                        </td><td>NA      </td></tr>\n\t<tr><th scope=row>5</th><td>234109_x_at</td><td>one cut homeobox 3                        </td><td>ONECUT3 </td></tr>\n\t<tr><th scope=row>6</th><td>212833_at  </td><td>solute carrier family 25 member 46        </td><td>SLC25A46</td></tr>\n</tbody>\n</table>\n",
      "text/markdown": "\nA data.frame: 6 × 3\n\n| <!--/--> | PROBEID &lt;chr&gt; | GENENAME &lt;chr&gt; | SYMBOL &lt;chr&gt; |\n|---|---|---|---|\n| 1 | 222178_s_at | NA                                         | NA       |\n| 2 | 224687_at   | ankyrin repeat and IBR domain containing 1 | ANKIB1   |\n| 3 | 207488_at   | NA                                         | NA       |\n| 4 | 239226_at   | NA                                         | NA       |\n| 5 | 234109_x_at | one cut homeobox 3                         | ONECUT3  |\n| 6 | 212833_at   | solute carrier family 25 member 46         | SLC25A46 |\n\n",
      "text/latex": "A data.frame: 6 × 3\n\\begin{tabular}{r|lll}\n  & PROBEID & GENENAME & SYMBOL\\\\\n  & <chr> & <chr> & <chr>\\\\\n\\hline\n\t1 & 222178\\_s\\_at & NA                                         & NA      \\\\\n\t2 & 224687\\_at   & ankyrin repeat and IBR domain containing 1 & ANKIB1  \\\\\n\t3 & 207488\\_at   & NA                                         & NA      \\\\\n\t4 & 239226\\_at   & NA                                         & NA      \\\\\n\t5 & 234109\\_x\\_at & one cut homeobox 3                         & ONECUT3 \\\\\n\t6 & 212833\\_at   & solute carrier family 25 member 46         & SLC25A46\\\\\n\\end{tabular}\n",
      "text/plain": "  PROBEID     GENENAME                                   SYMBOL  \n1 222178_s_at NA                                         NA      \n2 224687_at   ankyrin repeat and IBR domain containing 1 ANKIB1  \n3 207488_at   NA                                         NA      \n4 239226_at   NA                                         NA      \n5 234109_x_at one cut homeobox 3                         ONECUT3 \n6 212833_at   solute carrier family 25 member 46         SLC25A46"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(annotLookup)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "outputs": [],
   "source": [
    "# Merge DE result  data frame with annotation table\n",
    "data = merge(annotLookup, data, by=\"PROBEID\")"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "outputs": [
    {
     "data": {
      "text/html": "<table class=\"dataframe\">\n<caption>A data.frame: 6 × 9</caption>\n<thead>\n\t<tr><th></th><th scope=col>PROBEID</th><th scope=col>GENENAME</th><th scope=col>SYMBOL</th><th scope=col>logFC</th><th scope=col>AveExpr</th><th scope=col>t</th><th scope=col>P.Value</th><th scope=col>adj.P.Val</th><th scope=col>B</th></tr>\n\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n</thead>\n<tbody>\n\t<tr><th scope=row>1</th><td>1007_s_at</td><td>discoidin domain receptor tyrosine kinase 1 </td><td>DDR1  </td><td>-0.23858465</td><td>1.0967772</td><td>-2.08257585</td><td>0.04210211</td><td>0.2531265</td><td>-4.284228</td></tr>\n\t<tr><th scope=row>2</th><td>1053_at  </td><td>replication factor C subunit 2              </td><td>RFC2  </td><td> 0.05830367</td><td>0.9657758</td><td> 1.42311356</td><td>0.16052793</td><td>0.4674726</td><td>-5.359907</td></tr>\n\t<tr><th scope=row>3</th><td>117_at   </td><td>heat shock protein family A (Hsp70) member 6</td><td>HSPA6 </td><td>-0.01138494</td><td>1.1501582</td><td>-0.08100885</td><td>0.93573821</td><td>0.9861536</td><td>-6.337785</td></tr>\n\t<tr><th scope=row>4</th><td>121_at   </td><td>paired box 8                                </td><td>PAX8  </td><td> 0.02301084</td><td>1.0581847</td><td> 0.87166663</td><td>0.38729816</td><td>0.7236528</td><td>-5.968512</td></tr>\n\t<tr><th scope=row>5</th><td>1255_g_at</td><td>guanylate cyclase activator 1A              </td><td>GUCA1A</td><td> 0.36999950</td><td>1.4080351</td><td> 1.78615022</td><td>0.07976123</td><td>0.3351480</td><td>-4.812134</td></tr>\n\t<tr><th scope=row>6</th><td>1294_at  </td><td>ubiquitin like modifier activating enzyme 7 </td><td>UBA7  </td><td>-0.11221434</td><td>1.0859983</td><td>-1.61593670</td><td>0.11200995</td><td>0.3925252</td><td>-5.082923</td></tr>\n</tbody>\n</table>\n",
      "text/markdown": "\nA data.frame: 6 × 9\n\n| <!--/--> | PROBEID &lt;chr&gt; | GENENAME &lt;chr&gt; | SYMBOL &lt;chr&gt; | logFC &lt;dbl&gt; | AveExpr &lt;dbl&gt; | t &lt;dbl&gt; | P.Value &lt;dbl&gt; | adj.P.Val &lt;dbl&gt; | B &lt;dbl&gt; |\n|---|---|---|---|---|---|---|---|---|---|\n| 1 | 1007_s_at | discoidin domain receptor tyrosine kinase 1  | DDR1   | -0.23858465 | 1.0967772 | -2.08257585 | 0.04210211 | 0.2531265 | -4.284228 |\n| 2 | 1053_at   | replication factor C subunit 2               | RFC2   |  0.05830367 | 0.9657758 |  1.42311356 | 0.16052793 | 0.4674726 | -5.359907 |\n| 3 | 117_at    | heat shock protein family A (Hsp70) member 6 | HSPA6  | -0.01138494 | 1.1501582 | -0.08100885 | 0.93573821 | 0.9861536 | -6.337785 |\n| 4 | 121_at    | paired box 8                                 | PAX8   |  0.02301084 | 1.0581847 |  0.87166663 | 0.38729816 | 0.7236528 | -5.968512 |\n| 5 | 1255_g_at | guanylate cyclase activator 1A               | GUCA1A |  0.36999950 | 1.4080351 |  1.78615022 | 0.07976123 | 0.3351480 | -4.812134 |\n| 6 | 1294_at   | ubiquitin like modifier activating enzyme 7  | UBA7   | -0.11221434 | 1.0859983 | -1.61593670 | 0.11200995 | 0.3925252 | -5.082923 |\n\n",
      "text/latex": "A data.frame: 6 × 9\n\\begin{tabular}{r|lllllllll}\n  & PROBEID & GENENAME & SYMBOL & logFC & AveExpr & t & P.Value & adj.P.Val & B\\\\\n  & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n\\hline\n\t1 & 1007\\_s\\_at & discoidin domain receptor tyrosine kinase 1  & DDR1   & -0.23858465 & 1.0967772 & -2.08257585 & 0.04210211 & 0.2531265 & -4.284228\\\\\n\t2 & 1053\\_at   & replication factor C subunit 2               & RFC2   &  0.05830367 & 0.9657758 &  1.42311356 & 0.16052793 & 0.4674726 & -5.359907\\\\\n\t3 & 117\\_at    & heat shock protein family A (Hsp70) member 6 & HSPA6  & -0.01138494 & 1.1501582 & -0.08100885 & 0.93573821 & 0.9861536 & -6.337785\\\\\n\t4 & 121\\_at    & paired box 8                                 & PAX8   &  0.02301084 & 1.0581847 &  0.87166663 & 0.38729816 & 0.7236528 & -5.968512\\\\\n\t5 & 1255\\_g\\_at & guanylate cyclase activator 1A               & GUCA1A &  0.36999950 & 1.4080351 &  1.78615022 & 0.07976123 & 0.3351480 & -4.812134\\\\\n\t6 & 1294\\_at   & ubiquitin like modifier activating enzyme 7  & UBA7   & -0.11221434 & 1.0859983 & -1.61593670 & 0.11200995 & 0.3925252 & -5.082923\\\\\n\\end{tabular}\n",
      "text/plain": "  PROBEID   GENENAME                                     SYMBOL logFC      \n1 1007_s_at discoidin domain receptor tyrosine kinase 1  DDR1   -0.23858465\n2 1053_at   replication factor C subunit 2               RFC2    0.05830367\n3 117_at    heat shock protein family A (Hsp70) member 6 HSPA6  -0.01138494\n4 121_at    paired box 8                                 PAX8    0.02301084\n5 1255_g_at guanylate cyclase activator 1A               GUCA1A  0.36999950\n6 1294_at   ubiquitin like modifier activating enzyme 7  UBA7   -0.11221434\n  AveExpr   t           P.Value    adj.P.Val B        \n1 1.0967772 -2.08257585 0.04210211 0.2531265 -4.284228\n2 0.9657758  1.42311356 0.16052793 0.4674726 -5.359907\n3 1.1501582 -0.08100885 0.93573821 0.9861536 -6.337785\n4 1.0581847  0.87166663 0.38729816 0.7236528 -5.968512\n5 1.4080351  1.78615022 0.07976123 0.3351480 -4.812134\n6 1.0859983 -1.61593670 0.11200995 0.3925252 -5.082923"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(data)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Enrichment analysis using ORA\n",
    "Over-representation (or enrichment) analysis is a statistical method that determines whether genes from pre-defined sets (ex: those belonging to a specific GO term or KEGG pathway) are present more than would be expected (over-represented) in a subset of your data. The p-value can be calculated by hyper-geometric distribution and it is given by:\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "![images/Module4/ora_p_value.png](images/Module4/ora_p_value.png)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "where N is the total number of genes in the background distribution, M is the number of genes within that distribution that are annotated (either directly or indirectly) to the gene set of interest, n is the size of the list of genes of interest and k is the number of genes within that list which are annotated to the gene set. The background distribution by default is all the genes that have annotation."
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Over Representation Analysis Using GO and KEGG\n",
    "Getting the DE genes from  `limma` result"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in data$adj.P.Val: object of type 'closure' is not subsettable\n",
     "output_type": "error",
     "traceback": [
      "Error in data$adj.P.Val: object of type 'closure' is not subsettable\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "# Getting a list of DE genes\n",
    "DEGenes <- data[!is.na(data$adj.P.Val) & data$adj.P.Val <= 0.05,]\n",
    "# Select gene symbol\n",
    "DEGenes <- DEGenes$SYMBOL\n",
    "# Remove gene with NA symbol\n",
    "DEGenes <- na.omit(DEGenes)\n",
    "# Remove duplicated genes\n",
    "DEGenes <- unique(DEGenes)\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in data$SYMBOL: object of type 'closure' is not subsettable\n",
     "output_type": "error",
     "traceback": [
      "Error in data$SYMBOL: object of type 'closure' is not subsettable\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "#Defining background set\n",
    "backgroundSet <- data$SYMBOL\n",
    "backgroundSet <- na.omit(backgroundSet)\n",
    "backgroundSet <- unique(backgroundSet)\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "outputs": [],
   "source": [
    "suppressMessages({\n",
    "  library(tidyverse)\n",
    "})"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# We create a function to read the gene list"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "outputs": [],
   "source": [
    "gmt2geneset <- function(path){\n",
    "  genesets <- read_tsv(path, col_names = F) %>% apply(MARGIN = 1, function(r){\n",
    "    genes = unique(r[-(1:2)])\n",
    "    list(\n",
    "      id = r[1],\n",
    "      description = r[2],\n",
    "      genes = genes[!is.na(genes)]\n",
    "    )\n",
    "  })\n",
    "\n",
    "  gs <- lapply(genesets, function(g) g$genes %>% as.character())\n",
    "  names(gs) <- lapply(genesets, function(g) g$id)\n",
    "  gs\n",
    "}"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"One or more parsing issues, see `problems()` for details\"\n"
     ]
    }
   ],
   "source": [
    "suppressMessages({geneset <- gmt2geneset(\"./data/GO_terms.gmt\")})"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "Perform over-representation analysis"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "outputs": [],
   "source": [
    "res <- sapply(geneset, function(gs){\n",
    "  # hypergeomestric test\n",
    "  wBallDraw <- intersect(gs, DEGenes) %>%  length() - 1\n",
    "  if (wBallDraw < 0) return(1)\n",
    "  wBall <- length(DEGenes)\n",
    "  bBall <- nrow(data) - length(DEGenes)\n",
    "  ballDraw <- length(intersect(gs, backgroundSet))\n",
    "  1 - phyper(wBallDraw, wBall, bBall, ballDraw)\n",
    "})"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "outputs": [],
   "source": [
    "res.df <- data.frame(\n",
    "  pathway = names(res),\n",
    "  pvalue = res\n",
    ")"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "outputs": [
    {
     "data": {
      "text/html": "<table class=\"dataframe\">\n<caption>A data.frame: 6 × 3</caption>\n<thead>\n\t<tr><th></th><th scope=col>pathway</th><th scope=col>pvalue</th><th scope=col>pvalue.adj</th></tr>\n\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n</thead>\n<tbody>\n\t<tr><th scope=row>GO:0007015</th><td>GO:0007015</td><td>1.110223e-16</td><td>1.540542e-16</td></tr>\n\t<tr><th scope=row>GO:0007409</th><td>GO:0007409</td><td>1.110223e-16</td><td>1.540542e-16</td></tr>\n\t<tr><th scope=row>GO:0009150</th><td>GO:0009150</td><td>1.110223e-16</td><td>1.540542e-16</td></tr>\n\t<tr><th scope=row>GO:0009259</th><td>GO:0009259</td><td>1.110223e-16</td><td>1.540542e-16</td></tr>\n\t<tr><th scope=row>GO:0019693</th><td>GO:0019693</td><td>1.110223e-16</td><td>1.540542e-16</td></tr>\n\t<tr><th scope=row>GO:0031098</th><td>GO:0031098</td><td>1.110223e-16</td><td>1.540542e-16</td></tr>\n</tbody>\n</table>\n",
      "text/markdown": "\nA data.frame: 6 × 3\n\n| <!--/--> | pathway &lt;chr&gt; | pvalue &lt;dbl&gt; | pvalue.adj &lt;dbl&gt; |\n|---|---|---|---|\n| GO:0007015 | GO:0007015 | 1.110223e-16 | 1.540542e-16 |\n| GO:0007409 | GO:0007409 | 1.110223e-16 | 1.540542e-16 |\n| GO:0009150 | GO:0009150 | 1.110223e-16 | 1.540542e-16 |\n| GO:0009259 | GO:0009259 | 1.110223e-16 | 1.540542e-16 |\n| GO:0019693 | GO:0019693 | 1.110223e-16 | 1.540542e-16 |\n| GO:0031098 | GO:0031098 | 1.110223e-16 | 1.540542e-16 |\n\n",
      "text/latex": "A data.frame: 6 × 3\n\\begin{tabular}{r|lll}\n  & pathway & pvalue & pvalue.adj\\\\\n  & <chr> & <dbl> & <dbl>\\\\\n\\hline\n\tGO:0007015 & GO:0007015 & 1.110223e-16 & 1.540542e-16\\\\\n\tGO:0007409 & GO:0007409 & 1.110223e-16 & 1.540542e-16\\\\\n\tGO:0009150 & GO:0009150 & 1.110223e-16 & 1.540542e-16\\\\\n\tGO:0009259 & GO:0009259 & 1.110223e-16 & 1.540542e-16\\\\\n\tGO:0019693 & GO:0019693 & 1.110223e-16 & 1.540542e-16\\\\\n\tGO:0031098 & GO:0031098 & 1.110223e-16 & 1.540542e-16\\\\\n\\end{tabular}\n",
      "text/plain": "           pathway    pvalue       pvalue.adj  \nGO:0007015 GO:0007015 1.110223e-16 1.540542e-16\nGO:0007409 GO:0007409 1.110223e-16 1.540542e-16\nGO:0009150 GO:0009150 1.110223e-16 1.540542e-16\nGO:0009259 GO:0009259 1.110223e-16 1.540542e-16\nGO:0019693 GO:0019693 1.110223e-16 1.540542e-16\nGO:0031098 GO:0031098 1.110223e-16 1.540542e-16"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "res.df$pvalue.adj <- p.adjust(res.df$pvalue, method = \"fdr\")\n",
    "res.df <- res.df[order(!res.df$pvalue.adj),]\n",
    "head(res.df)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "### ORA With KEGG Data"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# We read the gene list from the KEGG pathways gmt file"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"One or more parsing issues, see `problems()` for details\"\n"
     ]
    }
   ],
   "source": [
    "suppressMessages({geneset <- gmt2geneset(\"./data/KEGG_pathways.gmt\")})"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# Call the ORA method on the KEGG data"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "outputs": [
    {
     "data": {
      "text/html": "<table class=\"dataframe\">\n<caption>A data.frame: 6 × 3</caption>\n<thead>\n\t<tr><th></th><th scope=col>pathway</th><th scope=col>pvalue</th><th scope=col>pvalue.adj</th></tr>\n\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n</thead>\n<tbody>\n\t<tr><th scope=row>hsa00010</th><td>hsa00010</td><td>0.003514458</td><td>0.03722398</td></tr>\n\t<tr><th scope=row>hsa00020</th><td>hsa00020</td><td>0.033191479</td><td>0.15959191</td></tr>\n\t<tr><th scope=row>hsa00030</th><td>hsa00030</td><td>0.033191479</td><td>0.15959191</td></tr>\n\t<tr><th scope=row>hsa00040</th><td>hsa00040</td><td>0.257329575</td><td>0.61864850</td></tr>\n\t<tr><th scope=row>hsa00051</th><td>hsa00051</td><td>0.003817844</td><td>0.03722398</td></tr>\n\t<tr><th scope=row>hsa00052</th><td>hsa00052</td><td>0.257329575</td><td>0.61864850</td></tr>\n</tbody>\n</table>\n",
      "text/markdown": "\nA data.frame: 6 × 3\n\n| <!--/--> | pathway &lt;chr&gt; | pvalue &lt;dbl&gt; | pvalue.adj &lt;dbl&gt; |\n|---|---|---|---|\n| hsa00010 | hsa00010 | 0.003514458 | 0.03722398 |\n| hsa00020 | hsa00020 | 0.033191479 | 0.15959191 |\n| hsa00030 | hsa00030 | 0.033191479 | 0.15959191 |\n| hsa00040 | hsa00040 | 0.257329575 | 0.61864850 |\n| hsa00051 | hsa00051 | 0.003817844 | 0.03722398 |\n| hsa00052 | hsa00052 | 0.257329575 | 0.61864850 |\n\n",
      "text/latex": "A data.frame: 6 × 3\n\\begin{tabular}{r|lll}\n  & pathway & pvalue & pvalue.adj\\\\\n  & <chr> & <dbl> & <dbl>\\\\\n\\hline\n\thsa00010 & hsa00010 & 0.003514458 & 0.03722398\\\\\n\thsa00020 & hsa00020 & 0.033191479 & 0.15959191\\\\\n\thsa00030 & hsa00030 & 0.033191479 & 0.15959191\\\\\n\thsa00040 & hsa00040 & 0.257329575 & 0.61864850\\\\\n\thsa00051 & hsa00051 & 0.003817844 & 0.03722398\\\\\n\thsa00052 & hsa00052 & 0.257329575 & 0.61864850\\\\\n\\end{tabular}\n",
      "text/plain": "         pathway  pvalue      pvalue.adj\nhsa00010 hsa00010 0.003514458 0.03722398\nhsa00020 hsa00020 0.033191479 0.15959191\nhsa00030 hsa00030 0.033191479 0.15959191\nhsa00040 hsa00040 0.257329575 0.61864850\nhsa00051 hsa00051 0.003817844 0.03722398\nhsa00052 hsa00052 0.257329575 0.61864850"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "res <- sapply(geneset, function(gs){\n",
    "  # hypergeomestric test\n",
    "  wBallDraw <- intersect(gs, DEGenes) %>%  length() - 1\n",
    "  if (wBallDraw < 0) return(1)\n",
    "  wBall <- length(DEGenes)\n",
    "  bBall <- nrow(data) - length(DEGenes)\n",
    "  ballDraw <- length(intersect(gs, backgroundSet))\n",
    "  1 - phyper(wBallDraw, wBall, bBall, ballDraw)\n",
    "})\n",
    "\n",
    "res.df <- data.frame(\n",
    "  pathway = names(res),\n",
    "  pvalue = res\n",
    ")\n",
    "\n",
    "res.df$pvalue.adj <- p.adjust(res.df$pvalue, method = \"fdr\")\n",
    "res.df <- res.df[order(!res.df$pvalue.adj),]\n",
    "head(res.df)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Enrichment Analysis using FGSEA\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "This submodule describes FGSEA, one of the methods for evaluating pathway enrichment in transcriptional data and it stands for Fast preranked Gene Set Enrichment Analysis (GSEA). FGSEA quickly calculates arbitrarily low GSEA P-values for a collection of gene sets and is based on an algorithm that allows  it to make more permutations and get accurate p-values.\n",
    "It extends the GSEA algorithm which Calculates the Enrichment Score (the degree to which a set S is over-represented at the top or bottom of the ranked list L), estimates the p-value of the Enrichment Score using a permutation test and adjusts the estimated significance level to account for multiple hypothesis in addition to calculating the q-values for FDR control.\n",
    "\n",
    "The FGSEA method call is given by:\n",
    "\n",
    "fgseaRes <- fgsea(pathways = examplePathways,\n",
    "                  stats    = exampleRanks,\n",
    "                  minSize  = 15,\n",
    "                  maxSize  = 500)\n",
    "\n",
    "where:\n",
    "+ pathways = a list of gene sets to check,\n",
    "+ stats    = a named vector of gene-level statistics, where the names should be the same as the gene names in the pathways list,\n",
    "+ minSize and maxSize are the minimum and maximum number of pathways of interest.\n",
    "\n",
    "And the output is a table containing enrichment scores and p-values.\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "outputs": [],
   "source": [
    "# To run FGSEA Enrichment Analysis, we first install the fgsea package\n",
    "suppressMessages({if (!require(\"BiocManager\", quietly = TRUE))\n",
    "        install.packages(\"BiocManager\")\n",
    "  suppressWarnings(BiocManager::install(\"fgsea\", update = F))\n",
    "})"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "outputs": [],
   "source": [
    "# Loading the package\n",
    "suppressPackageStartupMessages({\n",
    "  library(\"fgsea\")\n",
    "})"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "outputs": [
    {
     "data": {
      "text/html": "<pre class=language-r><code>function (..., list = character(), package = NULL, lib.loc = NULL, \n<span style=white-space:pre-wrap>    verbose = getOption(\"verbose\"), envir = .GlobalEnv, overwrite = TRUE) </span>\n{\n<span style=white-space:pre-wrap>    fileExt &lt;- function(x) {</span>\n<span style=white-space:pre-wrap>        db &lt;- grepl(\"\\\\.[^.]+\\\\.(gz|bz2|xz)$\", x)</span>\n<span style=white-space:pre-wrap>        ans &lt;- sub(\".*\\\\.\", \"\", x)</span>\n<span style=white-space:pre-wrap>        ans[db] &lt;- sub(\".*\\\\.([^.]+\\\\.)(gz|bz2|xz)$\", \"\\\\1\\\\2\", </span>\n<span style=white-space:pre-wrap>            x[db])</span>\n<span style=white-space:pre-wrap>        ans</span>\n<span style=white-space:pre-wrap>    }</span>\n<span style=white-space:pre-wrap>    my_read_table &lt;- function(...) {</span>\n<span style=white-space:pre-wrap>        lcc &lt;- Sys.getlocale(\"LC_COLLATE\")</span>\n<span style=white-space:pre-wrap>        on.exit(Sys.setlocale(\"LC_COLLATE\", lcc))</span>\n<span style=white-space:pre-wrap>        Sys.setlocale(\"LC_COLLATE\", \"C\")</span>\n<span style=white-space:pre-wrap>        read.table(...)</span>\n<span style=white-space:pre-wrap>    }</span>\n<span style=white-space:pre-wrap>    stopifnot(is.character(list))</span>\n<span style=white-space:pre-wrap>    names &lt;- c(as.character(substitute(list(...))[-1L]), list)</span>\n<span style=white-space:pre-wrap>    if (!is.null(package)) {</span>\n<span style=white-space:pre-wrap>        if (!is.character(package)) </span>\n<span style=white-space:pre-wrap>            stop(\"'package' must be a character vector or NULL\")</span>\n<span style=white-space:pre-wrap>    }</span>\n<span style=white-space:pre-wrap>    paths &lt;- find.package(package, lib.loc, verbose = verbose)</span>\n<span style=white-space:pre-wrap>    if (is.null(lib.loc)) </span>\n<span style=white-space:pre-wrap>        paths &lt;- c(path.package(package, TRUE), if (!length(package)) getwd(), </span>\n<span style=white-space:pre-wrap>            paths)</span>\n<span style=white-space:pre-wrap>    paths &lt;- unique(normalizePath(paths[file.exists(paths)]))</span>\n<span style=white-space:pre-wrap>    paths &lt;- paths[dir.exists(file.path(paths, \"data\"))]</span>\n<span style=white-space:pre-wrap>    dataExts &lt;- tools:::.make_file_exts(\"data\")</span>\n<span style=white-space:pre-wrap>    if (length(names) == 0L) {</span>\n<span style=white-space:pre-wrap>        db &lt;- matrix(character(), nrow = 0L, ncol = 4L)</span>\n<span style=white-space:pre-wrap>        for (path in paths) {</span>\n<span style=white-space:pre-wrap>            entries &lt;- NULL</span>\n<span style=white-space:pre-wrap>            packageName &lt;- if (file_test(\"-f\", file.path(path, </span>\n<span style=white-space:pre-wrap>                \"DESCRIPTION\"))) </span>\n<span style=white-space:pre-wrap>                basename(path)</span>\n<span style=white-space:pre-wrap>            else \".\"</span>\n<span style=white-space:pre-wrap>            if (file_test(\"-f\", INDEX &lt;- file.path(path, \"Meta\", </span>\n<span style=white-space:pre-wrap>                \"data.rds\"))) {</span>\n<span style=white-space:pre-wrap>                entries &lt;- readRDS(INDEX)</span>\n<span style=white-space:pre-wrap>            }</span>\n<span style=white-space:pre-wrap>            else {</span>\n<span style=white-space:pre-wrap>                dataDir &lt;- file.path(path, \"data\")</span>\n<span style=white-space:pre-wrap>                entries &lt;- tools::list_files_with_type(dataDir, </span>\n<span style=white-space:pre-wrap>                  \"data\")</span>\n<span style=white-space:pre-wrap>                if (length(entries)) {</span>\n<span style=white-space:pre-wrap>                  entries &lt;- unique(tools::file_path_sans_ext(basename(entries)))</span>\n<span style=white-space:pre-wrap>                  entries &lt;- cbind(entries, \"\")</span>\n<span style=white-space:pre-wrap>                }</span>\n<span style=white-space:pre-wrap>            }</span>\n<span style=white-space:pre-wrap>            if (NROW(entries)) {</span>\n<span style=white-space:pre-wrap>                if (is.matrix(entries) &amp;&amp; ncol(entries) == 2L) </span>\n<span style=white-space:pre-wrap>                  db &lt;- rbind(db, cbind(packageName, dirname(path), </span>\n<span style=white-space:pre-wrap>                    entries))</span>\n<span style=white-space:pre-wrap>                else warning(gettextf(\"data index for package %s is invalid and will be ignored\", </span>\n<span style=white-space:pre-wrap>                  sQuote(packageName)), domain = NA, call. = FALSE)</span>\n<span style=white-space:pre-wrap>            }</span>\n<span style=white-space:pre-wrap>        }</span>\n<span style=white-space:pre-wrap>        colnames(db) &lt;- c(\"Package\", \"LibPath\", \"Item\", \"Title\")</span>\n<span style=white-space:pre-wrap>        footer &lt;- if (missing(package)) </span>\n<span style=white-space:pre-wrap>            paste0(\"Use \", sQuote(paste(\"data(package =\", \".packages(all.available = TRUE))\")), </span>\n<span style=white-space:pre-wrap>                \"\\n\", \"to list the data sets in all *available* packages.\")</span>\n<span style=white-space:pre-wrap>        else NULL</span>\n<span style=white-space:pre-wrap>        y &lt;- list(title = \"Data sets\", header = NULL, results = db, </span>\n<span style=white-space:pre-wrap>            footer = footer)</span>\n<span style=white-space:pre-wrap>        class(y) &lt;- \"packageIQR\"</span>\n<span style=white-space:pre-wrap>        return(y)</span>\n<span style=white-space:pre-wrap>    }</span>\n<span style=white-space:pre-wrap>    paths &lt;- file.path(paths, \"data\")</span>\n<span style=white-space:pre-wrap>    for (name in names) {</span>\n<span style=white-space:pre-wrap>        found &lt;- FALSE</span>\n<span style=white-space:pre-wrap>        for (p in paths) {</span>\n<span style=white-space:pre-wrap>            tmp_env &lt;- if (overwrite) </span>\n<span style=white-space:pre-wrap>                envir</span>\n<span style=white-space:pre-wrap>            else new.env()</span>\n<span style=white-space:pre-wrap>            if (file_test(\"-f\", file.path(p, \"Rdata.rds\"))) {</span>\n<span style=white-space:pre-wrap>                rds &lt;- readRDS(file.path(p, \"Rdata.rds\"))</span>\n<span style=white-space:pre-wrap>                if (name %in% names(rds)) {</span>\n<span style=white-space:pre-wrap>                  found &lt;- TRUE</span>\n<span style=white-space:pre-wrap>                  if (verbose) </span>\n<span style=white-space:pre-wrap>                    message(sprintf(\"name=%s:\\t found in Rdata.rds\", </span>\n<span style=white-space:pre-wrap>                      name), domain = NA)</span>\n<span style=white-space:pre-wrap>                  thispkg &lt;- sub(\".*/([^/]*)/data$\", \"\\\\1\", p)</span>\n<span style=white-space:pre-wrap>                  thispkg &lt;- sub(\"_.*$\", \"\", thispkg)</span>\n<span style=white-space:pre-wrap>                  thispkg &lt;- paste0(\"package:\", thispkg)</span>\n<span style=white-space:pre-wrap>                  objs &lt;- rds[[name]]</span>\n<span style=white-space:pre-wrap>                  lazyLoad(file.path(p, \"Rdata\"), envir = tmp_env, </span>\n<span style=white-space:pre-wrap>                    filter = function(x) x %in% objs)</span>\n<span style=white-space:pre-wrap>                  break</span>\n<span style=white-space:pre-wrap>                }</span>\n<span style=white-space:pre-wrap>                else if (verbose) </span>\n<span style=white-space:pre-wrap>                  message(sprintf(\"name=%s:\\t NOT found in names() of Rdata.rds, i.e.,\\n\\t%s\\n\", </span>\n<span style=white-space:pre-wrap>                    name, paste(names(rds), collapse = \",\")), </span>\n<span style=white-space:pre-wrap>                    domain = NA)</span>\n<span style=white-space:pre-wrap>            }</span>\n<span style=white-space:pre-wrap>            if (file_test(\"-f\", file.path(p, \"Rdata.zip\"))) {</span>\n<span style=white-space:pre-wrap>                warning(\"zipped data found for package \", sQuote(basename(dirname(p))), </span>\n<span style=white-space:pre-wrap>                  \".\\nThat is defunct, so please re-install the package.\", </span>\n<span style=white-space:pre-wrap>                  domain = NA)</span>\n<span style=white-space:pre-wrap>                if (file_test(\"-f\", fp &lt;- file.path(p, \"filelist\"))) </span>\n<span style=white-space:pre-wrap>                  files &lt;- file.path(p, scan(fp, what = \"\", quiet = TRUE))</span>\n<span style=white-space:pre-wrap>                else {</span>\n<span style=white-space:pre-wrap>                  warning(gettextf(\"file 'filelist' is missing for directory %s\", </span>\n<span style=white-space:pre-wrap>                    sQuote(p)), domain = NA)</span>\n<span style=white-space:pre-wrap>                  next</span>\n<span style=white-space:pre-wrap>                }</span>\n<span style=white-space:pre-wrap>            }</span>\n<span style=white-space:pre-wrap>            else {</span>\n<span style=white-space:pre-wrap>                files &lt;- list.files(p, full.names = TRUE)</span>\n<span style=white-space:pre-wrap>            }</span>\n<span style=white-space:pre-wrap>            files &lt;- files[grep(name, files, fixed = TRUE)]</span>\n<span style=white-space:pre-wrap>            if (length(files) &gt; 1L) {</span>\n<span style=white-space:pre-wrap>                o &lt;- match(fileExt(files), dataExts, nomatch = 100L)</span>\n<span style=white-space:pre-wrap>                paths0 &lt;- dirname(files)</span>\n<span style=white-space:pre-wrap>                paths0 &lt;- factor(paths0, levels = unique(paths0))</span>\n<span style=white-space:pre-wrap>                files &lt;- files[order(paths0, o)]</span>\n<span style=white-space:pre-wrap>            }</span>\n<span style=white-space:pre-wrap>            if (length(files)) {</span>\n<span style=white-space:pre-wrap>                for (file in files) {</span>\n<span style=white-space:pre-wrap>                  if (verbose) </span>\n<span style=white-space:pre-wrap>                    message(\"name=\", name, \":\\t file= ...\", .Platform$file.sep, </span>\n<span style=white-space:pre-wrap>                      basename(file), \"::\\t\", appendLF = FALSE, </span>\n<span style=white-space:pre-wrap>                      domain = NA)</span>\n<span style=white-space:pre-wrap>                  ext &lt;- fileExt(file)</span>\n<span style=white-space:pre-wrap>                  if (basename(file) != paste0(name, \".\", ext)) </span>\n<span style=white-space:pre-wrap>                    found &lt;- FALSE</span>\n<span style=white-space:pre-wrap>                  else {</span>\n<span style=white-space:pre-wrap>                    found &lt;- TRUE</span>\n<span style=white-space:pre-wrap>                    zfile &lt;- file</span>\n<span style=white-space:pre-wrap>                    zipname &lt;- file.path(dirname(file), \"Rdata.zip\")</span>\n<span style=white-space:pre-wrap>                    if (file.exists(zipname)) {</span>\n<span style=white-space:pre-wrap>                      Rdatadir &lt;- tempfile(\"Rdata\")</span>\n<span style=white-space:pre-wrap>                      dir.create(Rdatadir, showWarnings = FALSE)</span>\n<span style=white-space:pre-wrap>                      topic &lt;- basename(file)</span>\n<span style=white-space:pre-wrap>                      rc &lt;- .External(C_unzip, zipname, topic, </span>\n<span style=white-space:pre-wrap>                        Rdatadir, FALSE, TRUE, FALSE, FALSE)</span>\n<span style=white-space:pre-wrap>                      if (rc == 0L) </span>\n<span style=white-space:pre-wrap>                        zfile &lt;- file.path(Rdatadir, topic)</span>\n<span style=white-space:pre-wrap>                    }</span>\n<span style=white-space:pre-wrap>                    if (zfile != file) </span>\n<span style=white-space:pre-wrap>                      on.exit(unlink(zfile))</span>\n<span style=white-space:pre-wrap>                    switch(ext, R = , r = {</span>\n<span style=white-space:pre-wrap>                      library(\"utils\")</span>\n<span style=white-space:pre-wrap>                      sys.source(zfile, chdir = TRUE, envir = tmp_env)</span>\n<span style=white-space:pre-wrap>                    }, RData = , rdata = , rda = load(zfile, </span>\n<span style=white-space:pre-wrap>                      envir = tmp_env), TXT = , txt = , tab = , </span>\n<span style=white-space:pre-wrap>                      tab.gz = , tab.bz2 = , tab.xz = , txt.gz = , </span>\n<span style=white-space:pre-wrap>                      txt.bz2 = , txt.xz = assign(name, my_read_table(zfile, </span>\n<span style=white-space:pre-wrap>                        header = TRUE, as.is = FALSE), envir = tmp_env), </span>\n<span style=white-space:pre-wrap>                      CSV = , csv = , csv.gz = , csv.bz2 = , </span>\n<span style=white-space:pre-wrap>                      csv.xz = assign(name, my_read_table(zfile, </span>\n<span style=white-space:pre-wrap>                        header = TRUE, sep = \";\", as.is = FALSE), </span>\n<span style=white-space:pre-wrap>                        envir = tmp_env), found &lt;- FALSE)</span>\n<span style=white-space:pre-wrap>                  }</span>\n<span style=white-space:pre-wrap>                  if (found) </span>\n<span style=white-space:pre-wrap>                    break</span>\n<span style=white-space:pre-wrap>                }</span>\n<span style=white-space:pre-wrap>                if (verbose) </span>\n<span style=white-space:pre-wrap>                  message(if (!found) </span>\n<span style=white-space:pre-wrap>                    \"*NOT* \", \"found\", domain = NA)</span>\n<span style=white-space:pre-wrap>            }</span>\n<span style=white-space:pre-wrap>            if (found) </span>\n<span style=white-space:pre-wrap>                break</span>\n<span style=white-space:pre-wrap>        }</span>\n<span style=white-space:pre-wrap>        if (!found) {</span>\n<span style=white-space:pre-wrap>            warning(gettextf(\"data set %s not found\", sQuote(name)), </span>\n<span style=white-space:pre-wrap>                domain = NA)</span>\n<span style=white-space:pre-wrap>        }</span>\n<span style=white-space:pre-wrap>        else if (!overwrite) {</span>\n<span style=white-space:pre-wrap>            for (o in ls(envir = tmp_env, all.names = TRUE)) {</span>\n<span style=white-space:pre-wrap>                if (exists(o, envir = envir, inherits = FALSE)) </span>\n<span style=white-space:pre-wrap>                  warning(gettextf(\"an object named %s already exists and will not be overwritten\", </span>\n<span style=white-space:pre-wrap>                    sQuote(o)))</span>\n<span style=white-space:pre-wrap>                else assign(o, get(o, envir = tmp_env, inherits = FALSE), </span>\n<span style=white-space:pre-wrap>                  envir = envir)</span>\n<span style=white-space:pre-wrap>            }</span>\n<span style=white-space:pre-wrap>            rm(tmp_env)</span>\n<span style=white-space:pre-wrap>        }</span>\n<span style=white-space:pre-wrap>    }</span>\n<span style=white-space:pre-wrap>    invisible(names)</span>\n}</code></pre>",
      "text/markdown": "```r\nfunction (..., list = character(), package = NULL, lib.loc = NULL, \n    verbose = getOption(\"verbose\"), envir = .GlobalEnv, overwrite = TRUE) \n{\n    fileExt <- function(x) {\n        db <- grepl(\"\\\\.[^.]+\\\\.(gz|bz2|xz)$\", x)\n        ans <- sub(\".*\\\\.\", \"\", x)\n        ans[db] <- sub(\".*\\\\.([^.]+\\\\.)(gz|bz2|xz)$\", \"\\\\1\\\\2\", \n            x[db])\n        ans\n    }\n    my_read_table <- function(...) {\n        lcc <- Sys.getlocale(\"LC_COLLATE\")\n        on.exit(Sys.setlocale(\"LC_COLLATE\", lcc))\n        Sys.setlocale(\"LC_COLLATE\", \"C\")\n        read.table(...)\n    }\n    stopifnot(is.character(list))\n    names <- c(as.character(substitute(list(...))[-1L]), list)\n    if (!is.null(package)) {\n        if (!is.character(package)) \n            stop(\"'package' must be a character vector or NULL\")\n    }\n    paths <- find.package(package, lib.loc, verbose = verbose)\n    if (is.null(lib.loc)) \n        paths <- c(path.package(package, TRUE), if (!length(package)) getwd(), \n            paths)\n    paths <- unique(normalizePath(paths[file.exists(paths)]))\n    paths <- paths[dir.exists(file.path(paths, \"data\"))]\n    dataExts <- tools:::.make_file_exts(\"data\")\n    if (length(names) == 0L) {\n        db <- matrix(character(), nrow = 0L, ncol = 4L)\n        for (path in paths) {\n            entries <- NULL\n            packageName <- if (file_test(\"-f\", file.path(path, \n                \"DESCRIPTION\"))) \n                basename(path)\n            else \".\"\n            if (file_test(\"-f\", INDEX <- file.path(path, \"Meta\", \n                \"data.rds\"))) {\n                entries <- readRDS(INDEX)\n            }\n            else {\n                dataDir <- file.path(path, \"data\")\n                entries <- tools::list_files_with_type(dataDir, \n                  \"data\")\n                if (length(entries)) {\n                  entries <- unique(tools::file_path_sans_ext(basename(entries)))\n                  entries <- cbind(entries, \"\")\n                }\n            }\n            if (NROW(entries)) {\n                if (is.matrix(entries) && ncol(entries) == 2L) \n                  db <- rbind(db, cbind(packageName, dirname(path), \n                    entries))\n                else warning(gettextf(\"data index for package %s is invalid and will be ignored\", \n                  sQuote(packageName)), domain = NA, call. = FALSE)\n            }\n        }\n        colnames(db) <- c(\"Package\", \"LibPath\", \"Item\", \"Title\")\n        footer <- if (missing(package)) \n            paste0(\"Use \", sQuote(paste(\"data(package =\", \".packages(all.available = TRUE))\")), \n                \"\\n\", \"to list the data sets in all *available* packages.\")\n        else NULL\n        y <- list(title = \"Data sets\", header = NULL, results = db, \n            footer = footer)\n        class(y) <- \"packageIQR\"\n        return(y)\n    }\n    paths <- file.path(paths, \"data\")\n    for (name in names) {\n        found <- FALSE\n        for (p in paths) {\n            tmp_env <- if (overwrite) \n                envir\n            else new.env()\n            if (file_test(\"-f\", file.path(p, \"Rdata.rds\"))) {\n                rds <- readRDS(file.path(p, \"Rdata.rds\"))\n                if (name %in% names(rds)) {\n                  found <- TRUE\n                  if (verbose) \n                    message(sprintf(\"name=%s:\\t found in Rdata.rds\", \n                      name), domain = NA)\n                  thispkg <- sub(\".*/([^/]*)/data$\", \"\\\\1\", p)\n                  thispkg <- sub(\"_.*$\", \"\", thispkg)\n                  thispkg <- paste0(\"package:\", thispkg)\n                  objs <- rds[[name]]\n                  lazyLoad(file.path(p, \"Rdata\"), envir = tmp_env, \n                    filter = function(x) x %in% objs)\n                  break\n                }\n                else if (verbose) \n                  message(sprintf(\"name=%s:\\t NOT found in names() of Rdata.rds, i.e.,\\n\\t%s\\n\", \n                    name, paste(names(rds), collapse = \",\")), \n                    domain = NA)\n            }\n            if (file_test(\"-f\", file.path(p, \"Rdata.zip\"))) {\n                warning(\"zipped data found for package \", sQuote(basename(dirname(p))), \n                  \".\\nThat is defunct, so please re-install the package.\", \n                  domain = NA)\n                if (file_test(\"-f\", fp <- file.path(p, \"filelist\"))) \n                  files <- file.path(p, scan(fp, what = \"\", quiet = TRUE))\n                else {\n                  warning(gettextf(\"file 'filelist' is missing for directory %s\", \n                    sQuote(p)), domain = NA)\n                  next\n                }\n            }\n            else {\n                files <- list.files(p, full.names = TRUE)\n            }\n            files <- files[grep(name, files, fixed = TRUE)]\n            if (length(files) > 1L) {\n                o <- match(fileExt(files), dataExts, nomatch = 100L)\n                paths0 <- dirname(files)\n                paths0 <- factor(paths0, levels = unique(paths0))\n                files <- files[order(paths0, o)]\n            }\n            if (length(files)) {\n                for (file in files) {\n                  if (verbose) \n                    message(\"name=\", name, \":\\t file= ...\", .Platform$file.sep, \n                      basename(file), \"::\\t\", appendLF = FALSE, \n                      domain = NA)\n                  ext <- fileExt(file)\n                  if (basename(file) != paste0(name, \".\", ext)) \n                    found <- FALSE\n                  else {\n                    found <- TRUE\n                    zfile <- file\n                    zipname <- file.path(dirname(file), \"Rdata.zip\")\n                    if (file.exists(zipname)) {\n                      Rdatadir <- tempfile(\"Rdata\")\n                      dir.create(Rdatadir, showWarnings = FALSE)\n                      topic <- basename(file)\n                      rc <- .External(C_unzip, zipname, topic, \n                        Rdatadir, FALSE, TRUE, FALSE, FALSE)\n                      if (rc == 0L) \n                        zfile <- file.path(Rdatadir, topic)\n                    }\n                    if (zfile != file) \n                      on.exit(unlink(zfile))\n                    switch(ext, R = , r = {\n                      library(\"utils\")\n                      sys.source(zfile, chdir = TRUE, envir = tmp_env)\n                    }, RData = , rdata = , rda = load(zfile, \n                      envir = tmp_env), TXT = , txt = , tab = , \n                      tab.gz = , tab.bz2 = , tab.xz = , txt.gz = , \n                      txt.bz2 = , txt.xz = assign(name, my_read_table(zfile, \n                        header = TRUE, as.is = FALSE), envir = tmp_env), \n                      CSV = , csv = , csv.gz = , csv.bz2 = , \n                      csv.xz = assign(name, my_read_table(zfile, \n                        header = TRUE, sep = \";\", as.is = FALSE), \n                        envir = tmp_env), found <- FALSE)\n                  }\n                  if (found) \n                    break\n                }\n                if (verbose) \n                  message(if (!found) \n                    \"*NOT* \", \"found\", domain = NA)\n            }\n            if (found) \n                break\n        }\n        if (!found) {\n            warning(gettextf(\"data set %s not found\", sQuote(name)), \n                domain = NA)\n        }\n        else if (!overwrite) {\n            for (o in ls(envir = tmp_env, all.names = TRUE)) {\n                if (exists(o, envir = envir, inherits = FALSE)) \n                  warning(gettextf(\"an object named %s already exists and will not be overwritten\", \n                    sQuote(o)))\n                else assign(o, get(o, envir = tmp_env, inherits = FALSE), \n                  envir = envir)\n            }\n            rm(tmp_env)\n        }\n    }\n    invisible(names)\n}\n```",
      "text/latex": "\\begin{minted}{r}\nfunction (..., list = character(), package = NULL, lib.loc = NULL, \n    verbose = getOption(\"verbose\"), envir = .GlobalEnv, overwrite = TRUE) \n\\{\n    fileExt <- function(x) \\{\n        db <- grepl(\"\\textbackslash{}\\textbackslash{}.{[}\\textasciicircum{}.{]}+\\textbackslash{}\\textbackslash{}.(gz\\textbar{}bz2\\textbar{}xz)\\$\", x)\n        ans <- sub(\".*\\textbackslash{}\\textbackslash{}.\", \"\", x)\n        ans{[}db{]} <- sub(\".*\\textbackslash{}\\textbackslash{}.({[}\\textasciicircum{}.{]}+\\textbackslash{}\\textbackslash{}.)(gz\\textbar{}bz2\\textbar{}xz)\\$\", \"\\textbackslash{}\\textbackslash{}1\\textbackslash{}\\textbackslash{}2\", \n            x{[}db{]})\n        ans\n    \\}\n    my\\_read\\_table <- function(...) \\{\n        lcc <- Sys.getlocale(\"LC\\_COLLATE\")\n        on.exit(Sys.setlocale(\"LC\\_COLLATE\", lcc))\n        Sys.setlocale(\"LC\\_COLLATE\", \"C\")\n        read.table(...)\n    \\}\n    stopifnot(is.character(list))\n    names <- c(as.character(substitute(list(...)){[}-1L{]}), list)\n    if (!is.null(package)) \\{\n        if (!is.character(package)) \n            stop(\"'package' must be a character vector or NULL\")\n    \\}\n    paths <- find.package(package, lib.loc, verbose = verbose)\n    if (is.null(lib.loc)) \n        paths <- c(path.package(package, TRUE), if (!length(package)) getwd(), \n            paths)\n    paths <- unique(normalizePath(paths{[}file.exists(paths){]}))\n    paths <- paths{[}dir.exists(file.path(paths, \"data\")){]}\n    dataExts <- tools:::.make\\_file\\_exts(\"data\")\n    if (length(names) == 0L) \\{\n        db <- matrix(character(), nrow = 0L, ncol = 4L)\n        for (path in paths) \\{\n            entries <- NULL\n            packageName <- if (file\\_test(\"-f\", file.path(path, \n                \"DESCRIPTION\"))) \n                basename(path)\n            else \".\"\n            if (file\\_test(\"-f\", INDEX <- file.path(path, \"Meta\", \n                \"data.rds\"))) \\{\n                entries <- readRDS(INDEX)\n            \\}\n            else \\{\n                dataDir <- file.path(path, \"data\")\n                entries <- tools::list\\_files\\_with\\_type(dataDir, \n                  \"data\")\n                if (length(entries)) \\{\n                  entries <- unique(tools::file\\_path\\_sans\\_ext(basename(entries)))\n                  entries <- cbind(entries, \"\")\n                \\}\n            \\}\n            if (NROW(entries)) \\{\n                if (is.matrix(entries) \\&\\& ncol(entries) == 2L) \n                  db <- rbind(db, cbind(packageName, dirname(path), \n                    entries))\n                else warning(gettextf(\"data index for package \\%s is invalid and will be ignored\", \n                  sQuote(packageName)), domain = NA, call. = FALSE)\n            \\}\n        \\}\n        colnames(db) <- c(\"Package\", \"LibPath\", \"Item\", \"Title\")\n        footer <- if (missing(package)) \n            paste0(\"Use \", sQuote(paste(\"data(package =\", \".packages(all.available = TRUE))\")), \n                \"\\textbackslash{}n\", \"to list the data sets in all *available* packages.\")\n        else NULL\n        y <- list(title = \"Data sets\", header = NULL, results = db, \n            footer = footer)\n        class(y) <- \"packageIQR\"\n        return(y)\n    \\}\n    paths <- file.path(paths, \"data\")\n    for (name in names) \\{\n        found <- FALSE\n        for (p in paths) \\{\n            tmp\\_env <- if (overwrite) \n                envir\n            else new.env()\n            if (file\\_test(\"-f\", file.path(p, \"Rdata.rds\"))) \\{\n                rds <- readRDS(file.path(p, \"Rdata.rds\"))\n                if (name \\%in\\% names(rds)) \\{\n                  found <- TRUE\n                  if (verbose) \n                    message(sprintf(\"name=\\%s:\\textbackslash{}t found in Rdata.rds\", \n                      name), domain = NA)\n                  thispkg <- sub(\".*/({[}\\textasciicircum{}/{]}*)/data\\$\", \"\\textbackslash{}\\textbackslash{}1\", p)\n                  thispkg <- sub(\"\\_.*\\$\", \"\", thispkg)\n                  thispkg <- paste0(\"package:\", thispkg)\n                  objs <- rds{[}{[}name{]}{]}\n                  lazyLoad(file.path(p, \"Rdata\"), envir = tmp\\_env, \n                    filter = function(x) x \\%in\\% objs)\n                  break\n                \\}\n                else if (verbose) \n                  message(sprintf(\"name=\\%s:\\textbackslash{}t NOT found in names() of Rdata.rds, i.e.,\\textbackslash{}n\\textbackslash{}t\\%s\\textbackslash{}n\", \n                    name, paste(names(rds), collapse = \",\")), \n                    domain = NA)\n            \\}\n            if (file\\_test(\"-f\", file.path(p, \"Rdata.zip\"))) \\{\n                warning(\"zipped data found for package \", sQuote(basename(dirname(p))), \n                  \".\\textbackslash{}nThat is defunct, so please re-install the package.\", \n                  domain = NA)\n                if (file\\_test(\"-f\", fp <- file.path(p, \"filelist\"))) \n                  files <- file.path(p, scan(fp, what = \"\", quiet = TRUE))\n                else \\{\n                  warning(gettextf(\"file 'filelist' is missing for directory \\%s\", \n                    sQuote(p)), domain = NA)\n                  next\n                \\}\n            \\}\n            else \\{\n                files <- list.files(p, full.names = TRUE)\n            \\}\n            files <- files{[}grep(name, files, fixed = TRUE){]}\n            if (length(files) > 1L) \\{\n                o <- match(fileExt(files), dataExts, nomatch = 100L)\n                paths0 <- dirname(files)\n                paths0 <- factor(paths0, levels = unique(paths0))\n                files <- files{[}order(paths0, o){]}\n            \\}\n            if (length(files)) \\{\n                for (file in files) \\{\n                  if (verbose) \n                    message(\"name=\", name, \":\\textbackslash{}t file= ...\", .Platform\\$file.sep, \n                      basename(file), \"::\\textbackslash{}t\", appendLF = FALSE, \n                      domain = NA)\n                  ext <- fileExt(file)\n                  if (basename(file) != paste0(name, \".\", ext)) \n                    found <- FALSE\n                  else \\{\n                    found <- TRUE\n                    zfile <- file\n                    zipname <- file.path(dirname(file), \"Rdata.zip\")\n                    if (file.exists(zipname)) \\{\n                      Rdatadir <- tempfile(\"Rdata\")\n                      dir.create(Rdatadir, showWarnings = FALSE)\n                      topic <- basename(file)\n                      rc <- .External(C\\_unzip, zipname, topic, \n                        Rdatadir, FALSE, TRUE, FALSE, FALSE)\n                      if (rc == 0L) \n                        zfile <- file.path(Rdatadir, topic)\n                    \\}\n                    if (zfile != file) \n                      on.exit(unlink(zfile))\n                    switch(ext, R = , r = \\{\n                      library(\"utils\")\n                      sys.source(zfile, chdir = TRUE, envir = tmp\\_env)\n                    \\}, RData = , rdata = , rda = load(zfile, \n                      envir = tmp\\_env), TXT = , txt = , tab = , \n                      tab.gz = , tab.bz2 = , tab.xz = , txt.gz = , \n                      txt.bz2 = , txt.xz = assign(name, my\\_read\\_table(zfile, \n                        header = TRUE, as.is = FALSE), envir = tmp\\_env), \n                      CSV = , csv = , csv.gz = , csv.bz2 = , \n                      csv.xz = assign(name, my\\_read\\_table(zfile, \n                        header = TRUE, sep = \";\", as.is = FALSE), \n                        envir = tmp\\_env), found <- FALSE)\n                  \\}\n                  if (found) \n                    break\n                \\}\n                if (verbose) \n                  message(if (!found) \n                    \"*NOT* \", \"found\", domain = NA)\n            \\}\n            if (found) \n                break\n        \\}\n        if (!found) \\{\n            warning(gettextf(\"data set \\%s not found\", sQuote(name)), \n                domain = NA)\n        \\}\n        else if (!overwrite) \\{\n            for (o in ls(envir = tmp\\_env, all.names = TRUE)) \\{\n                if (exists(o, envir = envir, inherits = FALSE)) \n                  warning(gettextf(\"an object named \\%s already exists and will not be overwritten\", \n                    sQuote(o)))\n                else assign(o, get(o, envir = tmp\\_env, inherits = FALSE), \n                  envir = envir)\n            \\}\n            rm(tmp\\_env)\n        \\}\n    \\}\n    invisible(names)\n\\}\n\\end{minted}",
      "text/plain": "function (..., list = character(), package = NULL, lib.loc = NULL, \n    verbose = getOption(\"verbose\"), envir = .GlobalEnv, overwrite = TRUE) \n{\n    fileExt <- function(x) {\n        db <- grepl(\"\\\\.[^.]+\\\\.(gz|bz2|xz)$\", x)\n        ans <- sub(\".*\\\\.\", \"\", x)\n        ans[db] <- sub(\".*\\\\.([^.]+\\\\.)(gz|bz2|xz)$\", \"\\\\1\\\\2\", \n            x[db])\n        ans\n    }\n    my_read_table <- function(...) {\n        lcc <- Sys.getlocale(\"LC_COLLATE\")\n        on.exit(Sys.setlocale(\"LC_COLLATE\", lcc))\n        Sys.setlocale(\"LC_COLLATE\", \"C\")\n        read.table(...)\n    }\n    stopifnot(is.character(list))\n    names <- c(as.character(substitute(list(...))[-1L]), list)\n    if (!is.null(package)) {\n        if (!is.character(package)) \n            stop(\"'package' must be a character vector or NULL\")\n    }\n    paths <- find.package(package, lib.loc, verbose = verbose)\n    if (is.null(lib.loc)) \n        paths <- c(path.package(package, TRUE), if (!length(package)) getwd(), \n            paths)\n    paths <- unique(normalizePath(paths[file.exists(paths)]))\n    paths <- paths[dir.exists(file.path(paths, \"data\"))]\n    dataExts <- tools:::.make_file_exts(\"data\")\n    if (length(names) == 0L) {\n        db <- matrix(character(), nrow = 0L, ncol = 4L)\n        for (path in paths) {\n            entries <- NULL\n            packageName <- if (file_test(\"-f\", file.path(path, \n                \"DESCRIPTION\"))) \n                basename(path)\n            else \".\"\n            if (file_test(\"-f\", INDEX <- file.path(path, \"Meta\", \n                \"data.rds\"))) {\n                entries <- readRDS(INDEX)\n            }\n            else {\n                dataDir <- file.path(path, \"data\")\n                entries <- tools::list_files_with_type(dataDir, \n                  \"data\")\n                if (length(entries)) {\n                  entries <- unique(tools::file_path_sans_ext(basename(entries)))\n                  entries <- cbind(entries, \"\")\n                }\n            }\n            if (NROW(entries)) {\n                if (is.matrix(entries) && ncol(entries) == 2L) \n                  db <- rbind(db, cbind(packageName, dirname(path), \n                    entries))\n                else warning(gettextf(\"data index for package %s is invalid and will be ignored\", \n                  sQuote(packageName)), domain = NA, call. = FALSE)\n            }\n        }\n        colnames(db) <- c(\"Package\", \"LibPath\", \"Item\", \"Title\")\n        footer <- if (missing(package)) \n            paste0(\"Use \", sQuote(paste(\"data(package =\", \".packages(all.available = TRUE))\")), \n                \"\\n\", \"to list the data sets in all *available* packages.\")\n        else NULL\n        y <- list(title = \"Data sets\", header = NULL, results = db, \n            footer = footer)\n        class(y) <- \"packageIQR\"\n        return(y)\n    }\n    paths <- file.path(paths, \"data\")\n    for (name in names) {\n        found <- FALSE\n        for (p in paths) {\n            tmp_env <- if (overwrite) \n                envir\n            else new.env()\n            if (file_test(\"-f\", file.path(p, \"Rdata.rds\"))) {\n                rds <- readRDS(file.path(p, \"Rdata.rds\"))\n                if (name %in% names(rds)) {\n                  found <- TRUE\n                  if (verbose) \n                    message(sprintf(\"name=%s:\\t found in Rdata.rds\", \n                      name), domain = NA)\n                  thispkg <- sub(\".*/([^/]*)/data$\", \"\\\\1\", p)\n                  thispkg <- sub(\"_.*$\", \"\", thispkg)\n                  thispkg <- paste0(\"package:\", thispkg)\n                  objs <- rds[[name]]\n                  lazyLoad(file.path(p, \"Rdata\"), envir = tmp_env, \n                    filter = function(x) x %in% objs)\n                  break\n                }\n                else if (verbose) \n                  message(sprintf(\"name=%s:\\t NOT found in names() of Rdata.rds, i.e.,\\n\\t%s\\n\", \n                    name, paste(names(rds), collapse = \",\")), \n                    domain = NA)\n            }\n            if (file_test(\"-f\", file.path(p, \"Rdata.zip\"))) {\n                warning(\"zipped data found for package \", sQuote(basename(dirname(p))), \n                  \".\\nThat is defunct, so please re-install the package.\", \n                  domain = NA)\n                if (file_test(\"-f\", fp <- file.path(p, \"filelist\"))) \n                  files <- file.path(p, scan(fp, what = \"\", quiet = TRUE))\n                else {\n                  warning(gettextf(\"file 'filelist' is missing for directory %s\", \n                    sQuote(p)), domain = NA)\n                  next\n                }\n            }\n            else {\n                files <- list.files(p, full.names = TRUE)\n            }\n            files <- files[grep(name, files, fixed = TRUE)]\n            if (length(files) > 1L) {\n                o <- match(fileExt(files), dataExts, nomatch = 100L)\n                paths0 <- dirname(files)\n                paths0 <- factor(paths0, levels = unique(paths0))\n                files <- files[order(paths0, o)]\n            }\n            if (length(files)) {\n                for (file in files) {\n                  if (verbose) \n                    message(\"name=\", name, \":\\t file= ...\", .Platform$file.sep, \n                      basename(file), \"::\\t\", appendLF = FALSE, \n                      domain = NA)\n                  ext <- fileExt(file)\n                  if (basename(file) != paste0(name, \".\", ext)) \n                    found <- FALSE\n                  else {\n                    found <- TRUE\n                    zfile <- file\n                    zipname <- file.path(dirname(file), \"Rdata.zip\")\n                    if (file.exists(zipname)) {\n                      Rdatadir <- tempfile(\"Rdata\")\n                      dir.create(Rdatadir, showWarnings = FALSE)\n                      topic <- basename(file)\n                      rc <- .External(C_unzip, zipname, topic, \n                        Rdatadir, FALSE, TRUE, FALSE, FALSE)\n                      if (rc == 0L) \n                        zfile <- file.path(Rdatadir, topic)\n                    }\n                    if (zfile != file) \n                      on.exit(unlink(zfile))\n                    switch(ext, R = , r = {\n                      library(\"utils\")\n                      sys.source(zfile, chdir = TRUE, envir = tmp_env)\n                    }, RData = , rdata = , rda = load(zfile, \n                      envir = tmp_env), TXT = , txt = , tab = , \n                      tab.gz = , tab.bz2 = , tab.xz = , txt.gz = , \n                      txt.bz2 = , txt.xz = assign(name, my_read_table(zfile, \n                        header = TRUE, as.is = FALSE), envir = tmp_env), \n                      CSV = , csv = , csv.gz = , csv.bz2 = , \n                      csv.xz = assign(name, my_read_table(zfile, \n                        header = TRUE, sep = \";\", as.is = FALSE), \n                        envir = tmp_env), found <- FALSE)\n                  }\n                  if (found) \n                    break\n                }\n                if (verbose) \n                  message(if (!found) \n                    \"*NOT* \", \"found\", domain = NA)\n            }\n            if (found) \n                break\n        }\n        if (!found) {\n            warning(gettextf(\"data set %s not found\", sQuote(name)), \n                domain = NA)\n        }\n        else if (!overwrite) {\n            for (o in ls(envir = tmp_env, all.names = TRUE)) {\n                if (exists(o, envir = envir, inherits = FALSE)) \n                  warning(gettextf(\"an object named %s already exists and will not be overwritten\", \n                    sQuote(o)))\n                else assign(o, get(o, envir = tmp_env, inherits = FALSE), \n                  envir = envir)\n            }\n            rm(tmp_env)\n        }\n    }\n    invisible(names)\n}\n<bytecode: 0x00000000155d4f68>\n<environment: namespace:utils>"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "ERROR",
     "evalue": "Error in data$adj.P.Val: object of type 'closure' is not subsettable\n",
     "output_type": "error",
     "traceback": [
      "Error in data$adj.P.Val: object of type 'closure' is not subsettable\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "# Get the gene list and their statistic from DE results\n",
    "data\n",
    "stats = data$adj.P.Val\n",
    "names(stats) = data$SYMBOL\n",
    "head(stats)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "### FGSEA Using GO"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "The package `fgsea` has its own function named `gmtPathways` to load the genesets from the `gmt` file. It is recommended to use this function for compatibility."
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "outputs": [
    {
     "data": {
      "text/html": "<dl>\n\t<dt>$`GO:0000002`</dt>\n\t\t<dd><style>\n.list-inline {list-style: none; margin:0; padding: 0}\n.list-inline>li {display: inline-block}\n.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n</style>\n<ol class=list-inline><li>'AKT3'</li><li>'DNA2'</li><li>'LIG3'</li><li>'LONP1'</li><li>'MEF2A'</li><li>'MGME1'</li></ol>\n</dd>\n\t<dt>$`GO:0000003`</dt>\n\t\t<dd><style>\n.list-inline {list-style: none; margin:0; padding: 0}\n.list-inline>li {display: inline-block}\n.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n</style>\n<ol class=list-inline><li>'A1CF'</li><li>'AAAS'</li><li>'ABAT'</li><li>'ABCC2'</li><li>'ABCC8'</li><li>'ABHD2'</li></ol>\n</dd>\n\t<dt>$`GO:0000012`</dt>\n\t\t<dd><style>\n.list-inline {list-style: none; margin:0; padding: 0}\n.list-inline>li {display: inline-block}\n.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n</style>\n<ol class=list-inline><li>'APLF'</li><li>'APTX'</li><li>'ERCC6'</li><li>'ERCC8'</li><li>'LIG4'</li><li>'PARP1'</li></ol>\n</dd>\n\t<dt>$`GO:0000018`</dt>\n\t\t<dd><style>\n.list-inline {list-style: none; margin:0; padding: 0}\n.list-inline>li {display: inline-block}\n.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n</style>\n<ol class=list-inline><li>'ACTR2'</li><li>'ALYREF'</li><li>'ANKLE1'</li><li>'ARID2'</li><li>'BCL6'</li><li>'BLM'</li></ol>\n</dd>\n\t<dt>$`GO:0000027`</dt>\n\t\t<dd><style>\n.list-inline {list-style: none; margin:0; padding: 0}\n.list-inline>li {display: inline-block}\n.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n</style>\n<ol class=list-inline><li>'BOP1'</li><li>'BRIX1'</li><li>'DDX28'</li><li>'DHX30'</li><li>'FASTKD2'</li><li>'MDN1'</li></ol>\n</dd>\n</dl>\n",
      "text/markdown": "$`GO:0000002`\n:   1. 'AKT3'\n2. 'DNA2'\n3. 'LIG3'\n4. 'LONP1'\n5. 'MEF2A'\n6. 'MGME1'\n\n\n\n$`GO:0000003`\n:   1. 'A1CF'\n2. 'AAAS'\n3. 'ABAT'\n4. 'ABCC2'\n5. 'ABCC8'\n6. 'ABHD2'\n\n\n\n$`GO:0000012`\n:   1. 'APLF'\n2. 'APTX'\n3. 'ERCC6'\n4. 'ERCC8'\n5. 'LIG4'\n6. 'PARP1'\n\n\n\n$`GO:0000018`\n:   1. 'ACTR2'\n2. 'ALYREF'\n3. 'ANKLE1'\n4. 'ARID2'\n5. 'BCL6'\n6. 'BLM'\n\n\n\n$`GO:0000027`\n:   1. 'BOP1'\n2. 'BRIX1'\n3. 'DDX28'\n4. 'DHX30'\n5. 'FASTKD2'\n6. 'MDN1'\n\n\n\n\n\n",
      "text/latex": "\\begin{description}\n\\item[\\$`GO:0000002`] \\begin{enumerate*}\n\\item 'AKT3'\n\\item 'DNA2'\n\\item 'LIG3'\n\\item 'LONP1'\n\\item 'MEF2A'\n\\item 'MGME1'\n\\end{enumerate*}\n\n\\item[\\$`GO:0000003`] \\begin{enumerate*}\n\\item 'A1CF'\n\\item 'AAAS'\n\\item 'ABAT'\n\\item 'ABCC2'\n\\item 'ABCC8'\n\\item 'ABHD2'\n\\end{enumerate*}\n\n\\item[\\$`GO:0000012`] \\begin{enumerate*}\n\\item 'APLF'\n\\item 'APTX'\n\\item 'ERCC6'\n\\item 'ERCC8'\n\\item 'LIG4'\n\\item 'PARP1'\n\\end{enumerate*}\n\n\\item[\\$`GO:0000018`] \\begin{enumerate*}\n\\item 'ACTR2'\n\\item 'ALYREF'\n\\item 'ANKLE1'\n\\item 'ARID2'\n\\item 'BCL6'\n\\item 'BLM'\n\\end{enumerate*}\n\n\\item[\\$`GO:0000027`] \\begin{enumerate*}\n\\item 'BOP1'\n\\item 'BRIX1'\n\\item 'DDX28'\n\\item 'DHX30'\n\\item 'FASTKD2'\n\\item 'MDN1'\n\\end{enumerate*}\n\n\\end{description}\n",
      "text/plain": "$`GO:0000002`\n[1] \"AKT3\"  \"DNA2\"  \"LIG3\"  \"LONP1\" \"MEF2A\" \"MGME1\"\n\n$`GO:0000003`\n[1] \"A1CF\"  \"AAAS\"  \"ABAT\"  \"ABCC2\" \"ABCC8\" \"ABHD2\"\n\n$`GO:0000012`\n[1] \"APLF\"  \"APTX\"  \"ERCC6\" \"ERCC8\" \"LIG4\"  \"PARP1\"\n\n$`GO:0000018`\n[1] \"ACTR2\"  \"ALYREF\" \"ANKLE1\" \"ARID2\"  \"BCL6\"   \"BLM\"   \n\n$`GO:0000027`\n[1] \"BOP1\"    \"BRIX1\"   \"DDX28\"   \"DHX30\"   \"FASTKD2\" \"MDN1\"   \n"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load the pathways into a named list\n",
    "GO_term_hallmark <- gmtPathways(\"./data/GO_terms.gmt\")\n",
    "\n",
    "# Show the first few GO terms, and within those, show only the first few genes.\n",
    "tmp = lapply(GO_term_hallmark,head)\n",
    "tmp[1:5]"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in fgsea(pathways = GO_term_hallmark, stats = stats, minSize = 15, : object 'stats' not found\n",
     "output_type": "error",
     "traceback": [
      "Error in fgsea(pathways = GO_term_hallmark, stats = stats, minSize = 15, : object 'stats' not found\nTraceback:\n",
      "1. fgsea(pathways = GO_term_hallmark, stats = stats, minSize = 15, \n .     maxSize = 500, nPermSimple = 10000)"
     ]
    }
   ],
   "source": [
    "# Running fgsea\n",
    "fgseaRes <- fgsea(pathways = GO_term_hallmark,\n",
    "                  stats    = stats,\n",
    "                  minSize  = 15,\n",
    "                  maxSize  = 500,\n",
    "                  nPermSimple = 10000\n",
    "                  )"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "outputs": [
    {
     "data": {
      "text/html": "<table class=\"dataframe\">\n<caption>A data.table: 6 × 7</caption>\n<thead>\n\t<tr><th scope=col>pathway</th><th scope=col>pval</th><th scope=col>padj</th><th scope=col>log2err</th><th scope=col>ES</th><th scope=col>NES</th><th scope=col>size</th></tr>\n\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n</thead>\n<tbody>\n\t<tr><td>GO:0065007</td><td>2.546484e-11</td><td>1.589006e-08</td><td>0.8634154</td><td>0.3431229</td><td>1.538333</td><td>459</td></tr>\n\t<tr><td>GO:0050789</td><td>7.479736e-10</td><td>2.333678e-07</td><td>0.8012156</td><td>0.3388931</td><td>1.514710</td><td>432</td></tr>\n\t<tr><td>GO:0050794</td><td>1.598783e-09</td><td>3.325469e-07</td><td>0.7881868</td><td>0.3366914</td><td>1.502073</td><td>414</td></tr>\n\t<tr><td>GO:0032501</td><td>4.659098e-09</td><td>7.177948e-07</td><td>0.7614608</td><td>0.3641697</td><td>1.592389</td><td>269</td></tr>\n\t<tr><td>GO:0050896</td><td>5.751561e-09</td><td>7.177948e-07</td><td>0.7614608</td><td>0.3535281</td><td>1.556321</td><td>313</td></tr>\n\t<tr><td>GO:0032502</td><td>1.143077e-08</td><td>1.188800e-06</td><td>0.7477397</td><td>0.3665868</td><td>1.599132</td><td>253</td></tr>\n</tbody>\n</table>\n",
      "text/markdown": "\nA data.table: 6 × 7\n\n| pathway &lt;chr&gt; | pval &lt;dbl&gt; | padj &lt;dbl&gt; | log2err &lt;dbl&gt; | ES &lt;dbl&gt; | NES &lt;dbl&gt; | size &lt;int&gt; |\n|---|---|---|---|---|---|---|\n| GO:0065007 | 2.546484e-11 | 1.589006e-08 | 0.8634154 | 0.3431229 | 1.538333 | 459 |\n| GO:0050789 | 7.479736e-10 | 2.333678e-07 | 0.8012156 | 0.3388931 | 1.514710 | 432 |\n| GO:0050794 | 1.598783e-09 | 3.325469e-07 | 0.7881868 | 0.3366914 | 1.502073 | 414 |\n| GO:0032501 | 4.659098e-09 | 7.177948e-07 | 0.7614608 | 0.3641697 | 1.592389 | 269 |\n| GO:0050896 | 5.751561e-09 | 7.177948e-07 | 0.7614608 | 0.3535281 | 1.556321 | 313 |\n| GO:0032502 | 1.143077e-08 | 1.188800e-06 | 0.7477397 | 0.3665868 | 1.599132 | 253 |\n\n",
      "text/latex": "A data.table: 6 × 7\n\\begin{tabular}{lllllll}\n pathway & pval & padj & log2err & ES & NES & size\\\\\n <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <int>\\\\\n\\hline\n\t GO:0065007 & 2.546484e-11 & 1.589006e-08 & 0.8634154 & 0.3431229 & 1.538333 & 459\\\\\n\t GO:0050789 & 7.479736e-10 & 2.333678e-07 & 0.8012156 & 0.3388931 & 1.514710 & 432\\\\\n\t GO:0050794 & 1.598783e-09 & 3.325469e-07 & 0.7881868 & 0.3366914 & 1.502073 & 414\\\\\n\t GO:0032501 & 4.659098e-09 & 7.177948e-07 & 0.7614608 & 0.3641697 & 1.592389 & 269\\\\\n\t GO:0050896 & 5.751561e-09 & 7.177948e-07 & 0.7614608 & 0.3535281 & 1.556321 & 313\\\\\n\t GO:0032502 & 1.143077e-08 & 1.188800e-06 & 0.7477397 & 0.3665868 & 1.599132 & 253\\\\\n\\end{tabular}\n",
      "text/plain": "  pathway    pval         padj         log2err   ES        NES      size\n1 GO:0065007 2.546484e-11 1.589006e-08 0.8634154 0.3431229 1.538333 459 \n2 GO:0050789 7.479736e-10 2.333678e-07 0.8012156 0.3388931 1.514710 432 \n3 GO:0050794 1.598783e-09 3.325469e-07 0.7881868 0.3366914 1.502073 414 \n4 GO:0032501 4.659098e-09 7.177948e-07 0.7614608 0.3641697 1.592389 269 \n5 GO:0050896 5.751561e-09 7.177948e-07 0.7614608 0.3535281 1.556321 313 \n6 GO:0032502 1.143077e-08 1.188800e-06 0.7477397 0.3665868 1.599132 253 "
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(fgseaRes[order(pval), ][,-8])"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in eval(expr, envir, enclos): object 'fgseaRes' not found\n",
     "output_type": "error",
     "traceback": [
      "Error in eval(expr, envir, enclos): object 'fgseaRes' not found\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "topPathwaysUp <- fgseaRes[ES > 0][head(order(pval), n=5), pathway]\n",
    "topPathwaysDown <- fgseaRes[ES < 0][head(order(pval), n=5), pathway]\n",
    "topPathways <- c(topPathwaysUp, rev(topPathwaysDown))\n",
    "plotGseaTable(GO_term_hallmark[topPathways], stats, fgseaRes,\n",
    "              gseaParam=0.5)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "### FGSEA Using KEGG"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "outputs": [
    {
     "data": {
      "text/html": "<dl>\n\t<dt>$hsa00010</dt>\n\t\t<dd><style>\n.list-inline {list-style: none; margin:0; padding: 0}\n.list-inline>li {display: inline-block}\n.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n</style>\n<ol class=list-inline><li>'HK3'</li><li>'HK1'</li><li>'HK2'</li><li>'HKDC1'</li><li>'GCK'</li><li>'GPI'</li></ol>\n</dd>\n\t<dt>$hsa00020</dt>\n\t\t<dd><style>\n.list-inline {list-style: none; margin:0; padding: 0}\n.list-inline>li {display: inline-block}\n.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n</style>\n<ol class=list-inline><li>'CS'</li><li>'ACLY'</li><li>'ACO2'</li><li>'ACO1'</li><li>'IDH1'</li><li>'IDH2'</li></ol>\n</dd>\n\t<dt>$hsa00030</dt>\n\t\t<dd><style>\n.list-inline {list-style: none; margin:0; padding: 0}\n.list-inline>li {display: inline-block}\n.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n</style>\n<ol class=list-inline><li>'GPI'</li><li>'G6PD'</li><li>'PGLS'</li><li>'H6PD'</li><li>'PGD'</li><li>'RPE'</li></ol>\n</dd>\n\t<dt>$hsa00040</dt>\n\t\t<dd><style>\n.list-inline {list-style: none; margin:0; padding: 0}\n.list-inline>li {display: inline-block}\n.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n</style>\n<ol class=list-inline><li>'GUSB'</li><li>'KL'</li><li>'UGT2A1'</li><li>'UGT2A3'</li><li>'UGT2B17'</li><li>'UGT2B11'</li></ol>\n</dd>\n\t<dt>$hsa00051</dt>\n\t\t<dd><style>\n.list-inline {list-style: none; margin:0; padding: 0}\n.list-inline>li {display: inline-block}\n.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n</style>\n<ol class=list-inline><li>'MPI'</li><li>'PMM2'</li><li>'PMM1'</li><li>'GMPPB'</li><li>'GMPPA'</li><li>'GMDS'</li></ol>\n</dd>\n</dl>\n",
      "text/markdown": "$hsa00010\n:   1. 'HK3'\n2. 'HK1'\n3. 'HK2'\n4. 'HKDC1'\n5. 'GCK'\n6. 'GPI'\n\n\n\n$hsa00020\n:   1. 'CS'\n2. 'ACLY'\n3. 'ACO2'\n4. 'ACO1'\n5. 'IDH1'\n6. 'IDH2'\n\n\n\n$hsa00030\n:   1. 'GPI'\n2. 'G6PD'\n3. 'PGLS'\n4. 'H6PD'\n5. 'PGD'\n6. 'RPE'\n\n\n\n$hsa00040\n:   1. 'GUSB'\n2. 'KL'\n3. 'UGT2A1'\n4. 'UGT2A3'\n5. 'UGT2B17'\n6. 'UGT2B11'\n\n\n\n$hsa00051\n:   1. 'MPI'\n2. 'PMM2'\n3. 'PMM1'\n4. 'GMPPB'\n5. 'GMPPA'\n6. 'GMDS'\n\n\n\n\n\n",
      "text/latex": "\\begin{description}\n\\item[\\$hsa00010] \\begin{enumerate*}\n\\item 'HK3'\n\\item 'HK1'\n\\item 'HK2'\n\\item 'HKDC1'\n\\item 'GCK'\n\\item 'GPI'\n\\end{enumerate*}\n\n\\item[\\$hsa00020] \\begin{enumerate*}\n\\item 'CS'\n\\item 'ACLY'\n\\item 'ACO2'\n\\item 'ACO1'\n\\item 'IDH1'\n\\item 'IDH2'\n\\end{enumerate*}\n\n\\item[\\$hsa00030] \\begin{enumerate*}\n\\item 'GPI'\n\\item 'G6PD'\n\\item 'PGLS'\n\\item 'H6PD'\n\\item 'PGD'\n\\item 'RPE'\n\\end{enumerate*}\n\n\\item[\\$hsa00040] \\begin{enumerate*}\n\\item 'GUSB'\n\\item 'KL'\n\\item 'UGT2A1'\n\\item 'UGT2A3'\n\\item 'UGT2B17'\n\\item 'UGT2B11'\n\\end{enumerate*}\n\n\\item[\\$hsa00051] \\begin{enumerate*}\n\\item 'MPI'\n\\item 'PMM2'\n\\item 'PMM1'\n\\item 'GMPPB'\n\\item 'GMPPA'\n\\item 'GMDS'\n\\end{enumerate*}\n\n\\end{description}\n",
      "text/plain": "$hsa00010\n[1] \"HK3\"   \"HK1\"   \"HK2\"   \"HKDC1\" \"GCK\"   \"GPI\"  \n\n$hsa00020\n[1] \"CS\"   \"ACLY\" \"ACO2\" \"ACO1\" \"IDH1\" \"IDH2\"\n\n$hsa00030\n[1] \"GPI\"  \"G6PD\" \"PGLS\" \"H6PD\" \"PGD\"  \"RPE\" \n\n$hsa00040\n[1] \"GUSB\"    \"KL\"      \"UGT2A1\"  \"UGT2A3\"  \"UGT2B17\" \"UGT2B11\"\n\n$hsa00051\n[1] \"MPI\"   \"PMM2\"  \"PMM1\"  \"GMPPB\" \"GMPPA\" \"GMDS\" \n"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load the pathways into a named list\n",
    "KEGG_hallmark <- gmtPathways(\"./data/KEGG_pathways.gmt\")\n",
    "\n",
    "# Show the first few GO terms, and within those, show only the first few genes.\n",
    "tmp = lapply(KEGG_hallmark,head)\n",
    "tmp[1:5]"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "\"There are ties in the preranked stats (62.67% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.\"\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "\"There are duplicate gene names, fgsea may produce unexpected results.\"\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "\"All values in the stats vector are greater than zero and scoreType is \"std\", maybe you should switch to scoreType = \"pos\".\"\n"
     ]
    }
   ],
   "source": [
    "# Running fgsea\n",
    "fgseaRes <- fgsea(pathways = KEGG_hallmark,\n",
    "                  stats    = stats,\n",
    "                  minSize  = 15,\n",
    "                  maxSize  = 500)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "outputs": [
    {
     "data": {
      "text/html": "<table class=\"dataframe\">\n<caption>A data.table: 6 × 7</caption>\n<thead>\n\t<tr><th scope=col>pathway</th><th scope=col>pval</th><th scope=col>padj</th><th scope=col>log2err</th><th scope=col>ES</th><th scope=col>NES</th><th scope=col>size</th></tr>\n\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n</thead>\n<tbody>\n\t<tr><td>hsa05168</td><td>9.927117e-14</td><td>3.156823e-11</td><td>0.9545416</td><td>0.3559273</td><td>1.591875</td><td>468</td></tr>\n\t<tr><td>hsa04060</td><td>2.105054e-13</td><td>3.347035e-11</td><td>0.9436322</td><td>0.3884407</td><td>1.707546</td><td>291</td></tr>\n\t<tr><td>hsa04151</td><td>1.479326e-10</td><td>1.568086e-08</td><td>0.8266573</td><td>0.3566070</td><td>1.578857</td><td>353</td></tr>\n\t<tr><td>hsa04530</td><td>9.306426e-10</td><td>7.398608e-08</td><td>0.8012156</td><td>0.4137364</td><td>1.757101</td><td>164</td></tr>\n\t<tr><td>hsa04510</td><td>5.558773e-09</td><td>3.535380e-07</td><td>0.7614608</td><td>0.3850561</td><td>1.658115</td><td>199</td></tr>\n\t<tr><td>hsa05202</td><td>2.349362e-08</td><td>1.245162e-06</td><td>0.7337620</td><td>0.3821628</td><td>1.639908</td><td>188</td></tr>\n</tbody>\n</table>\n",
      "text/markdown": "\nA data.table: 6 × 7\n\n| pathway &lt;chr&gt; | pval &lt;dbl&gt; | padj &lt;dbl&gt; | log2err &lt;dbl&gt; | ES &lt;dbl&gt; | NES &lt;dbl&gt; | size &lt;int&gt; |\n|---|---|---|---|---|---|---|\n| hsa05168 | 9.927117e-14 | 3.156823e-11 | 0.9545416 | 0.3559273 | 1.591875 | 468 |\n| hsa04060 | 2.105054e-13 | 3.347035e-11 | 0.9436322 | 0.3884407 | 1.707546 | 291 |\n| hsa04151 | 1.479326e-10 | 1.568086e-08 | 0.8266573 | 0.3566070 | 1.578857 | 353 |\n| hsa04530 | 9.306426e-10 | 7.398608e-08 | 0.8012156 | 0.4137364 | 1.757101 | 164 |\n| hsa04510 | 5.558773e-09 | 3.535380e-07 | 0.7614608 | 0.3850561 | 1.658115 | 199 |\n| hsa05202 | 2.349362e-08 | 1.245162e-06 | 0.7337620 | 0.3821628 | 1.639908 | 188 |\n\n",
      "text/latex": "A data.table: 6 × 7\n\\begin{tabular}{lllllll}\n pathway & pval & padj & log2err & ES & NES & size\\\\\n <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <int>\\\\\n\\hline\n\t hsa05168 & 9.927117e-14 & 3.156823e-11 & 0.9545416 & 0.3559273 & 1.591875 & 468\\\\\n\t hsa04060 & 2.105054e-13 & 3.347035e-11 & 0.9436322 & 0.3884407 & 1.707546 & 291\\\\\n\t hsa04151 & 1.479326e-10 & 1.568086e-08 & 0.8266573 & 0.3566070 & 1.578857 & 353\\\\\n\t hsa04530 & 9.306426e-10 & 7.398608e-08 & 0.8012156 & 0.4137364 & 1.757101 & 164\\\\\n\t hsa04510 & 5.558773e-09 & 3.535380e-07 & 0.7614608 & 0.3850561 & 1.658115 & 199\\\\\n\t hsa05202 & 2.349362e-08 & 1.245162e-06 & 0.7337620 & 0.3821628 & 1.639908 & 188\\\\\n\\end{tabular}\n",
      "text/plain": "  pathway  pval         padj         log2err   ES        NES      size\n1 hsa05168 9.927117e-14 3.156823e-11 0.9545416 0.3559273 1.591875 468 \n2 hsa04060 2.105054e-13 3.347035e-11 0.9436322 0.3884407 1.707546 291 \n3 hsa04151 1.479326e-10 1.568086e-08 0.8266573 0.3566070 1.578857 353 \n4 hsa04530 9.306426e-10 7.398608e-08 0.8012156 0.4137364 1.757101 164 \n5 hsa04510 5.558773e-09 3.535380e-07 0.7614608 0.3850561 1.658115 199 \n6 hsa05202 2.349362e-08 1.245162e-06 0.7337620 0.3821628 1.639908 188 "
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(fgseaRes[order(pval), ][,-8])"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "outputs": [
    {
     "data": {
      "text/plain": "plot without title",
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAM1BMVEUAAAAzMzNNTU1oaGh8\nfHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD////p0TvgAAAACXBIWXMAABJ0\nAAASdAHeZh94AAAgAElEQVR4nO2diYKrKpRFeTjFWA7//7WPSTkMYuoWVZiw1+u+MQhH3LAB\nid3FNgDAj2GlKwDAJwAjAZABGAmADMBIAGQARgIgAzASABmAkQDIAIwEQAZgJAAyACMBkAEY\nCYAMwEgAZABGAiADMBIAGYCRAMgAjARABmAkADIAIwGQARgJgAzASABkAEYCIAMwEgAZgJEA\nyACMBEAGYCQAMgAjAZABGAmADMBIpzAF7xcvfdInC9SIso6dqF03Zgr3m/fD2EQuwg5k0jo0\njLW57uLX7kOGTYcu3R9uzN7e3HVSw/TJInU6mHi8dv/K7xqJk4s4RlrNbfA127UyBQrDwkj/\niBkyWzZEkgsbaWKs/xKfX93eSX/I7xqJPexFnCv1rBUDweJr/JNrZQr03bAw0ilGvNXrqrcw\nEt9XS6IrZlkX/a6RGrYcF3GuxJiaitZsl4eR7scunvqcxBMJHzazMlGJA+OP7Yv1Ms+kO3bP\nvo6cK2tU8f0zJ0/W7YfLIGembWwYH3Vtl05WzEncb2RtZMHjXkhmfbODnDqmVjy1TFs+GJt1\nfWNG+n4wKTzRtxFepHf077U0gTeqkNCEi+kSS7sfQGekh17KD8RI8llfzAZc5er12kRktTk7\nprr406xrcmJC0wRJq2qtHjseTuJ+Q52smK0hyaxudpCZR3063/O/jN2rCodGGliwl3MZ7GHu\nqlXT3CIOnTv691rugWk8cRWmdIOR/hktnF6/M/aUlrAdQSi+ij7XCNGf2/48LT1jc056supZ\nnu2AsGqWSdZGPM1NpGI0cS/Vrpt3L3tm+VX5SIwMszydcRoVsfX04W42qHOinzbDV7K4H4zP\n28zFPegR6iHuL2idf6vlHpjGe5pUGOnfOXbtVpuyWSOZIVaOiGKB18nu1x6eUZka9QDwCyu7\nvU2PHtmpK61yAXVUjCbupb68EDazWti0+kTOZd1+qVFOcREjbVMvJX79krp6k7pVqWzD7Jnt\nZ0Y6ApN4eu6fYKQf4P6OtEyP1jHS/m8r+usgngIe2lMk5yiHzK9fWNkFRiKd01bM67GkJ0Tu\nRcxNxmeDWMnMc/baylHlpKt/Pbi/VL0Ipj/kXL/oBZh3R/9cy2Md4sWDkX6AI1x79EnfSJNw\nCm+2Zl/l2Zzm6Sr/ys55Rvq+kWL3IocMM3c+5HNTpp+n7HXlrsxpV59fn7ZJz5Zj1KCU8O/o\nn2tpPoJ4MNIPoML1rBmnJWokscL4EsPiILfEVienSJukwfLz1I9fRzVJVak34jcUvRcxHc3q\n0UAyDU3mZ6RNmn8Ouvpx/Hr/p/1djF9c1jO4o3+vpZnq/Hgw0g8IWvzESAPrhWEm8W/v5hQD\nbTv/xsqO/o606schsqOw/9v5Dzu0D4ZGkpMn9/NmwayWWBN09c5sDvq/1SWD6YcWqfXA1Oo5\nvKN/q+URmMTTIn7BSD/ANdLXNu+r5mVzjCRUllMR00+rJKd8MOC/sbJTT7+derNhUPuFam9J\nPJN1tGIk0bkh5172dPMc85D/PPPv2kke4eJLSDcK6b7a13fb9801qfWilQ/v6N9qeQQm8Sbs\n2v0YKtxgHje+ZEeTXZfOS3od1OpRleRU/f03VnaCr/1dO73I04t6+WBDKmYTnRsiNfSMNMt+\n+ST1z8MuJDddnTy77XVpU+W9YPqXHfWl0QWDO/q3Wh6BaRuqH+N6GOkHOMIJKdsvtTf61fhG\neqido4d5X8zmVOuuX1nZSZ6d6Jjtw9hkFP7uvanSJro3ZGvoP1A91IsP8s2GjD46rjuFRtrm\nXt7F81vBOvEIo788zerVv6N/rOURmLbhA282lGdiv7Oyq5icD2/fCwwjleMba3/wGsWMtMBI\npfjW2h+8RikjDYz85BArnrUygMLthhnIRSkjder/cipRPGtlAKgUGAmADMBIAGQARgIgAzAS\nABmAkQDIAIwEQAZgJAAyACMBkAEYCYAMwEgAZABGAiADMBIAGYCRAMgAjARABmAkADIAIwGQ\nARgJgAzASABkAEYCIAMwEgAZgJEAyACMBEAGYKSCsFcoXUnwEmingrxkpDN/oeVuBZqjID8z\nUiQNFAPiFyS3kXTyBlMVAIIX5FeNRPKVvs8agMgF+Ssj2S/gt4C2Bfl7I+1ZQG6gaUGKGenI\nDXIBLQtS3EjwUzYgYkHuYiRsSfwcyFeQexlpPwT/AnQryC2NZL6X1ubdgGAFubOR9Ad4FWhV\nkB/46O+MhOnpJaBRQX7go781kv0GToA6Bfk3C8W7vk3+PSOZ8iACZCnIvxjorOvb5N82EpZ6\nMaBJQf7BP6dd3yb/jZH200ADMQrybfckur5N/lMj6dSyMt4CaFCQb7sn0fVtcgEjycPKJ6i6\n774w3zVPquvb5HJGqtlLFd96eb5rnlTXt8lFjVRtf6r2xu/Ad82T6vo2ubCRKp2W6rzrm/Bd\n86S6vk2GkUpQ513fhO+aJ9X1bTKMVII67/omfNc8qa5vk2GkEtR51zfhu+ZJdX2bDCOVoM67\nvgnfNU+q69tkGKkEdd71OZfdYOCMD6vOenQ2wbgfzD1j/SKPVpv37GL/zpsZaWSRG98FbacL\nxd0MFyXopb5SzbnKlpqdok2ywUjk0Q8MI7lcGalVPaARRzPtDfKbPphUIhetsXB9uCQu9u+8\nl5FmJ5VKpwV9XCn+SCScX2rlqebU7UOcNKgEfuYkEnkO7hJGcrkw0hfj8zZz9iW17Gy6SDIF\nuciwdmzYtl7+IxqnT1zs33krI1l59Fcr3cjaVU4Nc1joJMNFCedSXao5VcsMpBVn1q8y/kmD\nkcje/UhgJJcLIw1MrimecjwcyaAoGtcUfCr3rIwfoRIRv2ueVNe3yfczkpVn/35I18oxaVuU\nalGCDOkSzqWeLNWcnK2b0zxdssFIZP9+VKnzC1WJEGhgXLXzJNTSa/FJjGxcNVzH5EJNjagj\nG22pYZefjJRmWSE9VTlWHg2Rbh9t2tOyfgYnQT7UjE52cqkl0t/D+EHz6DKpyP79qLSrC1UG\nY50ca0fZ2swcPfTRsDmzTMem3thrm48zDdseXC0RRDG9tEus/yth9oZ5It3ltB1koAmqrRwT\n0ku1bLk00sBGL2VV8ZKR/ftR9bm4UG0wuQIf5W4Cl3PLUx4x9lTLhM0zkqP1cUalq2FulE+z\n3G+oOvGMdEjXqCn+67zDBxlIwqQel1rmbuHtWR+i2S6M9GThCnGU4dKR/ftRCckL1QdTK3C9\n5o+oSIyk7LUeI9pxRm429Goa0jMZJiSJ0/GIdA/WrducWIIFGUhCp55yVrrrYy+l1t8XRho7\n7rfPwmW0ZOTgflRC8kL1cTxgy73QbjYPPMv0aH0jaVa1FU7PyCKLTB3laCc8hSlpi/ZoLZ3a\ng07trgUZbILdG4lcqpHb2NfPSF77rLzVMRKRo/cDI7lYI8lnHfMrUHuIyn0j2QdQ5n80ZlRr\n/qrudybWo1WaGGn4I9Xhgww2IbbJuF+qVwuKayOt7m5D25gY55Hj9wMjuRAjiZXy0Egb9KwZ\np8U83UpjLWTK90QmW6jX298VcWokxXw12gQZZv3wakP53Z1s5b9es6VpFy8RRvo3HCNtxBHa\nSA81zk1y0aZ/hjg8ZYroDIt8kNaz14rtb4nT8Yh0+nD0HkaCvCQDSehY7E2hV42018JadDq2\njlKRw/tRCWeXqRRrpEbv1TV6A8I87pI3Gwb1CDTsihtlRbuo392fKsNqsgGn4xHp1OsFX0rr\nOEEGkvCUrRG4kF4qNR+pQGtnn5EWu919GRlGusAa6anHs6/9FSy9n9ccG7crP35dsgX3rTrV\nIm3wW0S9OKsjIp05PJ2QnAy0sErQCntvM75qJPOuXXvk68kcdhUZRrqALO3Umw3SPFLh9mtS\nbaff6FZZ5WHj/0SvipkMm81bPe5jBpFuEeJ2qbe/SQZdmJYYm/1V+/BS/nGIrYXZkiWLwYvI\nMBIAvwGMBEAGYCQAMgAjAZABGAmADMBIAGQARgIgAzASABmAkQDIAIwEQAZgJAAyACMBkAEY\nCYAMwEgAZABGAiADMBIAGYCRAMgAjARABmAkADIAIwGQARgJgAzASABkAEYCIAMwEgAZgJEA\nyACMBEAGYCQAMgAjlYP85RHv39f/ug+4CWiocrxkJEa+uhlKVx9Q0Bzl+KGRfPC3AUsC7cuR\n30jk76qWvrnagODl+F0j7X/yBxPVnwCRy/FnRtLf0di/CbQtRwEjqf8wRf0C0LQcBY20/wty\nAS3LcQsjwU55gIrluI2R9mPw70C/ctzMSDpnaVHeFQhXjpsa6SgMvgEEK8ftjYQF3+tAqXK8\ng5HgpReBTOV4JyMxLPbSQJ5yvKORYKgTIEs53tRIKkspzW4LFCnHWxsJT08uEKMcH2EkhtWe\nAhqU45OMVP38VPntFwVG+iAqv/2ifKaRKu1Rld72LfhQI9XZpeq863sAI30Qdd71PfhYI9XY\nqWq857vwuUaqsFdVeMu3AUb6ICq85dvwwUaqr1vVd8dpLrvAwBkf1v3bFwtS556xfonkjVyr\nHiONNMVWOTwXMjaBiskS+8lVNsScyBhkuCpBLhvUAEZyuTJSqzpAY76tnPmpkzrka5A3dq1q\njDSzmJF45FzAYAWNRzu7FFcFE74IMlyUIJcNawAjuVwY6YvxeZs5+9JfO52dpnJxuHZsCPLG\nrlWLkYQKoa6TViZ6jhRl/SongP4iWnBykEUG1p3mDDJclCCXjdQARnK5MNLAJvHvkz3Ut6fp\nMST1KS20rXKwdfNGr1WJkUbWhrquvDs9R+j266ajhSc5W7dkewYZ0iXIZWM1gJFchEAD46rv\nT0KtdlJHYubh0iCiWeXTz6xHrWXXk6T2dmXg5K0aMbiE3bNT3TZ+LhLC5qEl5PPTmLqUWT4m\nAvsZTEIqcqzOMJILY50cdEc57DBz9NBHw3a0p/5o2WJGbpvasO3B1WrEzVs1c0SFWekZPxey\nsjYWTbUVORWEG5hjhpAgg0lIRo7VGa3sIsRbhYcaOdHPcmHWyLSnWsZtrjkeIjkwkjEi32Ak\nh0AFMyFFz4WMapnsR5tkY62te46EezLj1jOCDHvCVWQY6QqmnoC1JSIqEnOoNVvESHKzoZcP\nRjASIdh+ILsH1wot3F0fmxLajCuLnhSMHT9/Qo1m2BOuIsNIV1hnDGJymc0DzzI9Wt9IjdyR\njRhJFln0REYi1o6vwkCGqUuFVt66CWTnQ++knF6qv1rbBRlUwmVkGOkC4oyH/FmByw2D9hCV\nH+boVU/Q2Tk10nHIYSSLrwJn5+cCWv+3uMBItNc74dar3YYgg0q4jAwjXeBMMdPQyKmlZ804\n6X0FvRO3iCmfbDuTVLpbS1KB1/GcvcwLIy1Nu8Sj2YKnRrp2aWgJ9kJkGOkCx0gbUXUxGwxy\nHprU/qc1kk01h4vc7yGpIPhlaTw95zGxNkg7npGm4NRxUv8qtJy/WRJkIAnJyNE6w0gu1kiN\n3qtr9AbErJ+R/LcVdHaSKpphlZsNzxfebKgJr+N19EWcpJGWiI/2Ek+psDBldEtAvaewdufP\nSEEGkpCMHK0zjORijfTU883X/raX3s9r3N8XjJ4k9WEPvbxV462OGraG56L05LH/yLgf6IdX\n7q789pP8Sn6SYX/WPRKSkWN1hpFcyNJOvdmgppNeHkxqeFrVG91edidVFDOHXt6q8Yzk9MOk\nkVjKSPL9g/1V+zCckL9J7tnZDPssZkskI8NIAPwOMBIAGYCRAMgAjARABmAkADIAIwGQARgJ\ngAzASABkAEYCIAMwEgAZgJEAyACMBEAGYCQAMgAjAZABGAmADMBIAGQARgIgAzASABmAkQDI\nAIwEQAZgJAAyACMBkAEYCYAMwEgAZABGAiADMBIAGYCRAMgAjFQO5uD8oRiaDt4BtFM5XjOS\nk2Hb/wcOuxlojnL80Ej2bzRg8ioPxC9HdiO5+cBfAr3L8btGIrFK32gNQORy/JmR4KjfB9qW\n4++NhH2KXwOalqOskWCorEDLcpQ30p4b/BiIWI7bGOlIA/8M9CvH3YwEO/0AKFeOexqJ4Veo\nfwGKlePWRrLxwStAqXK8g5GwGfEiEKkcb2Qkpv4D50CdcrybkfTXwqLdFehSjvc0EswUBZqU\nA0b6IKBJOT7ASOg+O1CiHJ9jJIZ3y+u++7J8mpFYzd2p3jsvzycaqdr+VO2N34APNVKdXarO\nu74HMNIHUedd3wMY6YOo867vAYz0QdR51/cARvog6rzrewAjfRB13vU57EqQgTM+rPu3ryP7\nePSpo6uS1JNr1WMkVwgi0uoKGuW0cDr3tyNvc89Yv5zkteFidYaRXK6M1KpGbMy3le/ZZ1Nw\ndpp5Tob7YCP5t+0KQURauDriZ703WTid222qFyJvk65L3Hs2XLTOMJLLhZG+GJ+3mbMv/bU7\nHMMPI3U295F6di2///udP2aQdzSSJwQRqWeD+Hdg/blM54WTub2meiHyxkWJtVM1CiDhonWG\nkVwujDSwSfz7ZA/17bkPjCNrj6PHkdmmnl2rEiP5QhCRTHpypXZaOJnbbapXIj+VQVbGY5lJ\nuGidYSQXoc7AuBJ/EjK3kzoSMw9X41TH5HxuBsXlaAjRAoeRRhtruPJlJfhCEJHM2jjeea8K\niy8N42M8t9NUL0Xu2XwemYSL1hkN7cJYJwfeUTYYM0cPfTRs3mDUsmVf0G22+abemI6kVo4v\nBBHpYZZJ5/NGovCm2oq10dzXc50fuWHbg7N+jUYm4aJ1RkO7CPFW4aFGjjuznMobmfZUy7jN\nbZ2HSLYNcRjJaQEYyeB5wYo0yid3d1Z5ufAkG2tt1aIryH1tJP+sGUR5NDINF6szGtqFqYdT\nvdyPtA+RU83ygZGU6dZhX37ASAavx1qR9HSfeuw5L9wxOXus3urtB0aSmw29rEwY2R1Cwzqj\noV32B3E5dbNuNovmZXq0vpEauU0aGEmz7ruuMJIhIoQSaZTLJNF5k1PSWWG65RLk/hcjyeZe\n4pFJuGid0dAu1khyvWx+LGgPUfkhZ68mrBMjHd9gJENMCDUcmaE/+XPPWWFno5J5mvN/MFIi\nMgkXrTMa2oUYSayUh0aq1bNmnPS+gt67WcSUz6zWpKAbBkY6OPfC6emXC29bzEi2qV6N3FEj\n+ZFJOGx/v4BjpI2oupjVsZyHJrVxGjUSV6PV0XwwksERgoikB/qT326uCnfeY6yT2zbVq5F1\niUXuY4SRSbhondHQLtZIjd6ra/QGxKyfkfyfy4Ol3aDWz8PeDjCSwRGCiCQOV5Pw/cJP2Rri\nkSW62fDCmw1eZPF0tMpnn2csMgkXrTMa2sUa6annGyXccSjs5fy+EBhp5cdvTu75ynEmbipS\nG/wU9O3C3ot6e+7mu5HNblwbj9z4J93IaGgXsrRTbzao8ayXB5ManvSLv152eiQzNGN4vnLc\nHktFcgX9ZuGxCV/Xpi3xvciyyfcSQWQaLhIZDQ1ABmAkADIAIwGQARgJgAzASABkAEYCIAMw\nEgAZgJEAyACMBEAGYCQAMgAjAZABGAmADMBIAGQARgIgAzASABmAkQDIAIwEQAZgJAAyACMB\nkAEYCYAMwEgAZABGAiADMBIAGYCRAMgAjARABmAkADIAIwGQARipIIywHX/Vajv+vBVD87wN\naKmCvGQk/eX4ILnBjUBzFOSHRmIqhDpjgpW+oYqB9gX5FSMdCeAvgeIF+V0jmb90Wu72qgI6\nF+TvjEQSwK8AcQvy10bS8dDkvwFULUhZI8FROYGYBSluJGZzgJ8BFQtyFyOxvSz4Z6BfQe5l\npP0/8C9At4Lc0khM5UG/+CYQrCB3NxI6x+tAq4K8gZHw+9OLQKSCvJWRsNpLAnUK8nZGgplO\ngTIFeUsj6eug33hAkIK8s5Hw8OQCLQry9kZi6D87EKIgn2MkXbZm6r77wnyYkVjNmxH13vkN\ngJE+h3rv/AbASJ9DvXd+Az7USFX2qSpv+i7ASJ9DlTd9F2Ckz6HKm74LMNLnUOVN3wUY6XOo\n8qYTsCtBBs74sO7fvvZOxvZ+vPaM9XMsb+xi1RhppIn2pgWzFGxJK95O5DtVOMZxqQv5nVoc\nfJ12ABtujUSGkVyujNQq6RvzbeW6L5AW4epoDvNGL1aLkWYWMxIXx5M+Ou/vWsWHTSAKxzgu\ndSU/rcWBadHTeshwi64Bd9wPI7lcGOmL8XmbOfvSXzszqLJuzzCwXv7ThXmjF6vESEKFMHFS\nynAh0tqx4UyjkbWrnIQO4xCFYxyXupaf1OKgO+sAJFyvaqvqYYGRXC6MNDC5yHiaAfJpOtxo\nB0zO1j2Kmzd+sTqMJMwQ6rryTskjO+XqTguUVnX0xTqNKBzBXupafluLg71FQ0g4RjQ9gJFc\nhDoD40r8STSJXpxPYpziqiU7JudzMwMte6ONbPSicD9v1QgXhP2zU4boz9dopqzpta2XrJ03\nNow72ttLvSa/rsXOYh3vRybhzOrPNT+M5MJYJ4fcUbqDmaOHPho2bzBq2aIPOjb1xmmKQRkr\nOnBVyRxRYdZzTMO2B2d9akuAfuxohTfVVo7HZr9EWv7ZXVMeLRpGJuEeZmnnzHVoZRcml+Sj\nfKbkcqx8yiPGnmrS39zWeYjk3UhU9qf2HIxECVQwU4EZuE5XdsJpcib4csrvCk/q+allk1Pi\nO0ZyJyTbomFkGm6Uuw3uTAgjeTC1JNdPAZH2IXKqWd48FEinrWaU3MaOq8EKRiL4KszmUZ3J\nZ/i1P3+SebBu3Wb3GWtXWNtg9VZv3zDS7GwYkBYNI7tDqLeRuMFIPrsz5NTNutks4Jfp0fpG\nauSWLW2m1W619tJTMBLBV2Eww5Tex14Su9RqrznYTDMK73svkUu9Iv/gDJakRcPIJNwo58O1\ndx+M0cou1khy7W5+LGgPUfkhZ6/awGkm+0XtQnEYyeKrsP9cc93dRY/ljyCDUtjZ8jx6vfl8\nRX7nRyPaomFkEq4x05VjfrSyCzGSWCkPjVSrZ8046adQvXeziCmfbFy7ZY9Dmxf4/fnYTOte\nG23mYMo69uDNt8BIL8jvbunRFg0jk3DY/n4Bx0gbaa/FrI7lqDWpTVYru/5pY9Fbo/qwoXmB\n75TjBwMt0uLvblu0oKPt8kThznuMdS71gvzuzxa0RcPIJJyenFZsf6ewRmr0Xl2jNyDM467/\nc7nOPqhFs1pwq9+710620Is/rdeBZ6Ru//lIGEK9uPA8K6gE/WpsBqLwUypMTUYv9YL8XeRH\nLF06jEzCieZeTaOTcueXqRJrpKcenZRwx6Gwl/P7gs6+8uOHJvMmmMrg5a0ab93VHNvOjwuR\njLadLUwU1g+v7ktvx6Wu5be1cJcg0ciNf1k3MozkQpZ26s0GNZ718mBSjalf/PWyq9TGLBPs\noZe3ajwjkQlKyJwUaRHqd/sWn/ogYo9N+Or4Hvta/sA+5CiITMNFIsNIAGQARgIgAzASABmA\nkQDIAIwEQAZgJAAyACMBkAEYCYAMwEgAZABGAiADMBIAGYCRAMgAjARABmAkADIAIwGQARgJ\ngAzASABkAEYCIAMwEgAZgJEAyACMBEAGYCQAMgAjAZABGAmADMBIAGQARgIgAzASABmAkcrB\nHLy/zxP++SVwZ9BO5XjRSPrrRjJtcNjtQHOU46dG0ue2IwcoCOQvR34jMf2fPgf+Euhdjl83\nEmOw1F8BlcvxZ0bSx2Vv9tOBvOX4eyPpvyyNNv8FIGo5ChoJjsoNtCzHDYyERV8uIGI5bmMk\nhtnpx0C/ctzNSNjh+wFQrhw3NhIc9V0gWDlubaQjFbwElCrHOxiJoYu8BlQqx/sYCb3kEkhU\njnczkr5wCaXeAOhSjrc1EjpNCDQpx3sbCXOTA9Qox/sbCWY6gBLl+BwjMTiqegEKAiN9ENUL\nUJBPNFK1/anaG78BH2qkOrtUnXd9D2CkD6LOu74HMNIHUedd3wMY6YOo867vAYz0QdR51/cA\nRvog6rzrc9iVIANnfFj3b197FzMGUIwsmjdyrXqMNLqpc89Yv/iHKZXihUOI5uKwnc4Db6uM\nMycvFY1M2omUTVyoRq6M1CoFG/Nt5Sr77Ag77wdu3ti1qjHS7Oo6qeh8dQ+juNpuVyWI5vrw\ncSr/xlUG66TwUvHIe8twehpGcrkw0hfj8zZz9qW/dswYqbNZxFkWyxu7Vi1GOjQxcKHM2rHB\nPYziaOsXDiCaj6xd5aQzx/JJBtbLf2z48FLxyJrJbVgYyeXCSAOTa4WnGeeepgONZNwT7Wdi\nuHmj16rESFaTzegmXbDKIZ0cxhl9/ZIliOat6ujLuUU5WzenwYNLnURWrNx1HYzkInQdGFdq\nTaL19Rp7EjMPVw3SMbk2N0PXsvePkY02wLC3jZO3aqwmGjJNJGYMDdU2LDE2jNPzRHNzRdZe\n1M0a0rlUIrL57q4tYSQXxjo5FI9SVmaOHvpo2GzrqI+WLfqgY1NvnLbNXqbr7YvPZ/ZUaNj2\n4Kxf3cM4VFu/8KbailqFaP6a/AMxD71UKrK6JX+mQyu7MLmyHuUzJZcj31MeMfZUy7jNlfMh\nkncjObLDSCGOCma04u5hHE9bp8SkHoNaNpGTx0ejppCvpPxPRv1ALpWMrPN63kcruzC1stYj\n2uSd2Rw5yfpBOW09BjcYKcQzktwv6OXzBjk8K+ho65TQvXkli2ei+YN16za3SfnHjpMrk0sl\nI2+y8Xu/nonL1MjuDDnps242y/FlerS+kRq5AUubad13umGkEM9IUtdFz/bHYYrVZnAKH7sz\n7mXUh9rd7q7k7/1nsPU68r714Nxg+jLVYY0kV+Ji9JPrg/YQlR9y9kpKr384nxxGssSEev1J\nhmZwCrs7nq7mYtbij8vI4fbfdeTjGy11cQO1QYwkVspDI4ennjXjpPcV9N7NIqZ8ulntlD0+\nbV7g9ufOdsru20bqAhdu29HdA83nq7kuvPQLkSNbsTCSi2Okjai6mA0GOQ9Naj/XGkn/InE0\nn0a/8D4AAB7JSURBVClt8wK3u2plFvlYTw7jeNo6JbpggUU01wXH83Fsj9z4Cd1F5MiePIzk\nY43U6L26Rm9AmKdW//dtnX2Q8q7Hutl0mss3G2rCMZLovOqlg6dzGMfT1inxlAo7XiGaqxcX\nvppUZJFh7awnyKWSkaXPgl+/YCQXa6Snnm9UkxyHwl7OZqzOvvLjhyabGOStGnfZ+7DKPC5E\notoGhfXDKycvsFrNTcHEwprbQPqhilwqFVkeBz98wUguZGmn3mxQA1AvDybVJqt6B9jLrlKb\n0Uv081aN9/wotN2VIYdRiLZh4bHxXgQnmi+i2brU29+bH5k2Yypy7JEORgIgAzASABmAkQDI\nAIwEQAZgJAAyACMBkAEYCYAMwEgAZABGAiADMBIAGYCRAMgAjARABmAkADIAIwGQARgJgAzA\nSABkAEYCIAMwEgAZgJEAyACMBEAGYCQAMgAjAZABGAmADMBIAGQARgIgAzASABmAkQDIAIxU\njv3PWB0H5ttxZmNon3cBDVWOl4xELLX/AbltM9nAfUBzlOPHRjq+wlrFgfjlyG4k8zdpTFTw\nl0DvcvyykdgRtvSN1gBELsefGcmYCY76RaBtOQoYiaHFfwnIWo5yRtovWvb+PwpoWY4bGEl/\ngp8DFctxFyOZSGXFeHcgXznuZiQ8QP0AKFeOGxsJ89N3gWDluLeR8Pj0LSBVOd7ASAwd5EWg\nUzlgpA8COpXj/YwEY50CXcrxpkZCl4kBVcrx5kZC16FAjXK8v5HM9QBUKAmM9EFAhXLASB8E\nVCgHjPRBQIVywEgfBFQoB4z0QUCFcsBIHwRUKAeM9EFAhXLASB8EVCgHjPRBQAUXdiXIwBkf\nVp1176HbNjZ7KslAU0+uVY2RRjfX3DPWL/u3r2QIX8VVlp3Pclv5V9ISL1Yr2WAknFt9BYzk\ncmWkVvXKRhzNxEiDOuKrk4Gmnl2rFiPNrq6To8zKUyECFblKOHGSlX/R+fgSzxirVrLBSMNO\nkXwwksuFkb4Yn7eZsy/ZCN2eOrN+laNb72U4Uk+vVYmRhCBOLi5EWjs26G9dSvRAxUEeDlZ8\nByJ/r8IP5/IH1Uo2GInsVV8DI7lcGGlgk/j3yR5S8Mee2rGjLMlAUk+vVYeRRtY6KjxVH1wZ\nN99Sogcqcraeq0rkZ1fyB9VKNhiJ7FbfACO5CBUHxpVFJiFzO6kjMWZyNf50TK4U1GQ0sjEs\nSzOQ1MoRHc9RoacLs8XrzichvCymF8uHGtoMRH6zXnS7e7JazqUSkfvYuhKt7MJYJ8fvURqF\nmaOHPho2Z5jr2NQbe2lW1kbGQZVaObNnhIZtD67WUZt89liujeSrOOhRTLUVPUXkf5il3WM7\nw68WvVQqslP94/zVLVSGEG8VHmrkiDbLWbyRaU+1ANk8I3laj3LyD4ykUoHTY81opeaKhxD3\n2kiuik89qomHftFYa0vOUflHudvA/WVDolrkUsnItPr2/NUtVAZTT5Narsk7s3lyPuUW67HA\nW3i3hUbSqcAzknxa7+VkodZKl0byVBw7riaaTj0ureE6Wn3ohcT5hBRUi1wqGZlUn0S6uIXa\nMM/ratuAdbNZDC/To/WNpFn1hqjcw223MINJBZ6RpK6LlK6Rm8hXRoqo2MsBjO6kOJeRH6Oc\ntdY+eJI9rxa5VDIyqT6JlL6F6rBGkutg8zNEe4jKfSMdh62W1cvQOmJXjGek/aNX0/6VkSIq\nqj0zstNpGojI35hJJdkCwZXNpZKRo/uBMJILMZJYKQ+NbIeeNeOkn4j13s0STPlL0+of/pwM\nRyo42Wd2tudPiKtIevS2Hd2dyH+9/R2ePS6VjBzdJoeRXBwjbaS9FrPuliPoJFcN+vcMbZnp\n2HMgGUgqcHqdFmkR8rxgJF/FXfdGbZx6mZ32kQlrYvvbrxa5VDKyrT6NlLpMhVgjNXqvrtEb\nELN+RiK/bw9qDa5+pyOakgwLfGRxeqxwwSofX56xkx6BiupdhbWTzz5PKbZ4GLLrA7d9VtNK\nL1fruFQysl99HSl1mQqxRnrqkfJrfwVL7+c1x573yo9fl3oyrNoM/eVgWxG7Cvrz4f5ykJKI\nqqjzcVtYP7zS1+ka/2R6MHOqRS+VjPyIREYru5ClnXqzQZpHKtx+TWp40q8Uq6zysBl1dtsC\nNsMLy/96cI0ktSW/ZackoiqafIfu6v0D9zVs0j4bOXypWk6DJSO71dcR0hcCALwCjARABmAk\nADIAIwGQARgJgAzASABkAEYCIAMwEgAZgJEAyACMBEAGYCQAMgAjAZABGAmADMBIAGQARgIg\nAzASABmAkQDIAIwEQAZgJAAyACMBkAEYCYAMwEgAZABGAiADMBIAGYCRAMgAjARABmAkADIA\nI5WDuWz2gJ4qXUvwEmincrxmJCfHRvKWrj6goDnK8UMj7RPW/hVtWRKIX478RqKxSt9dZUDv\ncvyykcxXeOpPgMjl+DMjMXOt0jf8yUDcchQw0lEIZAaalqOokeCnvEDMcpQ3kilVWohPACKW\n4zZG0qdKy/HeQL5y3MxIexjwL0C3ctzTSOZkaXHeDQhWjlsbyeYErwClyvEORtpDgQsgUTne\nyEh4fLoC4pTj7YzEMDmdAl3K8aZGgpliQJNyvLORdLbCAt4JaFGO9zcS+bd2oEE5PsdIqhfV\n3ZXqvvuyfJqRqn7PqN47L88nGqna/lTtjd+AzzRSpT2q0tu+BZ9qpCqnpRrv+S7ASB9Ejfd8\nF2CkD6LGe74LMNIHUeM93wUY6YOo8Z5TXHaCgTM+rDrr3u8k434w94z1S5ghdq16jDQGaV/s\nELSd0poHhcNoB0f7EIUSfJHTFyVsy5M2toWTl6mPKyO1SulGHM2O6vN+MKlEvvoZotfybBL4\nJcjxrkaag7SVMyLoI6V5UDiMdmDbZ68/T4Xea7FtlyVsZNLGpHDqMhVyOYDxeZs5+5Kt2dl0\nkWQKcpFh7djgZYhfqxYjWXkOOp0ysnbd1p7N5zIFhSPRdkj7aCZ7GKMLI52UoJFtG1tgJJcL\nIw1MrkKecgQdyTAquoMp+FTyrnJUG9PjbEVGsvIcPE2hVvXMxe2U6cKRaLH2Uaw8OZjttSCc\nlSCRSRtbYCQXIezAuGqHSbSXXr1PYuDiqqk7JlfGaq4Z2WhLDXvvIWMrzVA3Vp6dZTeDSWft\n64VpwtgwTlUm7WO+O+svjyViyb1EInJ0/oSRXBjr5Ag7Sh8wc/TQR8Nm251JZafe2GubjzMN\n2x6c9bIxaIa6mYOZvmWLa6TzfhgUJgmqragJvXBzYqajtSDBTYlUZNLG5HzqQhXC5Jp9lM+U\nXI47T3nE2FOtAjbPSI7WxxmVLmd9L0PduB32IRTVKY0a6L/SK+rgrEmY1ANWyyb/hDVaakKy\ntbCYEsnIpI1JlVI3UCFMrdm1XJN3ZvPkFPZah339dpyRD6K9XEp7GerG6bBqgaRTHqxbt/n8\noScsTBJ0r1/Jpo5rpJn1iaikFiStv45M2phUKXUDFWIXHIMYeGazGF6mR+sbSbPqDVFyRhZZ\n9lSSoW6cDtvInWOTwuXgHtk8Oy1MEsiOi39CfQws9QMVrcXOXiIZOWhjdSJ1AxVCVu4P2cZc\nrjzaQ1QeLOn3Q7cJIxnqhqrQq95qUsTIzh8XIl0baW8gt314KqpTix1+tGYicvS5Dq3s4jwC\nT0Mjh52eNeOkH0v13s0STvl2teGkbjCShqpAttYNc3raPjeSG9Jrn/QPeWEtSIlk5LCNNxjJ\nxzHSRkYgbaSHGsYmubXD1TJ68aTXGRa5xeBlqJszI2mRxrRI589I/tKNtM/V7w8xIx0lkpFJ\nG5NwqRuoEGukRu/VNXoDwjwPk9+3B6npeizDTXuIlbP6pf4ZZKibcF7WKYN8uv9SWn+jsEl4\nytZwXOi82dCl3peIhj5KJCOTNiaBLi9VF9ZITz1cKcsch8Je+5b2yo9fl2zB/UenSIa68VbA\nx5ERKT1rB4X3A/3wysnro7Z95HFq89uJZD5siWTkB7nIEejyUnVBlnbqzQY1tPXyYFKtvap3\ngFVWediMbkFdLJqhbs6MtC1C3O5i0j41knz/wH0Nm7TPS0+np3tEyci2jW2g62sBAK6AkQDI\nAIwEQAZgJAAyACMBkAEYCYAMwEgAZABGAiADMBIAGYCRAMgAjARABmAkADIAIwGQARgJgAzA\nSABkAEYCIAMwEgAZgJEAyACMBEAGYCQAMgAjAZABGAmADMBIAGQARgIgAzASABmAkQDIAIwE\nQAZgpJJ4f87KHGz2O3vl/381uAFop5K8YqQ9WTYV2zb6FdwHNEdJfmQk9b8kCJqyJFC/JJmN\npN2kkmGrPwZ6l+RXjaQPNywC/wSIXJI/MpL+O7hw1G8CbUtSwEhY9v0O0LQkpYy0lyx8+58E\ntCxJcSNhGz0XkLEkNzES1no/BxKW5FZGMkHAPwHhSnJHI7HjNPgGEKwkNzbSnh+8BrQqyf2N\nhP7xIhCqJO9ipCM2OAPilOQNjYQOEwe6lORNjYROEwJNSvLORsIDlAO0KMm7G8kcoBPBSGX5\nJCNVvhdR992XBkb6GOq++9LASB9D3XdfGhjpY6j77ksDI30Mdd99aWCkj6Huuy8NjPQx1H33\npYGRPoa67740MNLHUPfdh1x2h4EzPqw6K9t7l2Dcu5yTPKbD1WMkVwiq0dwz1i9JmTwVx2Zv\ngghH+3gtccIXOX1RwrZ8rM4wksuVkVqldCOOZkf1mblG4k7q+dUqMZIrBJVu0nKd+SKi4pAq\nYdvHbYkTVh4xUryEjRytM4zkcjmA8XmbOfuSzdvZdJFEC04yQ5Aau1odRvKEoNJxIejaseEb\nhftVTlJ9LC9pH81kD2N0Ya1PStDIsTrDSC4XPX9gk/j3yR6yKR9H8shaWnDlXSQ1erUqjOQL\nQaR7qu64JuYNv3B3VCkCaR+FaYkzniwIc1aCRI7WGUZyEcIOjKt2mEQDtpM6EgMXV+NPx+TK\nWI2oIxttqcFp2I6tkdR68YUg0vVs/mbhPVmlycelkaSS9jHfU2vGJTLS7SUSkaN1RkO7MNbJ\nsXeUrc3M0UMfDdsxDsqPjk29sdc2OyPkzGKpFeMLQaRr2PbgarH2amHNyloVSNLaZOZOV3Nq\nySgfexY/9F4iFTlaZzS0ixBvFR4Sz5RcjjtPecTYU60CNs9IjtakRcgwCCMZPCMd0pmBK7kj\nEFNxlCutSTbW2qpFl5NzL5CekB6iXf3QpkQycrTOaGgXpp4mtVyTd2bz5BT2Wod9lWJbZCbP\nwTCSwRGCSMfkM/zak+fNq8KKRT3J6F6/kq0L10hzfEfCoJZpXui9RDJytM5oaJf9uVvtsnaz\nWQwv06P1jaRZ9YYoTRrCcQxEV2dqtpcSL7uKLxZeeavT920WL6f5GLyx0KWR+9de6L1EMnK0\nzmhoF2skuQ4WY498xGwPUXmwZbQf2iT3p4lfr/F7cLZfEIxMLxVuG5Pu7nr67cNTUXvlGS80\nP1ozETlaZzS0CzGSWCkPjRx2etaMk34s1Xs3SzjlW1md35dgJMOZkZKb2SeFl6ZdvPS9uzvt\n47REJGgw69gSycjROqOhXRwjbWQE0kZ6qGFskls7XC2jl0B6ui0OI+04QhDptKAL3R+7KCw3\nAsxRFyzdSPt4LREJGhrpKJGMHK0zGtrFGqnRe3WN3oCY9TMS+X17kJquxzL8aI+O/soAIxkc\nIYh04kljlQ/uz5cLkw78lK0her+deJw3G7rL36j8BjpKJCNH64yGdrFGeurh6mt/uUvv5zXH\nxu3Kj1+XbEGVY/XDAXcFTKV7+D/YXBXuyUSiH145eX20IeGa5Oa3E9p82BLJyLE6o6FdyNJO\nvdmghrZeHkxqeFrVO8AqqzxsRrfg5nkHRjK4j5JUOiEzT/5s6hV2VmRj472GTdrnJfFdI5ES\nyciROqOhAcgAjARABmAkADIAIwGQARgJgAzASABkAEYCIAMwEgAZgJEAyACMBEAGYCQAMgAj\nAZABGAmADMBIAGQARgIgAzASABmAkQDIAIwEQAZgJAAyACMBkAEYCYAMwEgAZABGAiADMBIA\nGYCRAMgAjARABmAkADIAI5WDxdkYPVO6kuA10FDleM1IjNHvmz69wWI3A81Rjh8a6fizjJi9\nbgDEL0d+I5EvpW+uNiB4OX7ZSPZL6RutAYhcjj8z0vG19B1/MNC2HH9vJFMe5AeqlqOckY7c\nIBfQshzljbSfAz8GKpbjPkZieID6KZCvHDczkkkorcqbAt3KcU8j7YHAt4Bg5bi1kWCn7wGp\nyvEORtpzgAugUTneykjYjkgDccrxdkZiWO6dAlnK8aZG2lMABYqU462NxLDYc4AU5fgAI5lP\nACMV5HOMZFJqpu67L8unGYnVPDdVfOvF+UQjVdufqr3xG/CZRqq0R1V627cARvogKr3tW/Cx\nRqqxU9V4z3cBRvogarznuwAjfRA13vNdgJE+iBrvOQW7EmTgjA/r/u2L+akryeDmjVyrHiON\nQdou3dwz1i8JxcfGU1HI2k5nuQ/Nya0nYrtnkw1mT8YCw0guV0ZqlYSN+bZy5qUuXB3yJcgb\nu9bHGslXcQ503aWbtF7nw83gZ9CyPq7aZ78Xfho5qFaywezJ+dCNACO5XBjpi/F5mzn70l87\nnZ2k9mzYZOP3Qd7YtaJ8oJGECr6uRrqNC5HWTqkWZWb9KmeOfk8YWbtua8/mF9pH+vRU/qBa\nyQYjJ2fWhedhJJcLIw1MrimeZjx8mmGJpJry8sPNG71WJUYSfd/XdZfuqSy0ns8bnRVU06qO\nvsSt52u+8kifP6tWssHIyTGWA0ZyEcoOjCuhJiGzXotPYvjkquE6JtdsZkha9oYgqWbBIjuG\nk7dqRKf3jHRIdzKzhCGYd8Ra+a98fhpJNl/zjp2vGYNqOYUTkUdGT+zBXrqNemCskyPrKOVi\n5uihj4bNmXDk2LjoA5L6MEu7h5e3auZAhUO6hm0PrlZvSVbtGwmVVbWVPRVoPp8vGSPVejVy\nx6beDKzk/MUd1AaTK/BRPlNyOVY+5RFjT7UW2VytHyI5MNI2yt0GNZbBSARXBSqd6rKpHQHJ\nyI5tukZNDV+y+KQel1p7ztc8OSEF1SKFk5G1yajLNhjJh6kVuH5umLwzmyOnmuVDI+np67HB\nSA6OCo50crOhP3+QVCzkUefBunWbW92jpU9Wsnh2NZ/tFsUr1XK8koisBtZ1cBd4aGUX64xB\nDJazWcAv06P1jdTIHdnASKNcTIiOMcJIDo4KjnRS4iXxI8Emtwzo6K9+YOjYvmdiNkOcy5iP\ngZ3+3BSrluOVdGRZKbfOaGUX4oyHWqTZH4RkGj/k7FUj6eyc2MsMZY2TCqgKVLqXRpvW6bJi\nmOIPr7vvDeRqzi+ld67LY0aKRw7rjFZ2cZp3GhrpiJ4147SYZ1BprEVM+XRn2aa6y4MjFdBe\n50oXnPZZmjZ88WHWD69uSE/zFzZMI7t2tBnPIkfqDCO5OEba9uFpkxLqJyA5mE5q49T2Bptq\nBi71uwhJBWdG0iIt3pM7ZfLOcTXnj7JHd8HSzdE8uk19Xi1aOBlZ18AbIWEkF2ukRu/VNXoD\nQj/dBj9+6+wkdWDydaxB6n35ZkNNhFOOThFPR+o9hedZwcBj+q0R0zpCYe0pg6N5d/0blVMt\nUjgZWbXu6j2AwUgu1khPPWh+7W976f28xt35NA1BUlt76OWtmr3HunP9tu9ynovUk8d+9c+q\n32ZUXVyLzcnKj2reXG1++9UKmjEe2dTAXWnASC5kaafebFBDWy8PJtV2+uVuL7uTag+9vFVz\naiQpc0ok5htpW0RzdGY2GBvv1XGq+QvbPG61aOFUZHnYeMtGGAmADMBIAGQARgIgAzASABmA\nkQDIAIwEQAZgJAAyACMBkAEYCYAMwEgAZABGAiADMBIAGYCRAMgAjARABmAkADIAIwGQARgJ\ngAzASABkAEYCIAMwEgAZgJEAyACMBEAGYCQAMgAjAZABGAmADMBIAGQARgIgAzBSSdRfstrM\nn7Pa9r9rpU5s9v/lNXgD0FAlecFI6g806QTdWru50HK3As1Rkp8ayWbFX9gsDOQvSWYj7Sf2\nU+DvgN4l+VUjsT0X+AMgdEn+xkhsL4/G/j2gbUkKGImhzX8FiFqSUkai2UEWIGZJyhsJ7Z8J\nCFmSGxkJHeFnQL+S3M1I6A3/DKQryU2NhE7xfaBZSW5tJHSN7wC1SnJ/I+mc4BKoVJL3MRIM\ndQHUKcn7GQl2OgG6lOQdjQQrRYEqJXljI7ENlqJAi5K8u5Gw3DuABiX5FCMd0eul7rsvzacZ\nqeK5qd47vwMfaKRaO1St930PPtFIlc5Kdd71XYCRPoY67/ouwEgfQ513fRdgpI+hzru+CzDS\nx1DnXd8FGOljqPOuz7nsBgNnfFjl0doz1s9+qmBkQd7Tq9VipNFNnaV2iyl7BIqyRlQcz7Mf\nua8jb2PjRT6qlYy8xQLDSC5XRmqVhI085Opw9lJFa+wxnNSTq1VipNlNnVRZvtruzs8kWrTM\nnHbv+byVrOaXkbfhqEVQrWTkGUa65sJIX4zP28zZl2yGXv7Tuamb/GR+3vOr1WGkQxMDF8qs\nHRv279O5SL3KpcQ+i0YINE9Enlm/ysnNRvardRJ5Vo3uASO5XBhpYJP498keckJa9/wkVTRM\na2LQ1NOrVWEkq4nmqfrqekwWK4/0zEMi+hGLRvE1T0Xu/Mh+tc4ij7EmhZFchKwD40qoSbRX\nO6mjTkz4apzqmFxikCFJqU5TRWOYpgnyVovVRNOz2TnfsfMHSW66u+3dNJp8yhmdSK7mqch7\nuKNmTrUSkUdGT+xhrq5TGYx1coAdpVzMHD300bCF4+OgNKWps5+p0l0syuyp0LDtwdXCSp+O\nL6Y0D7O0s5MAiabairU2t6d5MrJmtcVptVKROzb1ZmAl568uVBlCvFV4qJFD4Syn8kamPeXR\nvkqyH0/tLr/5YKQQRwUzWu2TTHraGOVuA3fnABNtko21tmrR5ZywRruckEZbmlQrGVmbjLps\ng5F8mHo41U8Vk3dm8xtq7LgaKWGkSzwjyaf63swyM91ICNHrAfep5Fg8S5+sZPHsan4RWbJw\nWvioVjKyGljXwV3goZVd9sdntTnazWbRvEyPNmKkTa6rRxjpBTwjSV0X88vA4I1YLqOc9Nfe\n67W7tP5GtKt5OrJk5c7q7ajWVWRZ1P1hA63sYo0k18vm94v2EJWHcvIg1XwGeWvGMxL94EmB\nGjM3uL3WN9LeQPwbkSUtDetMOunIwS3BSB7ESGKlPDSyAXvWjNOi0vTezeJN+V6qs2u3YNdO\n4vQ6Z+P5YlszOq8HiXt3dzS/3DBdmpb+zttRIyUjx6qUvlR1OEbaiKraSA+1WJjkckP/jqQW\nAiSVlPZS68bpdVqZRT+uRzeTLXom8H7cOYYqf+nmaH4RWe4oRAqraiUj7y3v2BRGcrFGavRe\nXaM3IGb9jOS/2bB2srG839NNM7/wZkM9OEYSo88qn3ue8kvn/ajkMTD5ftvgDkfHrqlQWPjF\n9mhH84vIi+cjWq1kZFWZ1XsAg5FcrJGeelL/2t/J0vt5jd355PaQpG6207ipdeM+Pz6IMs3F\nFnVr8x5u3A/0OfoeXvN65J5sKATVSkVe+fG7IrnB5E3UB1naqTcb1NDWy4NJDU/6XWSdVxw2\nevFAU20zu6l1423ECG25uxA+x6oYGEm+f+C+rk01v4jMAiPRaqUir7blbbSLuwAAvACMBEAG\nYCQAMgAjAZABGAmADMBIAGQARgIgAzASABmAkQDIAIwEQAZgJAAyACMBkAEYCYAMwEgAZABG\nAiADMBIAGYCRAMgAjARABmAkADIAIwGQARgJgAzASABkAEYCIAMwEgAZgJEAyACMBEAGYCQA\nMgAjAZABGAmADMBIAGQARgIgAzASABmAkQDIAIwEQAZgJAAyACMBkAEYCYAMwEgAZABGKsd/\nObPdN1gdwEjluG/fh5G+DYxUjvv2fRjp28BI5bhv34eRvg2MVI779n0Y6dvASABkAEYCIAMw\nEgAZgJEAyACMBEAGYKRScEH8hHM6/pk+e3y+mM1c84VgrwStExipEPz4xz/B6en4Z/rs8fli\nNhPwpdwvBKsTGKkQZ/2OkzN/ZSSdE0b6CTBSIc773QtG2i7O0uAv+e0FI+11hZGiwEiFeD8j\n7Y9IL1yzQmCkQvzYSC916lf6Pn8t2NW1YCRQgr8xUvr0LwWrEhipED810ot2e6Hvc7Nog5F+\nAoxUiB8aiW8v5Hp9NfbCjPR6sCqBkQrxMyNxN+tfGemFbLUCI5XiJ2828BffMsj+ZsNLwaoE\nRgIgAzASABmAkQDIAIwEQAZgJAAyACMBkAEYCYAMwEgAZABGAiADMBIAGYCRAMgAjARABmAk\nADIAIwGQARgJgAzASABkAEYCIAMwEgAZgJEAyACMBEAGYCQAMgAjAZABGAmADMBIAGQARgIg\nAzASABmAkQDIAIwEQAZgJAAyACMBkAEYCYAMwEgAZABGAiADMBIAGYCRAMgAjARABmAkADIA\nIwGQgf8BcGPv8M7MaXgAAAAASUVORK5CYII="
     },
     "metadata": {
      "image/png": {
       "width": 420,
       "height": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "topPathwaysUp <- fgseaRes[ES > 0][head(order(pval), n=5), pathway]\n",
    "topPathwaysDown <- fgseaRes[ES < 0][head(order(pval), n=5), pathway]\n",
    "topPathways <- c(topPathwaysUp, rev(topPathwaysDown))\n",
    "#Viewing the 5 most significantly up-regulated and down-regulated pathways each with the FGSEA internal plot function\n",
    "plotGseaTable(KEGG_hallmark[topPathways], stats, fgseaRes,\n",
    "              gseaParam=0.5)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Gene Set Enrichment Analysis using GSA"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "This submodule explains Gene Set Analysis (GSA), an Enrichment Analysis, a method that is commonly used to summarize high-dimensional gene expression data sets into sets according to its biological relevance. GSA takes the ranked gene lists from the initial stage of a gene expression analysis and aggregates the genes into sets based on shared biological or functional properties as specified by a reference knowledge base. Such databases often contain phenotype associations, molecular interactions and regulation and are referenced in the analysis of the resultant gene sets to find the relevance of the gene properties to the phenotype of interest.\n",
    "\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "### GSA Enrichment Analysis procedure starts with Data preparation"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "outputs": [],
   "source": [
    "# Install GSA from CRAN\n",
    "suppressMessages({if (!require(\"GSA\"))\n",
    "        suppressWarnings(install.packages(\"GSA\"))\n",
    "})"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "outputs": [],
   "source": [
    "suppressMessages({\n",
    "  library(GSA)\n",
    "})"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "outputs": [],
   "source": [
    "# Loading expression data with groups\n",
    "data <- readRDS(\"./data/GSE48350.rds\")\n",
    "expression_data <- data$expression_data\n",
    "norm_expression_data <- data$norm_expression_data\n",
    "groups <- data$groups"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "outputs": [],
   "source": [
    "expression_data$PROBEID <- rownames(expression_data)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "outputs": [],
   "source": [
    "probeIDs <- rownames(expression_data)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "outputs": [],
   "source": [
    "suppressMessages({\n",
    "  annotLookup <- AnnotationDbi::select(hgu133plus2.db, keys = probeIDs, columns = c('PROBEID', 'GENENAME', 'SYMBOL'))\n",
    "})"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "outputs": [],
   "source": [
    "# Merge DE result  data frame with annotation table\n",
    "new_expression_data = merge(annotLookup, expression_data, by=\"PROBEID\")"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "outputs": [],
   "source": [
    "# Remove NA value\n",
    "new_expression_data <- new_expression_data[!is.na(new_expression_data$SYMBOL),]\n",
    "# Remove duplicated genes symbol\n",
    "new_expression_data <-  new_expression_data[!duplicated(new_expression_data$SYMBOL,fromLast=FALSE),]\n",
    "rownames(new_expression_data) <- new_expression_data$SYMBOL"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "outputs": [],
   "source": [
    "# Drop PROBEID, GENENAME, and SYMBOL columns\n",
    "new_expression_data <- new_expression_data[,-c(1:3)]"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "outputs": [],
   "source": [
    "genenames= rownames(new_expression_data)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# GSA Method Call is Given By\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "GSA(x,y, genesets, genenames,\n",
    "method=c(\"maxmean\",\"mean\",\"absmean\"),\n",
    "resp.type=c(\"Quantitative\",\"Two class unpaired\",\"Survival\",\"Multiclass\",\n",
    "\"Two class paired\", \"tCorr\", \"taCorr\"),\n",
    "censoring.status=NULL,random.seed=NULL, knn.neighbors=10,\n",
    "s0=NULL, s0.perc=NULL,minsize=15,maxsize=500,\n",
    "restand=TRUE,restand.basis=c(\"catalog\",\"data\"),\n",
    "nperms=200,\n",
    "xl.mode=c(\"regular\",\"firsttime\",\"next20\",\"lasttime\"),\n",
    "xl.time=NULL, xl.prevfit=NULL)\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "where:\n",
    "+ x: Data x: p by n matrix of features (expression values), one observation per column (missing values allowed); y: n-vector of outcome measurements\n",
    "+ y: Vector of response values: 1,2 for two class problem, or 1,2,3 ... for multiclass\n",
    "problem, or real numbers for quantitative or survival problems\n",
    "+ genesets: Gene set collection (a list)\n",
    "+ genenames: Vector of genenames in expression dataset\n",
    "+ method: Method for summarizing a gene set: \"maxmean\" (default), \"mean\" or \"absmean\"\n",
    "+ resp.type: Problem type: \"quantitative\" for a continuous parameter; \"Two class unpaired\"\n",
    "; \"Survival\" for censored survival outcome; \"Multiclass\" : more than 2 groups,\n",
    "coded 1,2,3...; \"Two class paired\" for paired outcomes, coded -1,1 (first pair),\n",
    "-2,2 (second pair), etc\n",
    "+ censoring.status:\n",
    "Vector of censoring status values for survival problems, 1 mean death or failure,\n",
    "0 means censored\n",
    "+ random.seed: Optional initial seed for random number generator (integer)\n",
    "+ knn.neighbors: Number of nearest neighbors to use for imputation of missing features values\n",
    "+ s0: Exchangeability factor for denominator of test statistic; Default is automatic\n",
    "choice\n",
    "+ s0.perc: Percentile of standard deviation values to use for s0; default is automatic choice;\n",
    "-1 means s0=0 (different from s0.perc=0, meaning s0=zeroeth percentile of standard deviation values= min of sd values)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "### GSA Enrichment analysis using GO terms"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "outputs": [],
   "source": [
    "genesets = GO_term_hallmark\n",
    "GSA.obj<-GSA(as.matrix(new_expression_data),as.numeric(groups$groups), genenames=genenames, genesets=genesets, resp.type=\"Two class unpaired\", nperms=100)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# List the results from a GSA analysis\n",
    "res <- GSA.listsets(GSA.obj, geneset.names=names(genesets),FDRcut=.5)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "A table of the negative gene sets. “Negative” means that lower expression of most genes in the gene set correlates with higher values of the phenotype y. Eg for two classes coded 1,2, lower expression correlates with class 2. For survival data, lower expression correlates with higher risk, i.e shorter survival (Be careful, this can be confusing!"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "neg.table <-res$negative\n",
    "neg.table"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "A table of the positive gene sets. \"Positive\" means that higher expression of most genes in the gene set correlates with higher values of the phenotype y. See \"negative\" above for more info."
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "A table of the positive gene sets. \"Positive\" means that higher expression of most genes in the gene set correlates with higher values of the phenotype y. See \"negative\" above for more info."
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "pos.table <-res$positive\n",
    "pos.table"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# Individual gene scores from a gene set analysis\n",
    "# look at 10th gene set\n",
    "GSA.genescores(10, genesets, GSA.obj, genenames)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# Plot the result, this function makes a plot of the significant gene sets, based on a call to the GSA (Gene set analysis) function.\n",
    "GSA.plot(GSA.obj, fac=1, FDRcut = 0.5)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Enrichment analysis using KEGG pathways"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "perm= 10 / 100 \n",
      "perm= 20 / 100 \n",
      "perm= 30 / 100 \n",
      "perm= 40 / 100 \n",
      "perm= 50 / 100 \n",
      "perm= 60 / 100 \n",
      "perm= 70 / 100 \n",
      "perm= 80 / 100 \n",
      "perm= 90 / 100 \n",
      "perm= 100 / 100 \n"
     ]
    }
   ],
   "source": [
    "genesets = KEGG_hallmark\n",
    "GSA.obj<-GSA(as.matrix(new_expression_data),as.numeric(groups$groups), genenames=genenames, genesets=genesets, resp.type=\"Two class unpaired\", nperms=100)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "outputs": [],
   "source": [
    "# List the results from a GSA analysis\n",
    "res <- GSA.listsets(GSA.obj, geneset.names=names(genesets),FDRcut=.5)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "A table of the negative gene sets. \"Negative\" means that lower expression of most genes in the gene set correlates with higher values of the phenotype y. Eg for two classes coded 1,2, lower expression correlates with class 2. For survival data, lower expression correlates with higher risk, i.e shorter survival (Be careful, this can be confusing!)\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "outputs": [
    {
     "data": {
      "text/html": "<table class=\"dataframe\">\n<caption>A matrix: 14 × 5 of type chr</caption>\n<thead>\n\t<tr><th scope=col>Gene_set</th><th scope=col>Gene_set_name</th><th scope=col>Score</th><th scope=col>p-value</th><th scope=col>FDR</th></tr>\n</thead>\n<tbody>\n\t<tr><td>15 </td><td>hsa00190</td><td>-1.0545</td><td>0   </td><td>0     </td></tr>\n\t<tr><td>81 </td><td>hsa00970</td><td>-0.8368</td><td>0   </td><td>0     </td></tr>\n\t<tr><td>163</td><td>hsa04260</td><td>-0.6791</td><td>0   </td><td>0     </td></tr>\n\t<tr><td>246</td><td>hsa04932</td><td>-0.6198</td><td>0   </td><td>0     </td></tr>\n\t<tr><td>271</td><td>hsa05016</td><td>-0.7231</td><td>0   </td><td>0     </td></tr>\n\t<tr><td>273</td><td>hsa05020</td><td>-0.6949</td><td>0   </td><td>0     </td></tr>\n\t<tr><td>274</td><td>hsa05022</td><td>-0.5237</td><td>0   </td><td>0     </td></tr>\n\t<tr><td>208</td><td>hsa04714</td><td>-0.5285</td><td>0.01</td><td>0.2658</td></tr>\n\t<tr><td>268</td><td>hsa05010</td><td>-0.5025</td><td>0.01</td><td>0.2658</td></tr>\n\t<tr><td>269</td><td>hsa05012</td><td>-0.7434</td><td>0.01</td><td>0.2658</td></tr>\n\t<tr><td>316</td><td>hsa05208</td><td>-0.4672</td><td>0.01</td><td>0.2658</td></tr>\n\t<tr><td>348</td><td>hsa05415</td><td>-0.6448</td><td>0.01</td><td>0.2658</td></tr>\n\t<tr><td>212</td><td>hsa04723</td><td>-0.7089</td><td>0.02</td><td>0.4557</td></tr>\n\t<tr><td>270</td><td>hsa05014</td><td>-0.5696</td><td>0.02</td><td>0.4557</td></tr>\n</tbody>\n</table>\n",
      "text/markdown": "\nA matrix: 14 × 5 of type chr\n\n| Gene_set | Gene_set_name | Score | p-value | FDR |\n|---|---|---|---|---|\n| 15  | hsa00190 | -1.0545 | 0    | 0      |\n| 81  | hsa00970 | -0.8368 | 0    | 0      |\n| 163 | hsa04260 | -0.6791 | 0    | 0      |\n| 246 | hsa04932 | -0.6198 | 0    | 0      |\n| 271 | hsa05016 | -0.7231 | 0    | 0      |\n| 273 | hsa05020 | -0.6949 | 0    | 0      |\n| 274 | hsa05022 | -0.5237 | 0    | 0      |\n| 208 | hsa04714 | -0.5285 | 0.01 | 0.2658 |\n| 268 | hsa05010 | -0.5025 | 0.01 | 0.2658 |\n| 269 | hsa05012 | -0.7434 | 0.01 | 0.2658 |\n| 316 | hsa05208 | -0.4672 | 0.01 | 0.2658 |\n| 348 | hsa05415 | -0.6448 | 0.01 | 0.2658 |\n| 212 | hsa04723 | -0.7089 | 0.02 | 0.4557 |\n| 270 | hsa05014 | -0.5696 | 0.02 | 0.4557 |\n\n",
      "text/latex": "A matrix: 14 × 5 of type chr\n\\begin{tabular}{lllll}\n Gene\\_set & Gene\\_set\\_name & Score & p-value & FDR\\\\\n\\hline\n\t 15  & hsa00190 & -1.0545 & 0    & 0     \\\\\n\t 81  & hsa00970 & -0.8368 & 0    & 0     \\\\\n\t 163 & hsa04260 & -0.6791 & 0    & 0     \\\\\n\t 246 & hsa04932 & -0.6198 & 0    & 0     \\\\\n\t 271 & hsa05016 & -0.7231 & 0    & 0     \\\\\n\t 273 & hsa05020 & -0.6949 & 0    & 0     \\\\\n\t 274 & hsa05022 & -0.5237 & 0    & 0     \\\\\n\t 208 & hsa04714 & -0.5285 & 0.01 & 0.2658\\\\\n\t 268 & hsa05010 & -0.5025 & 0.01 & 0.2658\\\\\n\t 269 & hsa05012 & -0.7434 & 0.01 & 0.2658\\\\\n\t 316 & hsa05208 & -0.4672 & 0.01 & 0.2658\\\\\n\t 348 & hsa05415 & -0.6448 & 0.01 & 0.2658\\\\\n\t 212 & hsa04723 & -0.7089 & 0.02 & 0.4557\\\\\n\t 270 & hsa05014 & -0.5696 & 0.02 & 0.4557\\\\\n\\end{tabular}\n",
      "text/plain": "      Gene_set Gene_set_name Score   p-value FDR   \n [1,] 15       hsa00190      -1.0545 0       0     \n [2,] 81       hsa00970      -0.8368 0       0     \n [3,] 163      hsa04260      -0.6791 0       0     \n [4,] 246      hsa04932      -0.6198 0       0     \n [5,] 271      hsa05016      -0.7231 0       0     \n [6,] 273      hsa05020      -0.6949 0       0     \n [7,] 274      hsa05022      -0.5237 0       0     \n [8,] 208      hsa04714      -0.5285 0.01    0.2658\n [9,] 268      hsa05010      -0.5025 0.01    0.2658\n[10,] 269      hsa05012      -0.7434 0.01    0.2658\n[11,] 316      hsa05208      -0.4672 0.01    0.2658\n[12,] 348      hsa05415      -0.6448 0.01    0.2658\n[13,] 212      hsa04723      -0.7089 0.02    0.4557\n[14,] 270      hsa05014      -0.5696 0.02    0.4557"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "neg.table <-res$negative\n",
    "neg.table"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "A table of the positive gene sets. \"Positive\" means that higher expression of most genes in the gene set correlates with higher values of the phenotype y. See \"negative\" above for more info."
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "outputs": [
    {
     "data": {
      "text/html": "<table class=\"dataframe\">\n<caption>A matrix: 13 × 5 of type chr</caption>\n<thead>\n\t<tr><th scope=col>Gene_set</th><th scope=col>Gene_set_name</th><th scope=col>Score</th><th scope=col>p-value</th><th scope=col>FDR</th></tr>\n</thead>\n<tbody>\n\t<tr><td>179</td><td>hsa04520</td><td>0.4948</td><td>0   </td><td>0     </td></tr>\n\t<tr><td>200</td><td>hsa04662</td><td>0.3778</td><td>0   </td><td>0     </td></tr>\n\t<tr><td>223</td><td>hsa04810</td><td>0.1843</td><td>0   </td><td>0     </td></tr>\n\t<tr><td>185</td><td>hsa04612</td><td>0.5797</td><td>0.01</td><td>0.3544</td></tr>\n\t<tr><td>194</td><td>hsa04640</td><td>0.5286</td><td>0.01</td><td>0.3544</td></tr>\n\t<tr><td>197</td><td>hsa04658</td><td>0.4347</td><td>0.01</td><td>0.3544</td></tr>\n\t<tr><td>198</td><td>hsa04659</td><td>0.3344</td><td>0.01</td><td>0.3544</td></tr>\n\t<tr><td>343</td><td>hsa05332</td><td>0.891 </td><td>0.01</td><td>0.3544</td></tr>\n\t<tr><td>349</td><td>hsa05416</td><td>0.8136</td><td>0.01</td><td>0.3544</td></tr>\n\t<tr><td>7  </td><td>hsa00053</td><td>0.7591</td><td>0.02</td><td>0.4908</td></tr>\n\t<tr><td>251</td><td>hsa04940</td><td>0.8387</td><td>0.02</td><td>0.4908</td></tr>\n\t<tr><td>289</td><td>hsa05140</td><td>0.5441</td><td>0.02</td><td>0.4908</td></tr>\n\t<tr><td>342</td><td>hsa05330</td><td>0.9232</td><td>0.02</td><td>0.4908</td></tr>\n</tbody>\n</table>\n",
      "text/markdown": "\nA matrix: 13 × 5 of type chr\n\n| Gene_set | Gene_set_name | Score | p-value | FDR |\n|---|---|---|---|---|\n| 179 | hsa04520 | 0.4948 | 0    | 0      |\n| 200 | hsa04662 | 0.3778 | 0    | 0      |\n| 223 | hsa04810 | 0.1843 | 0    | 0      |\n| 185 | hsa04612 | 0.5797 | 0.01 | 0.3544 |\n| 194 | hsa04640 | 0.5286 | 0.01 | 0.3544 |\n| 197 | hsa04658 | 0.4347 | 0.01 | 0.3544 |\n| 198 | hsa04659 | 0.3344 | 0.01 | 0.3544 |\n| 343 | hsa05332 | 0.891  | 0.01 | 0.3544 |\n| 349 | hsa05416 | 0.8136 | 0.01 | 0.3544 |\n| 7   | hsa00053 | 0.7591 | 0.02 | 0.4908 |\n| 251 | hsa04940 | 0.8387 | 0.02 | 0.4908 |\n| 289 | hsa05140 | 0.5441 | 0.02 | 0.4908 |\n| 342 | hsa05330 | 0.9232 | 0.02 | 0.4908 |\n\n",
      "text/latex": "A matrix: 13 × 5 of type chr\n\\begin{tabular}{lllll}\n Gene\\_set & Gene\\_set\\_name & Score & p-value & FDR\\\\\n\\hline\n\t 179 & hsa04520 & 0.4948 & 0    & 0     \\\\\n\t 200 & hsa04662 & 0.3778 & 0    & 0     \\\\\n\t 223 & hsa04810 & 0.1843 & 0    & 0     \\\\\n\t 185 & hsa04612 & 0.5797 & 0.01 & 0.3544\\\\\n\t 194 & hsa04640 & 0.5286 & 0.01 & 0.3544\\\\\n\t 197 & hsa04658 & 0.4347 & 0.01 & 0.3544\\\\\n\t 198 & hsa04659 & 0.3344 & 0.01 & 0.3544\\\\\n\t 343 & hsa05332 & 0.891  & 0.01 & 0.3544\\\\\n\t 349 & hsa05416 & 0.8136 & 0.01 & 0.3544\\\\\n\t 7   & hsa00053 & 0.7591 & 0.02 & 0.4908\\\\\n\t 251 & hsa04940 & 0.8387 & 0.02 & 0.4908\\\\\n\t 289 & hsa05140 & 0.5441 & 0.02 & 0.4908\\\\\n\t 342 & hsa05330 & 0.9232 & 0.02 & 0.4908\\\\\n\\end{tabular}\n",
      "text/plain": "      Gene_set Gene_set_name Score  p-value FDR   \n [1,] 179      hsa04520      0.4948 0       0     \n [2,] 200      hsa04662      0.3778 0       0     \n [3,] 223      hsa04810      0.1843 0       0     \n [4,] 185      hsa04612      0.5797 0.01    0.3544\n [5,] 194      hsa04640      0.5286 0.01    0.3544\n [6,] 197      hsa04658      0.4347 0.01    0.3544\n [7,] 198      hsa04659      0.3344 0.01    0.3544\n [8,] 343      hsa05332      0.891  0.01    0.3544\n [9,] 349      hsa05416      0.8136 0.01    0.3544\n[10,] 7        hsa00053      0.7591 0.02    0.4908\n[11,] 251      hsa04940      0.8387 0.02    0.4908\n[12,] 289      hsa05140      0.5441 0.02    0.4908\n[13,] 342      hsa05330      0.9232 0.02    0.4908"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pos.table <-res$positive\n",
    "pos.table"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "outputs": [
    {
     "data": {
      "text/html": "<table class=\"dataframe\">\n<caption>A matrix: 43 × 2 of type chr</caption>\n<thead>\n\t<tr><th scope=col>Gene</th><th scope=col>Score</th></tr>\n</thead>\n<tbody>\n\t<tr><td>ALDH9A1</td><td>2.707 </td></tr>\n\t<tr><td>HADHA  </td><td>1.725 </td></tr>\n\t<tr><td>ACADVL </td><td>1.609 </td></tr>\n\t<tr><td>ADH1B  </td><td>1.513 </td></tr>\n\t<tr><td>HADHB  </td><td>1.175 </td></tr>\n\t<tr><td>ACAA1  </td><td>0.79  </td></tr>\n\t<tr><td>ALDH2  </td><td>0.768 </td></tr>\n\t<tr><td>ACSBG1 </td><td>0.73  </td></tr>\n\t<tr><td>ALDH3A2</td><td>0.726 </td></tr>\n\t<tr><td>ACSL1  </td><td>0.662 </td></tr>\n\t<tr><td>ALDH7A1</td><td>0.64  </td></tr>\n\t<tr><td>ACSL4  </td><td>0.581 </td></tr>\n\t<tr><td>ECI1   </td><td>0.533 </td></tr>\n\t<tr><td>ACSBG2 </td><td>0.474 </td></tr>\n\t<tr><td>CPT1B  </td><td>0.393 </td></tr>\n\t<tr><td>ADH6   </td><td>0.346 </td></tr>\n\t<tr><td>ACADL  </td><td>0.268 </td></tr>\n\t<tr><td>ADH4   </td><td>0.186 </td></tr>\n\t<tr><td>CPT2   </td><td>0.147 </td></tr>\n\t<tr><td>ADH1C  </td><td>0.146 </td></tr>\n\t<tr><td>ADH5   </td><td>0.136 </td></tr>\n\t<tr><td>ACOX1  </td><td>-0.005</td></tr>\n\t<tr><td>ECI2   </td><td>-0.063</td></tr>\n\t<tr><td>ACSL5  </td><td>-0.069</td></tr>\n\t<tr><td>ACADM  </td><td>-0.109</td></tr>\n\t<tr><td>ADH1A  </td><td>-0.123</td></tr>\n\t<tr><td>CPT1A  </td><td>-0.159</td></tr>\n\t<tr><td>ACADS  </td><td>-0.161</td></tr>\n\t<tr><td>HADH   </td><td>-0.171</td></tr>\n\t<tr><td>ADH7   </td><td>-0.378</td></tr>\n\t<tr><td>ACADSB </td><td>-0.542</td></tr>\n\t<tr><td>ACAA2  </td><td>-0.551</td></tr>\n\t<tr><td>ALDH1B1</td><td>-0.623</td></tr>\n\t<tr><td>ACAT1  </td><td>-0.689</td></tr>\n\t<tr><td>CYP2U1 </td><td>-1.043</td></tr>\n\t<tr><td>ACSL6  </td><td>-1.506</td></tr>\n\t<tr><td>ACOX3  </td><td>-1.508</td></tr>\n\t<tr><td>EHHADH </td><td>-1.76 </td></tr>\n\t<tr><td>ACAT2  </td><td>-1.793</td></tr>\n\t<tr><td>GCDH   </td><td>-1.917</td></tr>\n\t<tr><td>ACSL3  </td><td>-2.432</td></tr>\n\t<tr><td>CPT1C  </td><td>-2.561</td></tr>\n\t<tr><td>ECHS1  </td><td>-2.761</td></tr>\n</tbody>\n</table>\n",
      "text/markdown": "\nA matrix: 43 × 2 of type chr\n\n| Gene | Score |\n|---|---|\n| ALDH9A1 | 2.707  |\n| HADHA   | 1.725  |\n| ACADVL  | 1.609  |\n| ADH1B   | 1.513  |\n| HADHB   | 1.175  |\n| ACAA1   | 0.79   |\n| ALDH2   | 0.768  |\n| ACSBG1  | 0.73   |\n| ALDH3A2 | 0.726  |\n| ACSL1   | 0.662  |\n| ALDH7A1 | 0.64   |\n| ACSL4   | 0.581  |\n| ECI1    | 0.533  |\n| ACSBG2  | 0.474  |\n| CPT1B   | 0.393  |\n| ADH6    | 0.346  |\n| ACADL   | 0.268  |\n| ADH4    | 0.186  |\n| CPT2    | 0.147  |\n| ADH1C   | 0.146  |\n| ADH5    | 0.136  |\n| ACOX1   | -0.005 |\n| ECI2    | -0.063 |\n| ACSL5   | -0.069 |\n| ACADM   | -0.109 |\n| ADH1A   | -0.123 |\n| CPT1A   | -0.159 |\n| ACADS   | -0.161 |\n| HADH    | -0.171 |\n| ADH7    | -0.378 |\n| ACADSB  | -0.542 |\n| ACAA2   | -0.551 |\n| ALDH1B1 | -0.623 |\n| ACAT1   | -0.689 |\n| CYP2U1  | -1.043 |\n| ACSL6   | -1.506 |\n| ACOX3   | -1.508 |\n| EHHADH  | -1.76  |\n| ACAT2   | -1.793 |\n| GCDH    | -1.917 |\n| ACSL3   | -2.432 |\n| CPT1C   | -2.561 |\n| ECHS1   | -2.761 |\n\n",
      "text/latex": "A matrix: 43 × 2 of type chr\n\\begin{tabular}{ll}\n Gene & Score\\\\\n\\hline\n\t ALDH9A1 & 2.707 \\\\\n\t HADHA   & 1.725 \\\\\n\t ACADVL  & 1.609 \\\\\n\t ADH1B   & 1.513 \\\\\n\t HADHB   & 1.175 \\\\\n\t ACAA1   & 0.79  \\\\\n\t ALDH2   & 0.768 \\\\\n\t ACSBG1  & 0.73  \\\\\n\t ALDH3A2 & 0.726 \\\\\n\t ACSL1   & 0.662 \\\\\n\t ALDH7A1 & 0.64  \\\\\n\t ACSL4   & 0.581 \\\\\n\t ECI1    & 0.533 \\\\\n\t ACSBG2  & 0.474 \\\\\n\t CPT1B   & 0.393 \\\\\n\t ADH6    & 0.346 \\\\\n\t ACADL   & 0.268 \\\\\n\t ADH4    & 0.186 \\\\\n\t CPT2    & 0.147 \\\\\n\t ADH1C   & 0.146 \\\\\n\t ADH5    & 0.136 \\\\\n\t ACOX1   & -0.005\\\\\n\t ECI2    & -0.063\\\\\n\t ACSL5   & -0.069\\\\\n\t ACADM   & -0.109\\\\\n\t ADH1A   & -0.123\\\\\n\t CPT1A   & -0.159\\\\\n\t ACADS   & -0.161\\\\\n\t HADH    & -0.171\\\\\n\t ADH7    & -0.378\\\\\n\t ACADSB  & -0.542\\\\\n\t ACAA2   & -0.551\\\\\n\t ALDH1B1 & -0.623\\\\\n\t ACAT1   & -0.689\\\\\n\t CYP2U1  & -1.043\\\\\n\t ACSL6   & -1.506\\\\\n\t ACOX3   & -1.508\\\\\n\t EHHADH  & -1.76 \\\\\n\t ACAT2   & -1.793\\\\\n\t GCDH    & -1.917\\\\\n\t ACSL3   & -2.432\\\\\n\t CPT1C   & -2.561\\\\\n\t ECHS1   & -2.761\\\\\n\\end{tabular}\n",
      "text/plain": "      Gene    Score \n [1,] ALDH9A1 2.707 \n [2,] HADHA   1.725 \n [3,] ACADVL  1.609 \n [4,] ADH1B   1.513 \n [5,] HADHB   1.175 \n [6,] ACAA1   0.79  \n [7,] ALDH2   0.768 \n [8,] ACSBG1  0.73  \n [9,] ALDH3A2 0.726 \n[10,] ACSL1   0.662 \n[11,] ALDH7A1 0.64  \n[12,] ACSL4   0.581 \n[13,] ECI1    0.533 \n[14,] ACSBG2  0.474 \n[15,] CPT1B   0.393 \n[16,] ADH6    0.346 \n[17,] ACADL   0.268 \n[18,] ADH4    0.186 \n[19,] CPT2    0.147 \n[20,] ADH1C   0.146 \n[21,] ADH5    0.136 \n[22,] ACOX1   -0.005\n[23,] ECI2    -0.063\n[24,] ACSL5   -0.069\n[25,] ACADM   -0.109\n[26,] ADH1A   -0.123\n[27,] CPT1A   -0.159\n[28,] ACADS   -0.161\n[29,] HADH    -0.171\n[30,] ADH7    -0.378\n[31,] ACADSB  -0.542\n[32,] ACAA2   -0.551\n[33,] ALDH1B1 -0.623\n[34,] ACAT1   -0.689\n[35,] CYP2U1  -1.043\n[36,] ACSL6   -1.506\n[37,] ACOX3   -1.508\n[38,] EHHADH  -1.76 \n[39,] ACAT2   -1.793\n[40,] GCDH    -1.917\n[41,] ACSL3   -2.432\n[42,] CPT1C   -2.561\n[43,] ECHS1   -2.761"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Individual gene scores from a gene set analysis\n",
    "# look at 10th gene set\n",
    "GSA.genescores(10, genesets, GSA.obj, genenames)\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in xy.coords(x, y, xlabel, ylabel, log):\n",
      "\"2 x values <= 0 omitted from logarithmic plot\"\n"
     ]
    },
    {
     "data": {
      "text/plain": "plot without title",
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAZlBMVEUAAAAil+ZNTU1h0E9l\ntu1oaGh8fHx8wfCMjIyNyfKR3oSampqb0POh45enp6eu5qSysrK46bC9vb3G5PjHx8fO6PnQ\n0NDW7PrW8tHZ2dnc9Njh4eHi9t7l8vzp6ens+erw8PD///8VjzOsAAAACXBIWXMAABJ0AAAS\ndAHeZh94AAAgAElEQVR4nO3diYLyTJlA4aogiLZiiy3a4iC5/5scslcgSyX1plKVnGccmmbJ\n0nznz0IIKgXgTK09AcAWEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAk\nQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAk\nQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAk\nQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAk\nQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAk\nQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAk\nQAAhAQIICRBASIAAQgIEeAhJAZGZ8a9cPpwVRgFIIiRAACEBAggJEEBIgABCAgQQEiCAkAAB\nhAQIICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRAACEBAggJ6KQ/DTyakIAur2ymlERI\nwIc8mreSDoMlhROS47mNAFfmkqcnpEPfc8MJyfMogEaxHlcuiPLLQ1nO+6pdX0mEhL0rlkBl\nSVVMXfsasuURIQGdsmYO2eInLS7qaF7/V1xkiuu9HRES9qzcHjqUKZVLo+zaoSyo7KiOqQ8h\nYafq3QmHoiNzk6hcAuX12CEk7JPOV+rSQ7k/Ls3rSZuAJkSUISTsUL5Gl+ZbQvl6XbURdKhL\nmjpEQsJ21ds1B2OfwaF8Uyi7VqzUpcbm0KT1OQMhYUuqbNJmGZMWu9vqRU1eT1rtpSu2h9xH\nTEjYkCabcv2suCgXOcWbrPlmUbFtNGNbqA8hYTsO1UVrwVSHpMuVuUO5b2Fkj/YkhISYVYud\n+rIVUrlqVx34k6/WVUctNKt+IggJcSrjaTaGqpW5YgFU7tU+VG+yFm8PZdtF+dpds/dBCCEh\nQPXSpfyXXv84VHdWO9mahVLztPrwhOr4uWJ3d/5L/gbSAggJ4amPbKv2XtcrbM3bpalRT3O1\nOP60jsf8/1RXWS2BkBCOKpBy71v9FlCalnEZ20NpvWgyVu100VArpLRZtVtw0gkJwWgOsbYI\n6WDslcsvq1KKi9bSqPp0xIIICaFoRzSyapfWm0NptfCpf2k+XVQukzwgJISiCcliZ0NBV1tE\nulmjq+ryi5CwoPb+5db+5vpqsxv6UK/HWdL5Rd3Q4utvAwgJy2m/49l6B7S+ah6mM/LGTr0V\nVO2Da36rd3SvhZAgynybsz5Qp/it/pSCeVd9vNv4kqjYDmqfca5ZtfO2NdSNkDDXodqDZt5m\nliQbUvluarMAel8orYqQMFPzEbj22pvxwTjLVTtdfDq1WrSk9UpcFUi59NGfIS04f9MQEuYx\nFzAH80YjJLudDcWHG9J6H1wTTtlVveb2sWq33OxNRUiYpzOk9qrdu9Y//LqD1o7rKiNjoVRt\nGzXpmPeHgpAwU9eqXTrYkXnAqJlFahmS+CwIIiTM1bWzYYCuL9L2Nk6zZTS8aic13YsgJHii\n64v0bWdBvZI2uLPBy0TORkjwRXev2m0DIcGb7p0N20BImGbih7M31ksvryH9fp/zrxE7X36X\nGgWWNfUkB1tbg+vlMaTn0fhKvtMio8CyRjN6j6a1T2HTPIZ0UcnPPb/2uCXqssQosCSLw+He\n91ITkuxTcom619fvKlliFFiOzeHZOk0/S9pFRz5DUqrvF7FRYClW20YdIbGzQfQpOZZI0bI+\neCH4AxAW43cb6fbIr7GNFJUJe+p2svjp4HP398nYa3d8LjIKiJM7q++m+X0f6ZK/j5Scv3kf\nKRJkZIkjGzCAjGwREqo9bebl29266whsXVl24uLAIUKoPwd0qC/bdzef89atm6uO3p+wRxwi\nBJ1f6PRQX7bv1mn5dlCrpOKW8mZfkxouDhGCzi/0obmsN43qVbcmpOZDeIRk4A1Z1Kt2zdpd\n+25W7cZxiBCM3Qz12t3b3ZqdDSNYIqFmnAQVE3GI0G6xJJHEIUI7RUayOERojz63a1ifc8SR\nDbvTsXeAjJwR0s50rNKRkQD/IV2PSp1vi44CfchoKd7fRyr3OAzutCMkEWU19apc1/4FMhLi\nO6SLujzT9HFR18+7DTNHAYMuDv6pD0jQ7wfRkZEg3yElKt/v/VTHJUaBhi4uyiPl3s5hnyMj\nQb5DqhY2HCK0NF1c9IZERqJ8h/RVhcQhQkvTQ6t2ZCTMa0jn7+tN/byuPi8cIrQ8m50NEOI1\npHpHglIJhwgtq6zGXPKwFFqQz/eR7vfr9XzOdzlcBjsiJEf1Qsj8gtcD5zJZEEc2bI5xCNCh\nvmhfhzhC2pb2gXSE5A0hbUjHQd2s2nmyVki8jySt+yPf7GzwhJA2gRMnrI1Vu/hRUQAIKXJU\nFAZCillvRWwN+eb3nA2c+1tQ/7KIvQr+eQyJc38LGlqjI6MVeAyJc39LGdwuYnG0Co8hcaZV\nESN7F8hoHd7P2dD1i9goNm9sHx2Lo7WwRIrH+J5uMlqN320kzv09m8X7RSyOVuRz9zfn/p7L\n5l1XMlqV3/eROPf3DHbHLpDRujiyIWyWRwCxOFobIYXM9jg6MlodIQUmPwXd5xe2GlF9fnqP\njtZHSGHR9dkcdfu8qPWvb7ezOAoDIQVFp8XyqLgwitH1vfVvBRZHYSCkoOi0HZI27+gKiYwC\nQUhh0ZNW7VgcBYOQAjNlZwMZhYOQosXiKCSEFCsyCgohxYnFUWAIKUpkFBpCihCLo/AQUnzI\nKECEFBsWR0EipLiQUaAIKTB68F4yChUhBUYP3MfiKFyEFBY9cB8ZBYyQwqJ772FxFDRCCoru\nvYeMwkZIQdE9t7M4Ch0hhUT33E5GwSOkkOi1JwBzEVJA9NoTgNkIKSB6+G42lAJGSOHQ1ZW8\nmNfF4WC2c2BbKWCEFA5d/syLKS4OTTuH+gIBIqRg6PJnVcyhvmjdjCARUjB0+bMnJFbtgkZI\nwdDVle5VO3Y2BI2QQqGbq907GxAyQgqFXnsC4IKQAqHXngA4IaRA6LUnAE4IKQx67QmAG0IK\ng157AuCGkIKg154AOCKkIOj3G9jtHRlCCoF+v4GOYkNIIdDvNxBSbAgpAPr9BjqKDiEFQL/9\nTkfxIaQA6PavdBQhQlqffvudkCJESOvT7V/pKEaEtDrd/pWOokRIq9Ot3+goToS0Nt3+lZDi\nREhr063f6ChShLQy3fqNjmJFSCvT5i90FC1CWpdu/bbTkFQQHOfBy1MCHEUgtPnLTjsK4+Um\npKhp4/peOwrj5SakmGnj+m47CuPlJqSYaeM6Ia2KkCKmjev77SiMl5uQIqabqzvuKIyXm5Di\npZure+4ojJebkOKlm6uEtDZCipZuru66ozBebkKKlq6v7bujMF5uQoqVrq/tvKMwXm5CipWu\nrxFSAAgpUrq+tveOwni5CSlSurqy+47CeLkJKVK6/ElHYbzchBQnXV0hpDBebkKKky5/0pH1\ny63Urbpi8/Cb/UMnTITg0wlJgC5/0lE6IaSkumLx6KOyfuiUiRB8OiEJ0MUPOspYh6S+yytW\nj15kIgSfTkjudPmTkDLWIR3Vo7hi9ehFJkLw6YTkThc/6ChnHdJdnYsr2eX1qJJrcc8lUZfi\n1tv5tQJ4ScsTqmQ/nuqYP+aonq3nzJyI3onz8pQAR7EqXfygo8L7y/3vvx7++u+Oh6n0S/2m\nZUjnPJVTdscpu/aV3fpdnBDoYoT0ujdbjD2yhxrPGZ0Ix3lY5ikBjmJVOr+ko9Lby/23Q+Zv\nnw9TabF4yQq5qdMzfZ6yHXk3ldzTe1J085OmP3loqtrZ8JNvWX2/Hmk8Z3QiHOdhoacEOIo1\n6eIHIZXaL/d/Dn/+X/q/Px/+8/Gw1+Ou6lpcOWdraq+wztnVrIxbs1HUDinN4zu2nzM2EY7z\nsNRTAhzFmnR+SUeV9sv998N/X5f//Vwk5WlkmzrFsqc+r2NZUPHjcfs+vYX09Vq3e9Sre33n\ngiSk6Oj8ko5q7Zf7cDB/mA/LHvervoZCOtWlNCH9vtbtLtnGFSFti84u6KgxJaTX+tndqKe5\nvVz4HK+3x1tIaXLM/jeyR5yQoqOzC0JqTArpoY7FNlK9y8DYRsof8hHSRV3zHQ7n7t0MXRPh\nOA9LPSXAUaxHZxd0ZOjeRvr7x8OKx33n62Y/2a669JrtOGjttftN79U20iNtysp3MxjPGZsI\nx3lY6ikBjmI9OqWjtgl77TKJsTWUPOqr2a2X8tpvtlciOzSveM6xfO/IeM7IRDjOw1JPCXAU\nq9HZBSGZ7N9Hyt3qIxvUV9HEJVGn3/zWL/W6dsuWOb/HJqSfap3OeM7IRDjOw0JPCXAUq9Ep\nHb35OLLhz51HNowNpfuIhbkTsfzTCcmFTunoneu/4exwhuc5e6doxYnwPsZARrEWTUcfHF/u\n8gC7ZNWJICS/dEpIH1xf7utJqaPb8oiQIqPp6FMQLzchxUTTUYcgXm5Ciommow5BvNyEFBHN\nAqlLEC83IUWEBVKnIF5uQoqIpqMuQbzchBQPOuoWxMtNSPEgpG62L3f5xmvPwXJpeWTdxDOs\nTp0IuacT0kx01GNaSH2Hb6dFPlPPsDp1IuSeTkgz6bUnIFT2IWWXz9PgUXXTE5o2EXJPJ6R5\n9NoTEKxpIaXPwcPqCMnvKPzTa09AsCaG1Jxp9VicNfV2Uup0K+6YfIbVqRMh93RCmkWvPQHh\nen+5//FH/cd/dD3OXCKd6rOmXostp2s7JNszrPZNhOM8LPOUAEfh20GvPQXhenu5/6Qzf+p4\nXP7AR76N9FOeqOEn++j5Pfv9mKbN6YXsz7DaMxGO87DQUwIchWd0NKD9cv9T/+H/0v/7g/7n\n5+OqvXbP5tRBp7T5ArJ2SJZnWO2eCMd5WOopAY7CM0Ia0H65/6L/9br8V8ciyXwfyTiZ3UWp\n8/1e/Tb1DKvdE+E4D0s9JcBR+HVgE2lA++XW2vzRepz6uJ7/+E7KN5faIdmdYbV7IiYjJB/o\naJBrSK9VvMvxYxvJ8gyr3RMxGSF5cGCf3aA5IVXbSGfzzreQrM6w2j0RkxGSByyQhnVvI/3l\n83HGA429dsfiS5HqJdK0M6x2T4TjPCz1lABH4RELpBH2e+2MX5r3kX7qs6sWx9pNO8Nq90Q4\nzsNSTwlwFP7Q0Zhp7yNVrknryIbqOzGnnmG1ZyKmCickZVpmFKvIDvnWa09E2D6ObPhD95EN\nXidi+aezRJqEBdKoIF5uQgobC6RxQbzchBQ0OrIQxMtNSCHLPxOrV56I4AXxchNSyFgg2Qji\n5SakgLFAshLEy712SLdz/tW4w/vo3UYRreJkJ3rdiYhAEC/3yiGdyg8kDr9r7DSKaNGRpSBe\n7nVDuqrTMwvpqr7cpqN/FPEiJEtBvNzrhpSoZ1p+GbvbdPSPIlp0ZCuIl3vdkOrj1gnpXXk2\nSL3qRMQhiJd73ZCO5RLpXp77SEgQf1k3dGQviJc7iG2kW6JGzho2fxSRIiR7QbzcK++1O5eH\naw+fM8xpFFGiowmCeLlDeB9JnX/cpmJ4FBGqTpev15yIaATxcq8d0iKC+Ms6oKNJgni5CSlA\nhDRJEC/3+ru/c8nQFwQ4jSJCdDRNEC93GCE9eB+pUX+fmF5xImKiguA4D/OfcmtNBu8j1Qhp\nh1yWSEezo9+VpyocdLRHUttIsmIOqfmiWL3eRMA39toJo6N9kgrpd/B0sCKjiAMh7ZNrSBeh\nnR4Do4gKHe2UY0hNRxbn+583iqg0HRHSvjh/sO8n+9Lbx0mx1y6lox0T2GuXfd3tXfbwb0JC\nZARCupVfzC4o0pDoaL8cQzq/Vu0e6pj+ElKrI0LaG8eQbllA+Sm5OIsQHe2Z6+7v7+y3L5V9\nCbsgQkJkvB7Z8PtdfDL9fBnZxxdjSGZHhLQ7rttIE5ZET/Mg1+GdfBGGREf75vGg1YtKfu75\ntcctGV4XjC+kVkeEtD8C57Wzlah7ff2uBj9RG3lIeqWJwHocQ3qeT9aHNCj7RVl0IbFA2jvn\nVTv7g1Y3vESio93zGNJrG+lWfPnL1raR2h0R0h753P19Mj+aPrhtFXVIep2JwKr8vo90yd9H\nSs7fm3ofiQUS+Ki5OzoCIbl764iQ9sl/SNejUueRz9MSEiLjMaRix165x2H40KKIQqIjZHyH\ndFGXZ5o+LsNfTBZPSO8dEdJOuR4i9P2wf172xKQ4pug5fIrjaEKiIxTc35C1bqn1pc0bOUSI\nkFBwPdbu58u6JVV8BrD8ZROHCNERSgLbSL/fR6uWlDp/X28q+5bM52UThwh9dERIuyWzs+Ge\nvJZLY99rbhyUp1SygUOE6Ag1kZBuJ4tPvb5yu1+v53O+y+Ey/DEmQkJk3EN6fr8WR8fb81WT\n2In0owiJjtBwDek329lwKT5oJHduuxhC+uyIkHbM+aPm6nitVtOG98Sl2zqLEB3B5Po+0thR\nc4ZtnUWIkGDyeDqukbMICX7DtAcdHRHSnnk8HdeWztlAR2jzeDquDZ1FqKsjQto1j6fj2tAS\niQUS3nAWoRlYIOGdx5A2cxYhOsIHziI0HSHhAyc/mYyO8Mk5pNs5W6s7239QdvoowtLZESHt\nnWtIp2LzSCWiJQUcEh2hi2NIV3V6ZiFdp36HbLTvIxESujiGlJ3LpHUuBuuBRBoSHaGTwCFC\ns0KyH0VQujsiJAgcIpQ1dB8+vZbLKELS0xEhQWYb6ZaMnrFh9ihCwgIJPVz32p2tPl9Uiv2D\nfSyQ0EfkfSR1/rF4XvQf7KMj9PJ4ZMPIB/skRrGovo4ICe4fo5jwvNg/RsECCf1cd3+f7M/Z\nEPkH+1ggYYD7WYTU2J6DStxLJDrCENdtpEdx4m+bVbyoP9jX2xEhISOws+FxSZTVKl7MH+xj\ngYRBMnvtrnZn0Ir3g30skDBMYomUr93ZvJM0exRroyOMENlGSi6yn+sLLaT+jggJBYG9dl/W\nJ+SaNYoAEBLGOL+PJLpK1zWK9dERRnk8smHmKFY30BEhoeS8s+HnZHvQ6uxRrIuOYEHk5Ce2\nH6OYN4qVERIsOH+wL8neid3wB/voCDac99oVx89t9qPmQx0REmoCJz9pXxERTEh0BDtiS6Sx\n74+dPYpVERLssI00hI5gib12AwY7IiQY3N9Hsj75yexRrGW4I0KCwePJT8IahQUWSLBGSL1Y\nIMGea0jPS7a7LrnIHnQXQkh0hAkcQ3ok5Sn0N/f9SCMdERJaHEM6qa9sWfS8qLPUFL2PYiUs\nkDAFRzZ0Y4GESRxDyr5oLPPcWEh0hGkcQ7qoU/ZB89/T8HnqXEaxhrGOCAlvOLKhCwskTCRz\nZMNJ9Ei71UNigYSpeEP2Ex1hMkL6REiYzDWk6zFNH0d1lD233aohjXZESPjgGNIt2+2dZHsb\nREtaMyQ6wgzORzb85Odr+JHdbbdiSOMdERI+CRzZcM/eQ9rMG7IskDCHQEhnddtOSCyQMIvz\nqt39lp33ZCurdnSEedx3Nij1nS2Q7L+UeeIofLLoiJDQxXn3d/FtsLLfMxZySHrxiUCMeEPW\nwAIJcxFSg44wm0NI2Y4643vKV54qdzYdERK6EVKNBRLmY9WuwgIJDgipZNURIaGH06pdy8pT\n5YiO4ISQCoQEJ66rduf8a11+ky+h6ekYhQ90BDfOZxGqvmgs6rMI2XVESOjFCSJTOoI75xNE\nbuGrLwkJrpxX7ZLsM+a3JDsEXI7fkOgIzqROECl6Dn2/IVl2REgYIHOCyLPop5H8hkRHEMCR\nDYQEAbsPiY4gYe8h2XZESBi085CsOyIkDCIkO3rBicAG7DskFkgQsuuQ6AhS9hySfUeEhBHO\nId3O+WmLH0LT0zWKpbBAghiRQ4RetyWiJXkJiQUS5DiGdFWnZxbSVYl+ss9HSHQEQc4fo3gW\nH0WK7vNIEzoiJIwS+GDf5kPSi00ENsMxpGO5RMq+tU/Q8iGxQIIomW2kW6KuYpOUeghpSkeE\nhHHOZxEqP9gn+j1ji4dERxAm8j6SOst+PRIhITa7PLKBjiBtjyFN6oiQYMM1pOsxTR9HdfyV\nmqDPUUijI8hzDOmW7ftOsr0NoiUREiLjGNJJ/eTvIf3I7rZbMiQ6wgIEjmzIz/sdzZEN0zoi\nJNgRCOmsbpsNSS8zEdgc51W7+y077Xc0q3YskLAI950NKjvtt1Ki51pdLCQ6wjKcd38n+Tcj\nHWUPbVgqpIkdERJs7esNWULCQnYVEh1hKQ4hRfdlzFM7IiRY21FIdITl7GjVjpCwnP2EREdY\nkFRIv6LffblASJM7IiRM4BrSJZJtJDrCohxDajoK/MgGQsKinE8Q+ZOe1ONxCvzzSHSEZQkc\n/f39Whrdwz5odXpHhIRJBEK6Zee0C3obiY6wNMeQzq9Vu4c6pr+EhF2TOGdD/tUuAX8bxYyO\nCAnTuO7+/s5++1L5ZynkiIZER1je9o9smNMRIWEiQuqiBcePXXA6+lt0SrpGIYAFEnxwDmmR\nnOSGSUfwYuMhzeqIkDAZIX3SUmPHfmw7JBZI8GTTIdERfNlySPM6IiTMsOWTn7BAgjcbDokF\nEvzZ7pENMzsiJMyx2ZDoCD4R0hvtPmrs0FZDoiN4tdGQ5nZESJhnmyHRETwjpBbtOGLs1SZD\noiP4Fk5Icu/uzu6IkDBXOCGJjYKO4B8hGbTLaLFr2wtpfkeEhNk2FxIdYQ2E1NDzn4q921pI\ndIRVbCwkh44ICQ62FRIdYSWEVNEOz8XubSokOsJathSSS0eEBCcbComOsB5CKmiXJwPbCYmO\nsKLNhOTUESHB0VZCcuuIkOCIkDLa6dnAVkJigYR1bSMkOsLKNhGSY0eEBGeEREcQsIWQWCBh\ndRsIiY6wvvhDcu2IkCCAkLTj84F0AyGxQEIIYg+JjhCEyENy7oiQIIKQnIcAxB4SHSEQUYfk\n3hEhQUbMIdERgkFIgAD/IV2PSp1vAqOgI4TDY0jFF1qeiu+2vDiPQqAjQoIU3yFd1OWZpo+L\nujqOgo4QEt8hJeqZXX+qo+MoCAkh8R1S9YXlw19cPj4KOkJQfIf0VYWUOI1CoiNCghyvIZ2/\nrzf187r6vAzvbRgbBR0hMF5DKuRXk6fLKAgJgfH5PtL9fr2ez/kuh8tgR2OjEOmIkCAoxiMb\n6AjBiTAkmY4ICZK8hvT7fc63ks6XX4dRsEBCeDyG9Dyqxmn2KFggIUAeQ7qo5OeeX3vcktm7\nv+kIIfIYUqLu9fX77DdkCQkh8n70d9cvE0ZBRwhSZEskoY4ICcL8biPdHvm12dtIdIRA+dz9\nfTL22h1nHSJESAiU3/eRLvn7SMn5e977SFIdERKkxXRkAx0hWBGFJNYRIUFcRIcIsUBCuOI5\nRIgFEgIWziFCyvT5bDpCyGJ5Q1auI0LCAmI5RIgFEoIWyRKJBRLCFschQnSEwEVxiJBgR4SE\nRURxiBALJIQuhiMbWCAheBGEJNkRIWEZ4YdER4jAWiHZv49ESIhA8CHREWIQ+qqdaEeEhKUE\nHhIdIQ6EBAgI+4N9dIRIBP3BPtmOCAnLCeeDfZ+joCNEI+SPURASohHwB/voCPEId4kk3BEh\nYUnBfrBPuiNCwpKC/WAfCyTEJNQP9rFAQlQCPbKBjhCXMEMS74iQsKydhKSlBwi0hBmS+Ci0\n9ACBln2EpIWHB7whJEDALkLSsoMDPhASIGAPIWnRoQEdCAkQQEiAgB2EpCUHBnQiJEDA9kPS\ngsMCehASIGDzIWm5QQG9CAkQsPWQtNiQgAGEBAjYeEhaakDAIEICBBASIGDbIWmh4QAjCAkQ\nsOmQtMxggFGEBAjYckhaZCiABUICBGw4JC0xEMAKIQECthuSFhgGYImQAAGbDUm7DwKwRkiA\nAEICBGw1JO0+EYA9QgIEbDQkLTARgD1CAgRsMyQtMRGAPUICBGwyJC0yEYA9QgIEbDEkLTMR\ngD1CAgRsMCQtNBGAPUICBBASIGB7IWmpiQDsERIgYHMhabGJAOwREiBgayFpuYkA7BESIGBj\nIWnBiQDsERIgYFshacmJAOwREiBgUyFp0YkA7BESIICQAAFbCknLTgRgj5AAARsKSQtPBGCP\nkAAB2wlJS08EYI+QAAGbCUmLTwRgj5AAAVsJSctPBGCPkAABhAQI2EhIeoGJAOwREiBgGyHp\nJSYCsEdIgIBNhKQXmQjAHiEBArYQkl5mIgB7hAQI2EBIeqGJAOwREiAg/pD0UhMB2CMkQED8\nIQEBICRAACEBAvYV0uGw0ICxdx5DUm1LjOJTK51D/j9AnseQriuE1ErnUF8Awnyu2t2T09Kj\neNNOp/Uba3mQ5HUb6a4ukqPQeuwRb8sgY/nEWh5E+d3ZcFV3uVHo8TdjDznz9+pKfQFIiHev\nna4v2g6G3jU4QoKs7YRk5tNS3tC+i1U7iIo3pHzVrr346TCWFyDCa0i/3+d8z/f58us0ijIN\nrXsD6synLzbAmceQnkfjXaThHeGdozi8G3hI+p5N7yILkOAxpItKfoqddo9bMrwj3BjF4S2O\nzrW0j7Te9tURERbmMaTE2Pd9V8nQQ98XO53LoN4l01tRRITleT3Wru+X/lF8pjKyc8Hci0BE\n8CbMJZL6KGgsn+ZhzeNnTicwnd9tpNsjvza6jdQcgWDXT/3g+kkzpxGYx+fu75Ox1+74HHrk\npH7qp1TPnDt9wGx+30e65O8jJedvt/eRPjT79OZOGuAk4iMbKofimDoiwopiDOltLx4RYX0R\nHiLUPt6UiBCCiA4RKvEJCAQonEOEbE/oQEgIUKhvyA7go0QIT+CHCHViqwjBiXCJBIQnzEOE\nCAmRCfMQIUJCZLZxiBCwshiPbACCQ0iAAEICBKwVkq+vdQG8ICRAAKt2gABCAgQQUq/xrxcp\ncB0AAAc9SURBVF8CKhF+sM8TPf71S0Alvg/2eaLrC2BcOB/skxiFIF1fAOP4GEUfTUewF+MH\n+zxhZwPssUQCBPDBPkAAH+wDBAT6wT4gMtNzWHNxITTuZWdBZujM6kqD8TZwQvIydGZ1pcF4\nGzgheRk6s7rSYLwNnJC8DJ1ZXWkw3gZOSF6GzqyuNBhvAyckL0NnVlcajLeBE5KXoTOrKw3G\n28AJycvQmdWVBuNt4ITkZejM6kqD8TZwQvIydGZ1pcF4GzgheRk6s7rSYLwNnJC8DJ1ZXWkw\n3gbOodmAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQEC\nCAkQ4CekS6KSy7P7htZ918HpsRzM9fj2MMmJHDvL+tBQjPn7eNiEwRjzNzaY6UM3529k6B13\nd83f5Fntnr95s/r+rJ7Zm/l3NHgJqfgKmGPnDa377oPfA2A5mEt+NZn6Z7Eb+n0kpKGhGPP3\n8bAJgzHmb2ww04duzt/I0Dvu7pq/ybPaPX/zZvX9WT2zN/PvaPIR0q9K7uk9Ub8dN7Tue/0c\nmB7LwdzV1zP7T+PXIhN5V+e5QzHm7+NhEwZjzN/YYGYN/dz7sJGhdM/f5Fntnr95s9ox8K7Z\nm/l3bPER0kXdXpc/6rvjBvO+qzoNhWQ5mHMxiKnfcWM59GvzgKlDMebv42ETBmPM39hgZgz9\n2j3pNkPpnr/Js9o9f/Nm9eNZ3bM38+/Y4iOks8q+JNP4j4Fxg3mfugz+87cdTGFqSJZDv6rr\n3KEY8/c5uRMG08zf2GBmDN2Yv5Ghd0xS1/xNntXu+Zs3qx/P6p69mX/H9vQ6PNd6HO/LCOMG\n87778D9/28Hknuq0yESe1e3rtV06ZyjG/H08bMpgcvn8jQ1mxtCN+RsZ+ufdnfM3eVYL7/M3\nb1Y/ntU9ezP/ju1ROTzXehz2BYiFdM2X1vITeS42VvsqHStALKR8/hYJqZ6/ySF93OgQ0vv8\niYXUNXuE1DeY9JFMXUxbDl2pn9d/Li99K3i+Qirmb4GQjPlbM6SP+RMKqXv2CKlvMM9k4ord\nxOXds29XqaeQyvlbIKRy+Nn8rRjS5/wJhVQO/m32YgkpeZ9O44a3+4bmZcJgTtPfEpgw9IHp\nHBqK8fNzgNMGU87f2GBmDj3tm+vhoRi/Df7tLAbzOX/zZrXvWW9DnPl3bA/S4bm2ip0ij/e9\nRI9mh0l939C8WA/mcTw9lpzIgekcGorxvM8BThlMPX9jg5k19Ho6R4beeffn/E2e1e75mzer\nfc96m72Zf8f2IB2ea+s73/C/qUvHDW/3DYVkO5jb1B12U4aeqGc68DcfGooxfx8PmzKYZv7G\nBjNj6Mb8jQy98+7P+Zs8q93zN29WP57VPXsz/44tPkKyPrJhOCTLwTxmdWQ79Ev2135e+vYJ\nDh/ZUM+fy5ENxvwtcGSDMX/Tj2zomr/Js9o9f0JHNnTPXixHNqTHZpdx8Zc2bji2dicPrqba\nDeZLjRwO5zT0Z5Jf7f1v19BQjPlr3zppMOb8jQ1m+tDN+RsZ+sdQuudv6qz2zN+8WX0feM/s\nzfw7mryE9MwPri3Gp95uMK6mIyHZDUYZL8QCE5ldPfYf3TA0FGP+2rdOGow5f2ODmTmr5fyN\nDP1jKN3zN3VWe+Zv3qx2zunn7M38O5q8hARsHSEBAggJEEBIgABCAgQQEiCAkAABhAQIICRA\nACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRA\nACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRAACFFz+lb7SGEFyF6hBQCXoToEVIIeBGiR0gh\n4EUI2CuRi/lt2091zH8e1TO9nVX9NedVTMXl9aiS/q9dxzIIKWBKfauXU33DST1el4/XLfkd\nSl3Sj5DOqv0ceEFIAXstc+7pPVE/1Q0/6vt1+a1ur7t+sl9V+h7STZ2e6fP0egh8IqSAqTyH\nmzo3t2TrdkfVPCB9D+n8Wu3LVgLPKXwipICVuxGMvQlfr3W7R75Clz5u36eOkFRlhendM/7e\nAWtCqtr4fa3bXdRvmm0uVbkQUgj4ewfsM6Q0OWb/yxZNx+vt0RnSetO7Z/zZA6byZc9NfTU3\nXdQ13+GQ9/IW0m+xjcRuhjUQUsCqvXZGGq928r0JWWP3ZhvpqK7ZrjqV7cl7PSe9srPBM0IK\nmFL5llCriWPxFtGl3BL6LUK65o/Luyo2npLHKlO8X4QUsFcY59e2UOu2n3L59PWq7DffM57n\n8528VgDrIxvUFx15RkgBY8dBPHipAkZI8eClChghxYOXKmCEFA9eKkAAIQECCAkQQEiAAEIC\nBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEIC\nBBASIICQAAGEBAggJEAAIQECCAkQ8P+J954m+tf07gAAAABJRU5ErkJggg=="
     },
     "metadata": {
      "image/png": {
       "width": 420,
       "height": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot the result, this function makes a plot of the significant gene sets, based on a call to the GSA (Gene set analysis) function.\n",
    "GSA.plot(GSA.obj, fac=1, FDRcut = 0.5)\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "environment": {
   "kernel": "ir",
   "name": "r-cpu.4-1.m93",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/r-cpu.4-1:m93"
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  },
  "toc-showcode": true,
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
