
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Processing sequencing data from NCBI Gene Expression Omnibus (GEO) &#8212; Consensus Pathway Analysis using Google Cloud Infrastructure</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=62ba249389abaaa9ffc34bf36a076bdc1d65ee18" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=f31d14ad54b65d19161ba51d4ffff3a77ae00456"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Data Processing" href="Module1-Data_Processing2.html" />
    <link rel="prev" title="Setting up Google Cloud environment" href="VertexAI.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/Cover.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Consensus Pathway Analysis using Google Cloud Infrastructure</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Overview
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting Started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="VertexAI.html">
   Setting up Google Cloud environment
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 01
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Processing sequencing data from NCBI Gene Expression Omnibus (GEO)
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 02
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Module1-Data_Processing2.html">
   Data Processing
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://mybinder.org/v2/gh/executablebooks/jupyter-book/master?urlpath=tree/docs/Module1-Data_Processing.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Launch on Binder"
>
  

<span class="headerbtn__icon-container">
  
    <img src="_static/images/logo_binder.svg">
  </span>
<span class="headerbtn__text-container">Binder</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/executablebooks/jupyter-book"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2FModule1-Data_Processing.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="_sources/Module1-Data_Processing.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#learning-objectives">
   Learning Objectives:
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#accessing-public-geo-sequencing-data">
   Accessing public GEO sequencing data
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#searching-and-accessing-data-on-geo">
     Searching and accessing data on GEO
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#downloading-data-using-web-interface">
     Downloading data using web-interface
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#downloading-data-using-r-command-line">
     Downloading data using R command line
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Processing sequencing data from NCBI Gene Expression Omnibus (GEO)</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#learning-objectives">
   Learning Objectives:
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#accessing-public-geo-sequencing-data">
   Accessing public GEO sequencing data
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#searching-and-accessing-data-on-geo">
     Searching and accessing data on GEO
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#downloading-data-using-web-interface">
     Downloading data using web-interface
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#downloading-data-using-r-command-line">
     Downloading data using R command line
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="processing-sequencing-data-from-ncbi-gene-expression-omnibus-geo">
<h1>Processing sequencing data from NCBI Gene Expression Omnibus (GEO)<a class="headerlink" href="#processing-sequencing-data-from-ncbi-gene-expression-omnibus-geo" title="Permalink to this headline">#</a></h1>
<p><img alt="" src="_images/Data_Processing.png" /></p>
<section id="learning-objectives">
<h2>Learning Objectives:<a class="headerlink" href="#learning-objectives" title="Permalink to this headline">#</a></h2>
<ol class="simple">
<li><p>Understand the type of data that is accessible from GEO.</p></li>
<li><p>Demonstrate how to navigate the GEO website, search for dataset using accession number, and select samples.</p></li>
<li><p>Use the command-line interface to copy over data from GEO.</p></li>
</ol>
</section>
<section id="accessing-public-geo-sequencing-data">
<h2>Accessing public GEO sequencing data<a class="headerlink" href="#accessing-public-geo-sequencing-data" title="Permalink to this headline">#</a></h2>
<p>The Gene Expression Omnibus (GEO) is a public repository that archives and freely distributes comprehensive sets of microarray,
next-generation sequencing, and other forms of high-throughput functional genomic data submitted by the scientific community.
In addition to data storage, the website prodide set of web-based tools and applications to help users query and download the
studies and gene expression patterns stored in GEO.</p>
<section id="searching-and-accessing-data-on-geo">
<h3>Searching and accessing data on GEO<a class="headerlink" href="#searching-and-accessing-data-on-geo" title="Permalink to this headline">#</a></h3>
<p>Searching for GEO is relatively straight forward via GEO website. First, users need to navigate to  <a href="https://www.ncbi.nlm.nih.gov/geo/">https://www.ncbi.nlm.nih.gov/geo/</a>.
There are multiple ways for searching datasets but the simples way is to provide the accession number in the search box.
We will use an example dataset with the accession number <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE48350">GSE48350</a> for demonstration.
The GEO website interface with searching procedure for the example date are shown in the figure below:</p>
<p><img alt="" src="_images/GEO_Website.png" /></p>
<p>When the searching process is done, a webpage with detailed description of the example dataset such as published date, title, organism, experiment type, dataset summary etc. will be shown in the figure below:</p>
<p><img alt="" src="_images/GEO_Dataset_Page.png" /></p>
<p>We can see that the dataset GSE48350 was published in Apr 21, 2014 which focuses on Human Alzheimer’s
Disease using microarray sequencing technology. Samples were primary collected from 4 brain regions:
hippocampus (HC), entorhinal cortex (EC), superior frontal cortex (SCG), post-central gyrus (PCG)
of normal and disease patients.</p>
</section>
<section id="downloading-data-using-web-interface">
<h3>Downloading data using web-interface<a class="headerlink" href="#downloading-data-using-web-interface" title="Permalink to this headline">#</a></h3>
<p>At the bottom of the dataset page, users will find additional information about the dataset such as sequencing platform, number of samples, project ID and links to download the expression data. The exact places to find this information are shown in the figure below:</p>
<p><img alt="" src="_images/Manual_Download.png" /></p>
<p>As presented in the figure, the dataset GSE48350 were sequenced using Affymetrix Human Genome U133 Plus 2.0 Array and the whole dataset contains 253 samples. The summary table at the bottom of the page show the all the data generated rom the experiment.  User can click to “(http)” hyperlink to download all the samples or click “(custom)” to select and download the samples of interest. Note that, expression data download from this step are raw data and additional data processing need to be done locally for further analysis.</p>
</section>
<section id="downloading-data-using-r-command-line">
<h3>Downloading data using R command line<a class="headerlink" href="#downloading-data-using-r-command-line" title="Permalink to this headline">#</a></h3>
<p>If the users have programming background, it is more beneficial to automate the downloading task using R’s command line. Getting data from GEO is really quite easy using <a href="https://bioconductor.org/packages/release/bioc/html/GEOquery.html">GEOquery</a> R package available in <a href="https://bioconductor.org/">Bioconductor</a>. Before starting, user need to install GEOquery package using the following command.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">,</span> <span class="n">quietly</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">))</span>
    <span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>
<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;GEOquery&quot;</span><span class="p">)</span>

<span class="nf">library</span><span class="p">(</span><span class="s">&quot;GEOquery&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Bioconductor version &#39;3.14&#39; is out-of-date; the current release version &#39;3.15&#39;
  is available with R version &#39;4.2&#39;; see https://bioconductor.org/install
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;getOption(&quot;repos&quot;)&#39; replaces Bioconductor standard repositories, see
&#39;?repositories&#39; for details

replacement repositories:
    CRAN: https://cran.r-project.org
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Bioconductor version 3.14 (BiocManager 1.30.16), R 4.1.3 (2022-03-10)
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Warning message:
&quot;package(s) not installed when version(s) same as current; use `force = TRUE` to
  re-install: &#39;GEOquery&#39;&quot;
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Old packages: &#39;ade4&#39;, &#39;AnnotationHub&#39;, &#39;BiocFileCache&#39;, &#39;BiocManager&#39;, &#39;blob&#39;,
  &#39;broom&#39;, &#39;car&#39;, &#39;cli&#39;, &#39;clipr&#39;, &#39;clusterSim&#39;, &#39;colorspace&#39;, &#39;crayon&#39;, &#39;DBI&#39;,
  &#39;dbplyr&#39;, &#39;desc&#39;, &#39;devEMF&#39;, &#39;doParallel&#39;, &#39;dplyr&#39;, &#39;DT&#39;, &#39;e1071&#39;, &#39;evaluate&#39;,
  &#39;exactRankTests&#39;, &#39;ExperimentHub&#39;, &#39;extrafont&#39;, &#39;fansi&#39;, &#39;farver&#39;,
  &#39;fitdistrplus&#39;, &#39;flextable&#39;, &#39;FNN&#39;, &#39;formatR&#39;, &#39;future&#39;, &#39;future.apply&#39;,
  &#39;gdtools&#39;, &#39;generics&#39;, &#39;GenomeInfoDb&#39;, &#39;gert&#39;, &#39;glmnet&#39;, &#39;globals&#39;, &#39;glue&#39;,
  &#39;GPArotation&#39;, &#39;gplots&#39;, &#39;gtools&#39;, &#39;gWidgets2tcltk&#39;, &#39;haven&#39;, &#39;httr&#39;,
  &#39;ie2misc&#39;, &#39;igraph&#39;, &#39;insight&#39;, &#39;keras&#39;, &#39;km.ci&#39;, &#39;knitr&#39;, &#39;leiden&#39;, &#39;limma&#39;,
  &#39;lme4&#39;, &#39;lmtest&#39;, &#39;locfit&#39;, &#39;magrittr&#39;, &#39;maptools&#39;, &#39;matrixStats&#39;, &#39;mclust&#39;,
  &#39;missForest&#39;, &#39;mnormt&#39;, &#39;nloptr&#39;, &#39;officer&#39;, &#39;openssl&#39;, &#39;packageRank&#39;,
  &#39;parallelly&#39;, &#39;parsedate&#39;, &#39;pillar&#39;, &#39;pkgload&#39;, &#39;pkgsearch&#39;, &#39;plyr&#39;,
  &#39;polynom&#39;, &#39;processx&#39;, &#39;proxy&#39;, &#39;ps&#39;, &#39;psych&#39;, &#39;quantmod&#39;, &#39;quantreg&#39;,
  &#39;R.methodsS3&#39;, &#39;R.oo&#39;, &#39;R.utils&#39;, &#39;R2HTML&#39;, &#39;randomForest&#39;, &#39;ranger&#39;,
  &#39;RColorBrewer&#39;, &#39;RcppArmadillo&#39;, &#39;RcppEigen&#39;, &#39;RCurl&#39;, &#39;readr&#39;, &#39;readxl&#39;,
  &#39;reshape&#39;, &#39;reticulate&#39;, &#39;rgl&#39;, &#39;rhdf5&#39;, &#39;rlang&#39;, &#39;rmarkdown&#39;, &#39;roxygen2&#39;,
  &#39;rprojroot&#39;, &#39;rrtable&#39;, &#39;RSpectra&#39;, &#39;RSQLite&#39;, &#39;Rtsne&#39;, &#39;Rttf2pt1&#39;,
  &#39;S4Vectors&#39;, &#39;sass&#39;, &#39;scales&#39;, &#39;scattermore&#39;, &#39;scCAN&#39;, &#39;SCFA&#39;, &#39;scMappR&#39;,
  &#39;Seurat&#39;, &#39;SeuratObject&#39;, &#39;shinyWidgets&#39;, &#39;sjlabelled&#39;, &#39;sp&#39;,
  &#39;spatstat.core&#39;, &#39;spatstat.data&#39;, &#39;spatstat.geom&#39;, &#39;spatstat.sparse&#39;,
  &#39;spatstat.utils&#39;, &#39;survMisc&#39;, &#39;systemfonts&#39;, &#39;tensorflow&#39;, &#39;testthat&#39;,
  &#39;tibble&#39;, &#39;tidyr&#39;, &#39;tinytex&#39;, &#39;torch&#39;, &#39;tzdb&#39;, &#39;umap&#39;, &#39;usethis&#39;, &#39;uuid&#39;,
  &#39;vcd&#39;, &#39;vctrs&#39;, &#39;waldo&#39;, &#39;xfun&#39;, &#39;XML&#39;, &#39;zoo&#39;
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Loading required package: Biobase
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Loading required package: BiocGenerics
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Attaching package: &#39;BiocGenerics&#39;
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>The following objects are masked from &#39;package:stats&#39;:

    IQR, mad, sd, var, xtabs
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>The following objects are masked from &#39;package:base&#39;:

    anyDuplicated, append, as.data.frame, basename, cbind, colnames,
    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
    union, unique, unsplit, which.max, which.min
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    &#39;browseVignettes()&#39;. To cite Bioconductor, see
    &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Setting options(&#39;download.file.method.GEOquery&#39;=&#39;auto&#39;)
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Setting options(&#39;GEOquery.inmemory.gpl&#39;=FALSE)
</pre></div>
</div>
</div>
</div>
<p>We can use the <code class="docutils literal notranslate"><span class="pre">getGEO</span></code> function from the <code class="docutils literal notranslate"><span class="pre">GEOquery</span></code> package to download GEO dataset. First, users have to specify the accession ID of the dataset. For demonstration, we will use the same dataset <code class="docutils literal notranslate"><span class="pre">GSE48350</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">## change my_id to be the dataset that you want.</span>
<span class="n">my_id</span> <span class="o">&lt;-</span> <span class="s">&quot;GSE48350&quot;</span>
<span class="n">gse</span> <span class="o">&lt;-</span> <span class="nf">getGEO</span><span class="p">(</span><span class="n">my_id</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Found 1 file(s)
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>GSE48350_series_matrix.txt.gz
</pre></div>
</div>
</div>
</div>
<p>Some datasets on GEO may be derived from different microarray platforms. Therefore the object <code class="docutils literal notranslate"><span class="pre">gse</span></code> is a list of different datasets.
You can find out how many were used by checking the length of the <code class="docutils literal notranslate"><span class="pre">gse</span></code> object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">## check how many platforms used</span>
<span class="nf">length</span><span class="p">(</span><span class="n">gse</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">1</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">The</span> <span class="n">result</span> <span class="n">shows</span> <span class="n">that</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="n">Error</span> <span class="ow">in</span> <span class="n">parse</span><span class="p">(</span><span class="n">text</span> <span class="o">=</span> <span class="n">x</span><span class="p">,</span> <span class="n">srcfile</span> <span class="o">=</span> <span class="n">src</span><span class="p">):</span> <span class="o">&lt;</span><span class="n">text</span><span class="o">&gt;</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">5</span><span class="p">:</span> <span class="n">unexpected</span> <span class="n">symbol</span>
<span class="mi">1</span><span class="p">:</span> <span class="n">The</span> <span class="n">result</span>
        <span class="o">^</span>
<span class="ne">Traceback</span>:
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">gse</span> <span class="o">&lt;-</span> <span class="n">gse</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span>
<span class="n">gse</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">Usually</span> <span class="n">there</span> <span class="n">will</span> <span class="n">only</span> <span class="n">be</span> <span class="n">one</span> <span class="n">platform</span> <span class="n">and</span> <span class="n">the</span> <span class="n">dataset</span> <span class="n">we</span> <span class="n">want</span> <span class="n">to</span> <span class="n">analyse</span> <span class="n">will</span> <span class="n">be</span> <span class="n">the</span> <span class="n">first</span> <span class="n">object</span> <span class="n">in</span> <span class="n">the</span> <span class="nf">list </span><span class="p">(</span><span class="n">gse</span><span class="p">[[</span><span class="m">1</span><span class="p">]])</span><span class="n">.</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">summary</span><span class="p">(</span><span class="nf">exprs</span><span class="p">(</span><span class="n">gse</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>   GSM300166          GSM300167          GSM300168         GSM300169       
 Min.   :  0.0100   Min.   :  0.0100   Min.   : 0.0100   Min.   : 0.01942  
 1st Qu.:  0.6366   1st Qu.:  0.6519   1st Qu.: 0.3342   1st Qu.: 0.32045  
 Median :  0.8622   Median :  0.8828   Median : 0.7756   Median : 0.75063  
 Mean   :  0.9193   Mean   :  0.9492   Mean   : 0.9274   Mean   : 0.87928  
 3rd Qu.:  1.0903   3rd Qu.:  1.0880   3rd Qu.: 1.1929   3rd Qu.: 1.15987  
 Max.   :335.6746   Max.   :531.7116   Max.   :99.4657   Max.   :26.46015  
   GSM300170          GSM300171          GSM300172          GSM300173       
 Min.   : 0.01098   Min.   :  0.0100   Min.   :  0.0100   Min.   : 0.04317  
 1st Qu.: 0.32324   1st Qu.:  0.3139   1st Qu.:  0.3461   1st Qu.: 0.68439  
 Median : 0.77234   Median :  0.7294   Median :  0.7358   Median : 0.92708  
 Mean   : 0.93768   Mean   :  1.0313   Mean   :  0.9851   Mean   : 0.95096  
 3rd Qu.: 1.21978   3rd Qu.:  1.1992   3rd Qu.:  1.1835   3rd Qu.: 1.12190  
 Max.   :83.41583   Max.   :693.3533   Max.   :721.7593   Max.   :10.93424  
   GSM300174           GSM300175          GSM300176          GSM300177       
 Min.   :  0.03664   Min.   : 0.03605   Min.   : 0.04455   Min.   : 0.03746  
 1st Qu.:  0.65056   1st Qu.: 0.66651   1st Qu.: 0.72507   1st Qu.: 0.66970  
 Median :  0.89621   Median : 0.92130   Median : 0.95680   Median : 0.90137  
 Mean   :  0.93030   Mean   : 0.95004   Mean   : 0.95321   Mean   : 0.90062  
 3rd Qu.:  1.07786   3rd Qu.: 1.13891   3rd Qu.: 1.11961   3rd Qu.: 1.07227  
 Max.   :126.99336   Max.   :18.39787   Max.   :36.52705   Max.   :23.44349  
   GSM300178          GSM300179          GSM300180          GSM300181      
 Min.   : 0.03513   Min.   : 0.03495   Min.   : 0.03819   Min.   : 0.0411  
 1st Qu.: 0.62134   1st Qu.: 0.65650   1st Qu.: 0.68178   1st Qu.: 0.6630  
 Median : 0.86038   Median : 0.87945   Median : 0.90843   Median : 0.8929  
 Mean   : 0.91505   Mean   : 0.89371   Mean   : 0.91517   Mean   : 0.9119  
 3rd Qu.: 1.05592   3rd Qu.: 1.03777   3rd Qu.: 1.08511   3rd Qu.: 1.0666  
 Max.   :43.74532   Max.   :20.74707   Max.   :19.37128   Max.   :17.1122  
   GSM300182          GSM300183          GSM300185           GSM300186       
 Min.   : 0.03787   Min.   : 0.03533   Min.   :  0.01456   Min.   :  0.0100  
 1st Qu.: 0.64043   1st Qu.: 0.67396   1st Qu.:  0.32666   1st Qu.:  0.6251  
 Median : 0.88555   Median : 0.91363   Median :  0.75186   Median :  0.9259  
 Mean   : 0.90504   Mean   : 0.90907   Mean   :  0.93242   Mean   :  1.0681  
 3rd Qu.: 1.08388   3rd Qu.: 1.08255   3rd Qu.:  1.21301   3rd Qu.:  1.2476  
 Max.   :19.30724   Max.   :20.77829   Max.   :104.92921   Max.   :143.7138  
   GSM300187          GSM300188          GSM300189          GSM300190       
 Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.0100  
 1st Qu.:  0.5463   1st Qu.:  0.6357   1st Qu.:  0.6008   1st Qu.:  0.5629  
 Median :  0.8130   Median :  0.8885   Median :  0.8534   Median :  0.8442  
 Mean   :  0.9692   Mean   :  0.9681   Mean   :  1.0102   Mean   :  1.0721  
 3rd Qu.:  1.1536   3rd Qu.:  1.1408   3rd Qu.:  1.1191   3rd Qu.:  1.1752  
 Max.   :139.6595   Max.   :196.0111   Max.   :184.4251   Max.   :291.6673  
   GSM300191          GSM300192          GSM300193          GSM300194       
 Min.   :  0.0100   Min.   : 0.02184   Min.   :  0.0100   Min.   :  0.0100  
 1st Qu.:  0.6185   1st Qu.: 0.66305   1st Qu.:  0.5661   1st Qu.:  0.6583  
 Median :  0.8721   Median : 0.89910   Median :  0.8236   Median :  0.8922  
 Mean   :  0.9843   Mean   : 0.96226   Mean   :  0.9548   Mean   :  0.9116  
 3rd Qu.:  1.1592   3rd Qu.: 1.10925   3rd Qu.:  1.0829   3rd Qu.:  1.0688  
 Max.   :184.3417   Max.   :98.38803   Max.   :114.6595   Max.   :145.5392  
   GSM300195          GSM300196          GSM300197          GSM300198       
 Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.0120   Min.   :  0.0100  
 1st Qu.:  0.6674   1st Qu.:  0.6374   1st Qu.:  0.6151   1st Qu.:  0.6305  
 Median :  0.8947   Median :  0.8918   Median :  0.8844   Median :  0.8990  
 Mean   :  0.9360   Mean   :  1.1227   Mean   :  1.0836   Mean   :  1.0356  
 3rd Qu.:  1.0982   3rd Qu.:  1.1271   3rd Qu.:  1.1442   3rd Qu.:  1.1250  
 Max.   :141.9029   Max.   :496.3890   Max.   :399.4686   Max.   :625.9273  
   GSM300199          GSM300200          GSM300201           GSM300202       
 Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.01169   Min.   :  0.0100  
 1st Qu.:  0.6496   1st Qu.:  0.6404   1st Qu.:  0.66185   1st Qu.:  0.5735  
 Median :  0.8929   Median :  0.8944   Median :  0.89072   Median :  0.8445  
 Mean   :  1.0279   Mean   :  0.9511   Mean   :  0.92949   Mean   :  0.9564  
 3rd Qu.:  1.1261   3rd Qu.:  1.1264   3rd Qu.:  1.09865   3rd Qu.:  1.1254  
 Max.   :521.0724   Max.   :201.8175   Max.   :163.86461   Max.   :140.8394  
   GSM300203          GSM300204          GSM300205          GSM300206       
 Min.   :  0.0100   Min.   : 0.04201   Min.   : 0.04504   Min.   : 0.03707  
 1st Qu.:  0.6444   1st Qu.: 0.69534   1st Qu.: 0.65576   1st Qu.: 0.67520  
 Median :  0.8815   Median : 0.90327   Median : 0.89266   Median : 0.89029  
 Mean   :  0.9393   Mean   : 0.90615   Mean   : 0.93143   Mean   : 0.89243  
 3rd Qu.:  1.0947   3rd Qu.: 1.05673   3rd Qu.: 1.09594   3rd Qu.: 1.04216  
 Max.   :171.9267   Max.   :11.86873   Max.   :41.92615   Max.   :44.31767  
   GSM300207          GSM300208         GSM300209         GSM300210      
 Min.   : 0.09496   Min.   : 0.0339   Min.   : 0.0378   Min.   : 0.0439  
 1st Qu.: 0.67073   1st Qu.: 0.6254   1st Qu.: 0.6646   1st Qu.: 0.6878  
 Median : 0.90025   Median : 0.8622   Median : 0.9000   Median : 0.9280  
 Mean   : 0.94534   Mean   : 0.9803   Mean   : 0.9123   Mean   : 0.9147  
 3rd Qu.: 1.12179   3rd Qu.: 1.0859   3rd Qu.: 1.0942   3rd Qu.: 1.0794  
 Max.   :29.24678   Max.   :52.9163   Max.   :65.7118   Max.   :10.5618  
   GSM300211          GSM300212          GSM300213           GSM300214       
 Min.   : 0.04076   Min.   :  0.0100   Min.   :  0.01375   Min.   : 0.02087  
 1st Qu.: 0.66863   1st Qu.:  0.6050   1st Qu.:  0.60921   1st Qu.: 0.64353  
 Median : 0.91381   Median :  0.9212   Median :  0.85885   Median : 0.85259  
 Mean   : 0.97523   Mean   :  1.2176   Mean   :  0.97751   Mean   : 0.88436  
 3rd Qu.: 1.15610   3rd Qu.:  1.3335   3rd Qu.:  1.09871   3rd Qu.: 1.03969  
 Max.   :25.20849   Max.   :250.6645   Max.   :203.71402   Max.   :14.91582  
   GSM300215          GSM300216          GSM300217          GSM300218       
 Min.   : 0.02148   Min.   : 0.02095   Min.   :  0.0298   Min.   :  0.0100  
 1st Qu.: 0.59703   1st Qu.: 0.67674   1st Qu.:  0.6785   1st Qu.:  0.6831  
 Median : 0.84453   Median : 0.89393   Median :  0.8926   Median :  0.9121  
 Mean   : 0.91951   Mean   : 0.88336   Mean   :  0.9825   Mean   :  0.9497  
 3rd Qu.: 1.07942   3rd Qu.: 1.04526   3rd Qu.:  1.0784   3rd Qu.:  1.0947  
 Max.   :29.88398   Max.   :16.45487   Max.   :322.8984   Max.   :111.3580  
   GSM300219          GSM300220          GSM300221          GSM300222       
 Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.0100  
 1st Qu.:  0.6428   1st Qu.:  0.6633   1st Qu.:  0.7095   1st Qu.:  0.6876  
 Median :  0.9016   Median :  0.9022   Median :  0.9606   Median :  0.9257  
 Mean   :  0.9919   Mean   :  0.9357   Mean   :  1.0171   Mean   :  0.9561  
 3rd Qu.:  1.1415   3rd Qu.:  1.0901   3rd Qu.:  1.1823   3rd Qu.:  1.1120  
 Max.   :125.5235   Max.   :141.2652   Max.   :103.0937   Max.   :103.6204  
   GSM300223         GSM300224          GSM300225          GSM300226       
 Min.   :  0.010   Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.0100  
 1st Qu.:  0.582   1st Qu.:  0.6703   1st Qu.:  0.6976   1st Qu.:  0.5967  
 Median :  0.847   Median :  0.8944   Median :  0.9273   Median :  0.8473  
 Mean   :  1.000   Mean   :  0.9170   Mean   :  0.9424   Mean   :  0.9371  
 3rd Qu.:  1.138   3rd Qu.:  1.0767   3rd Qu.:  1.0917   3rd Qu.:  1.0801  
 Max.   :118.937   Max.   :132.6392   Max.   :113.8672   Max.   :149.0119  
   GSM300227          GSM300228           GSM300229          GSM300230       
 Min.   :  0.0100   Min.   :  0.01864   Min.   :  0.0100   Min.   :  0.0100  
 1st Qu.:  0.3218   1st Qu.:  0.59977   1st Qu.:  0.6461   1st Qu.:  0.6812  
 Median :  0.7507   Median :  0.84264   Median :  0.9027   Median :  0.8965  
 Mean   :  0.9563   Mean   :  0.95215   Mean   :  0.9725   Mean   :  0.9218  
 3rd Qu.:  1.2218   3rd Qu.:  1.12991   3rd Qu.:  1.1386   3rd Qu.:  1.0598  
 Max.   :224.7027   Max.   :149.91292   Max.   :229.9002   Max.   :105.3936  
   GSM300231          GSM300232          GSM300233         GSM300234      
 Min.   :  0.0100   Min.   :  0.0100   Min.   : 0.0100   Min.   :0.08865  
 1st Qu.:  0.6239   1st Qu.:  0.6354   1st Qu.: 0.6944   1st Qu.:0.69713  
 Median :  0.8769   Median :  0.8624   Median : 0.9281   Median :0.88687  
 Mean   :  0.9230   Mean   :  0.9112   Mean   : 0.9492   Mean   :0.88399  
 3rd Qu.:  1.0732   3rd Qu.:  1.0637   3rd Qu.: 1.1011   3rd Qu.:1.03223  
 Max.   :128.2421   Max.   :170.9801   Max.   :92.7360   Max.   :8.77896  
   GSM300235          GSM300236          GSM300237          GSM300238       
 Min.   : 0.02057   Min.   : 0.04699   Min.   : 0.04454   Min.   :  0.0100  
 1st Qu.: 0.63230   1st Qu.: 0.68478   1st Qu.: 0.71399   1st Qu.:  0.6510  
 Median : 0.87893   Median : 0.90417   Median : 0.93046   Median :  0.8794  
 Mean   : 0.92925   Mean   : 0.90269   Mean   : 0.93182   Mean   :  0.9447  
 3rd Qu.: 1.06560   3rd Qu.: 1.05101   3rd Qu.: 1.08743   3rd Qu.:  1.0838  
 Max.   :38.21748   Max.   :11.90255   Max.   :21.36143   Max.   :117.4163  
   GSM300239          GSM300240          GSM300241          GSM300242       
 Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.0100  
 1st Qu.:  0.6024   1st Qu.:  0.6434   1st Qu.:  0.6838   1st Qu.:  0.6874  
 Median :  0.8384   Median :  0.8842   Median :  0.9244   Median :  0.9148  
 Mean   :  0.9717   Mean   :  0.9403   Mean   :  0.9776   Mean   :  0.9552  
 3rd Qu.:  1.0506   3rd Qu.:  1.0693   3rd Qu.:  1.1303   3rd Qu.:  1.1071  
 Max.   :161.4674   Max.   :117.7097   Max.   :102.1145   Max.   :124.5769  
   GSM300243          GSM300244          GSM300245           GSM300246      
 Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.01189   Min.   :0.03533  
 1st Qu.:  0.5409   1st Qu.:  0.6313   1st Qu.:  0.67620   1st Qu.:0.64580  
 Median :  0.8148   Median :  0.8771   Median :  0.91279   Median :0.88000  
 Mean   :  1.0880   Mean   :  0.9351   Mean   :  1.02559   Mean   :0.92713  
 3rd Qu.:  1.2108   3rd Qu.:  1.0872   3rd Qu.:  1.10722   3rd Qu.:1.09020  
 Max.   :172.2145   Max.   :103.1746   Max.   :290.61795   Max.   :8.73023  
   GSM300247           GSM300248           GSM300249          GSM300250       
 Min.   :  0.04447   Min.   :   0.0100   Min.   :  0.0100   Min.   : 0.01699  
 1st Qu.:  0.69971   1st Qu.:   0.6038   1st Qu.:  0.6140   1st Qu.: 0.60785  
 Median :  0.92495   Median :   0.8751   Median :  0.8884   Median : 0.87486  
 Mean   :  0.95636   Mean   :   1.1166   Mean   :  0.9656   Mean   : 0.97831  
 3rd Qu.:  1.09870   3rd Qu.:   1.1931   3rd Qu.:  1.1319   3rd Qu.: 1.18884  
 Max.   :107.12526   Max.   :1054.4650   Max.   :256.1622   Max.   :14.95980  
   GSM300251          GSM300252          GSM300253          GSM300254       
 Min.   :  0.0100   Min.   : 0.01161   Min.   : 0.03465   Min.   : 0.04621  
 1st Qu.:  0.6116   1st Qu.: 0.58647   1st Qu.: 0.65324   1st Qu.: 0.68563  
 Median :  0.8821   Median : 0.85915   Median : 0.89737   Median : 0.95210  
 Mean   :  0.9927   Mean   : 0.99749   Mean   : 0.94833   Mean   : 1.02953  
 3rd Qu.:  1.1608   3rd Qu.: 1.13201   3rd Qu.: 1.11151   3rd Qu.: 1.20104  
 Max.   :168.5180   Max.   :62.38041   Max.   :16.46131   Max.   :17.73828  
   GSM300255          GSM300256          GSM300258          GSM300259       
 Min.   :  0.0100   Min.   :  0.0199   Min.   : 0.03505   Min.   : 0.03505  
 1st Qu.:  0.5681   1st Qu.:  0.5921   1st Qu.: 0.67815   1st Qu.: 0.64623  
 Median :  0.8272   Median :  0.8527   Median : 0.94947   Median : 0.91167  
 Mean   :  1.0118   Mean   :  0.9491   Mean   : 1.01388   Mean   : 0.99579  
 3rd Qu.:  1.2145   3rd Qu.:  1.1324   3rd Qu.: 1.20595   3rd Qu.: 1.15778  
 Max.   :124.2457   Max.   :117.1530   Max.   :16.55426   Max.   :23.04740  
   GSM300260          GSM300261          GSM300262          GSM300263       
 Min.   : 0.02695   Min.   :  0.0177   Min.   : 0.03516   Min.   : 0.03613  
 1st Qu.: 0.69568   1st Qu.:  0.6140   1st Qu.: 0.62587   1st Qu.: 0.67446  
 Median : 0.95686   Median :  0.8580   Median : 0.85730   Median : 0.91205  
 Mean   : 1.01674   Mean   :  0.9309   Mean   : 0.95006   Mean   : 0.90884  
 3rd Qu.: 1.19858   3rd Qu.:  1.0825   3rd Qu.: 1.12608   3rd Qu.: 1.06946  
 Max.   :12.03024   Max.   :128.2865   Max.   :75.89957   Max.   :13.53799  
   GSM300264          GSM300265          GSM300266         GSM300267       
 Min.   : 0.04537   Min.   : 0.01841   Min.   :0.02369   Min.   : 0.02345  
 1st Qu.: 0.72198   1st Qu.: 0.64875   1st Qu.:0.69646   1st Qu.: 0.59089  
 Median : 0.94350   Median : 0.88743   Median :0.93410   Median : 0.83558  
 Mean   : 0.94155   Mean   : 0.91975   Mean   :0.97869   Mean   : 0.92024  
 3rd Qu.: 1.10764   3rd Qu.: 1.07594   3rd Qu.:1.13732   3rd Qu.: 1.06608  
 Max.   :10.16412   Max.   :17.47691   Max.   :9.70494   Max.   :33.64462  
   GSM300268          GSM300269          GSM300270          GSM300271       
 Min.   : 0.01385   Min.   : 0.01721   Min.   : 0.02787   Min.   :  0.0100  
 1st Qu.: 0.65043   1st Qu.: 0.68075   1st Qu.: 0.72431   1st Qu.:  0.6896  
 Median : 0.88457   Median : 0.91825   Median : 0.95277   Median :  0.9317  
 Mean   : 0.90011   Mean   : 0.89812   Mean   : 0.94187   Mean   :  0.9666  
 3rd Qu.: 1.05509   3rd Qu.: 1.07451   3rd Qu.: 1.11231   3rd Qu.:  1.1104  
 Max.   :11.32207   Max.   :24.80721   Max.   :20.54625   Max.   :125.1418  
   GSM300272          GSM300273          GSM300274           GSM300275       
 Min.   :  0.0118   Min.   :  0.0100   Min.   :  0.01288   Min.   : 0.04162  
 1st Qu.:  0.5784   1st Qu.:  0.6178   1st Qu.:  0.71696   1st Qu.: 0.68152  
 Median :  0.8374   Median :  0.8693   Median :  0.95407   Median : 0.92559  
 Mean   :  0.9731   Mean   :  0.9143   Mean   :  0.98065   Mean   : 0.97090  
 3rd Qu.:  1.0923   3rd Qu.:  1.0924   3rd Qu.:  1.14896   3rd Qu.: 1.12422  
 Max.   :929.0001   Max.   :181.5166   Max.   :100.33138   Max.   :43.72514  
   GSM300276          GSM300277          GSM300278          GSM300279       
 Min.   :  0.0100   Min.   : 0.02591   Min.   : 0.02448   Min.   : 0.02353  
 1st Qu.:  0.5797   1st Qu.: 0.62983   1st Qu.: 0.69030   1st Qu.: 0.67695  
 Median :  0.8547   Median : 0.90902   Median : 0.97570   Median : 0.91166  
 Mean   :  1.0294   Mean   : 0.99638   Mean   : 1.08382   Mean   : 0.94342  
 3rd Qu.:  1.1676   3rd Qu.: 1.15420   3rd Qu.: 1.25245   3rd Qu.: 1.10394  
 Max.   :316.0058   Max.   :14.37378   Max.   :48.10394   Max.   :22.47344  
   GSM300280          GSM300281          GSM300282          GSM300283       
 Min.   : 0.03567   Min.   : 0.03646   Min.   : 0.01794   Min.   :  0.0100  
 1st Qu.: 0.63620   1st Qu.: 0.64927   1st Qu.: 0.69415   1st Qu.:  0.6197  
 Median : 0.89410   Median : 0.90538   Median : 0.95422   Median :  0.8724  
 Mean   : 0.94472   Mean   : 0.94032   Mean   : 1.00218   Mean   :  0.9557  
 3rd Qu.: 1.10887   3rd Qu.: 1.11570   3rd Qu.: 1.19959   3rd Qu.:  1.1029  
 Max.   :44.08281   Max.   :19.00818   Max.   :29.01290   Max.   :248.8288  
   GSM300284          GSM300285          GSM300286          GSM300287       
 Min.   :  0.0100   Min.   : 0.05183   Min.   : 0.02798   Min.   :  0.0100  
 1st Qu.:  0.6298   1st Qu.: 0.65461   1st Qu.: 0.62962   1st Qu.:  0.6519  
 Median :  0.8870   Median : 0.87610   Median : 0.87419   Median :  0.8648  
 Mean   :  0.9849   Mean   : 0.90370   Mean   : 0.90177   Mean   :  1.0189  
 3rd Qu.:  1.1372   3rd Qu.: 1.06947   3rd Qu.: 1.05481   3rd Qu.:  1.1840  
 Max.   :293.9077   Max.   :17.20291   Max.   :46.13750   Max.   :181.9658  
   GSM300288          GSM300289         GSM300290          GSM300291       
 Min.   : 0.03446   Min.   : 0.0100   Min.   :  0.0100   Min.   :  0.0100  
 1st Qu.: 0.66115   1st Qu.: 0.6854   1st Qu.:  0.6199   1st Qu.:  0.6395  
 Median : 0.88253   Median : 0.9391   Median :  0.8759   Median :  0.9062  
 Mean   : 0.94080   Mean   : 0.9974   Mean   :  1.0114   Mean   :  0.9946  
 3rd Qu.: 1.08031   3rd Qu.: 1.1561   3rd Qu.:  1.1474   3rd Qu.:  1.1657  
 Max.   :18.22758   Max.   :96.9125   Max.   :171.5643   Max.   :121.9132  
   GSM300292          GSM300293         GSM300294          GSM300295       
 Min.   :  0.0100   Min.   : 0.0100   Min.   :  0.0100   Min.   :  0.0100  
 1st Qu.:  0.6830   1st Qu.: 0.6621   1st Qu.:  0.6562   1st Qu.:  0.6438  
 Median :  0.9498   Median : 0.8968   Median :  0.9104   Median :  0.9012  
 Mean   :  1.0339   Mean   : 0.9474   Mean   :  0.9581   Mean   :  0.9601  
 3rd Qu.:  1.1986   3rd Qu.: 1.1050   3rd Qu.:  1.1218   3rd Qu.:  1.1191  
 Max.   :105.3159   Max.   :77.3347   Max.   :116.0118   Max.   :122.7439  
   GSM300296         GSM300297         GSM300298          GSM300299      
 Min.   : 0.0100   Min.   : 0.0100   Min.   :  0.0100   Min.   : 0.0100  
 1st Qu.: 0.6958   1st Qu.: 0.6836   1st Qu.:  0.6529   1st Qu.: 0.6659  
 Median : 0.9537   Median : 0.9437   Median :  0.9283   Median : 0.9054  
 Mean   : 1.0163   Mean   : 1.0164   Mean   :  0.9520   Mean   : 1.0125  
 3rd Qu.: 1.1845   3rd Qu.: 1.1783   3rd Qu.:  1.1302   3rd Qu.: 1.1707  
 Max.   :81.5376   Max.   :82.2354   Max.   :106.6485   Max.   :86.2981  
   GSM300300         GSM300301           GSM300302          GSM300303       
 Min.   : 0.0100   Min.   :   0.0167   Min.   : 0.02443   Min.   : 0.04432  
 1st Qu.: 0.6293   1st Qu.:   0.5753   1st Qu.: 0.61690   1st Qu.: 0.71413  
 Median : 0.8481   Median :   0.8250   Median : 0.87990   Median : 0.94998  
 Mean   : 0.9043   Mean   :   1.2715   Mean   : 0.92145   Mean   : 0.97311  
 3rd Qu.: 1.0472   3rd Qu.:   1.0612   3rd Qu.: 1.10378   3rd Qu.: 1.14614  
 Max.   :97.2785   Max.   :1438.7626   Max.   :83.08115   Max.   :16.41317  
   GSM300304         GSM300305          GSM300306           GSM300307       
 Min.   : 0.0197   Min.   :  0.0186   Min.   :  0.03083   Min.   : 0.04091  
 1st Qu.: 0.6740   1st Qu.:  0.5957   1st Qu.:  0.64115   1st Qu.: 0.68402  
 Median : 0.9217   Median :  0.8666   Median :  0.89985   Median : 0.91782  
 Mean   : 0.9775   Mean   :  1.0242   Mean   :  0.95530   Mean   : 0.98517  
 3rd Qu.: 1.1316   3rd Qu.:  1.1451   3rd Qu.:  1.13500   3rd Qu.: 1.15760  
 Max.   :25.6341   Max.   :570.2180   Max.   :103.79764   Max.   :16.18648  
   GSM300308          GSM300309          GSM300310         GSM300311       
 Min.   : 0.02225   Min.   : 0.01903   Min.   :0.03432   Min.   : 0.04164  
 1st Qu.: 0.63377   1st Qu.: 0.63188   1st Qu.:0.65531   1st Qu.: 0.68847  
 Median : 0.85786   Median : 0.88736   Median :0.90542   Median : 0.92338  
 Mean   : 0.92050   Mean   : 0.94824   Mean   :0.92688   Mean   : 0.98592  
 3rd Qu.: 1.04866   3rd Qu.: 1.09681   3rd Qu.:1.09336   3rd Qu.: 1.13954  
 Max.   :29.43575   Max.   :73.26019   Max.   :9.10163   Max.   :14.85199  
   GSM300312          GSM300313          GSM300314          GSM300315       
 Min.   : 0.02503   Min.   : 0.01903   Min.   : 0.02139   Min.   : 0.01602  
 1st Qu.: 0.68584   1st Qu.: 0.63704   1st Qu.: 0.64942   1st Qu.: 0.68730  
 Median : 0.91091   Median : 0.88976   Median : 0.90704   Median : 0.94324  
 Mean   : 0.92650   Mean   : 0.94362   Mean   : 0.94873   Mean   : 0.99373  
 3rd Qu.: 1.08553   3rd Qu.: 1.07836   3rd Qu.: 1.11316   3rd Qu.: 1.17411  
 Max.   :23.76547   Max.   :77.30228   Max.   :14.77693   Max.   :13.94632  
   GSM300316         GSM300317          GSM300318          GSM300319       
 Min.   : 0.0443   Min.   : 0.01724   Min.   : 0.02574   Min.   : 0.04441  
 1st Qu.: 0.6878   1st Qu.: 0.60077   1st Qu.: 0.63747   1st Qu.: 0.70189  
 Median : 0.9203   Median : 0.85598   Median : 0.89916   Median : 0.94542  
 Mean   : 0.9524   Mean   : 0.93867   Mean   : 0.98044   Mean   : 0.99355  
 3rd Qu.: 1.1167   3rd Qu.: 1.09901   3rd Qu.: 1.14161   3rd Qu.: 1.15856  
 Max.   :22.6159   Max.   :87.92602   Max.   :62.09858   Max.   :33.07560  
   GSM300320         GSM300321          GSM300322          GSM300323       
 Min.   : 0.0100   Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.0100  
 1st Qu.: 0.6905   1st Qu.:  0.6413   1st Qu.:  0.6466   1st Qu.:  0.7086  
 Median : 0.9225   Median :  0.8851   Median :  0.9098   Median :  0.9551  
 Mean   : 0.9637   Mean   :  0.9382   Mean   :  0.9600   Mean   :  0.9940  
 3rd Qu.: 1.1078   3rd Qu.:  1.0807   3rd Qu.:  1.1348   3rd Qu.:  1.1466  
 Max.   :83.2493   Max.   :117.1360   Max.   :105.8359   Max.   :147.5523  
   GSM300324        GSM300325          GSM300326          GSM300327       
 Min.   :0.0378   Min.   : 0.02492   Min.   : 0.07938   Min.   : 0.07164  
 1st Qu.:0.6605   1st Qu.: 0.59030   1st Qu.: 0.57543   1st Qu.: 0.68960  
 Median :0.8698   Median : 0.83026   Median : 0.80885   Median : 0.91148  
 Mean   :0.8959   Mean   : 0.89040   Mean   : 0.93886   Mean   : 0.93016  
 3rd Qu.:1.0431   3rd Qu.: 1.06096   3rd Qu.: 1.08607   3rd Qu.: 1.09133  
 Max.   :9.6998   Max.   :17.02787   Max.   :41.96609   Max.   :17.30909  
   GSM300328          GSM300329          GSM300330          GSM300331       
 Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.0100  
 1st Qu.:  0.6705   1st Qu.:  0.6051   1st Qu.:  0.6585   1st Qu.:  0.6914  
 Median :  0.8796   Median :  0.8309   Median :  0.8885   Median :  0.9198  
 Mean   :  0.9071   Mean   :  0.9097   Mean   :  0.8990   Mean   :  0.9417  
 3rd Qu.:  1.0345   3rd Qu.:  1.0267   3rd Qu.:  1.0517   3rd Qu.:  1.1145  
 Max.   :104.6215   Max.   :108.7841   Max.   :132.5643   Max.   :105.0428  
   GSM300332          GSM300333          GSM300335           GSM300338       
 Min.   : 0.04042   Min.   : 0.01795   Min.   :  0.04111   Min.   :  0.0100  
 1st Qu.: 0.63358   1st Qu.: 0.59186   1st Qu.:  0.69447   1st Qu.:  0.6862  
 Median : 0.86349   Median : 0.82857   Median :  0.93652   Median :  0.9050  
 Mean   : 0.90280   Mean   : 0.89598   Mean   :  0.95166   Mean   :  0.9268  
 3rd Qu.: 1.06843   3rd Qu.: 1.04435   3rd Qu.:  1.13255   3rd Qu.:  1.0713  
 Max.   :12.46614   Max.   :38.78039   Max.   :104.19598   Max.   :104.0315  
   GSM300339         GSM300340          GSM300341          GSM318840       
 Min.   : 0.0100   Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.0100  
 1st Qu.: 0.6215   1st Qu.:  0.6653   1st Qu.:  0.7029   1st Qu.:  0.6109  
 Median : 0.8574   Median :  0.9063   Median :  0.9268   Median :  0.8345  
 Mean   : 0.9266   Mean   :  0.9320   Mean   :  0.9565   Mean   :  0.9693  
 3rd Qu.: 1.0550   3rd Qu.:  1.0833   3rd Qu.:  1.1072   3rd Qu.:  1.0615  
 Max.   :93.1254   Max.   :161.6437   Max.   :106.4059   Max.   :196.9218  
   GSM350078          GSM1176196         GSM1176197         GSM1176198      
 Min.   : 0.04448   Min.   :  0.0100   Min.   : 0.02004   Min.   :  0.0180  
 1st Qu.: 0.65940   1st Qu.:  0.5974   1st Qu.: 0.68871   1st Qu.:  0.6154  
 Median : 0.92149   Median :  0.8561   Median : 0.90076   Median :  0.8545  
 Mean   : 0.97956   Mean   :  1.0601   Mean   : 0.92212   Mean   :  0.9761  
 3rd Qu.: 1.18199   3rd Qu.:  1.1689   3rd Qu.: 1.07213   3rd Qu.:  1.0995  
 Max.   :74.13313   Max.   :664.9006   Max.   :46.03111   Max.   :541.5968  
   GSM1176199         GSM1176200         GSM1176201         GSM1176202      
 Min.   : 0.05353   Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.0100  
 1st Qu.: 0.61809   1st Qu.:  0.6050   1st Qu.:  0.6652   1st Qu.:  0.6256  
 Median : 0.83626   Median :  0.8410   Median :  0.8899   Median :  0.8658  
 Mean   : 0.91422   Mean   :  0.9682   Mean   :  0.9661   Mean   :  0.9945  
 3rd Qu.: 1.05383   3rd Qu.:  1.0957   3rd Qu.:  1.0732   3rd Qu.:  1.1016  
 Max.   :53.84378   Max.   :547.5518   Max.   :412.6565   Max.   :524.7122  
   GSM1176203         GSM1176204        GSM1176205         GSM1176206      
 Min.   : 0.05854   Min.   : 0.0230   Min.   : 0.08918   Min.   : 0.02613  
 1st Qu.: 0.67265   1st Qu.: 0.6284   1st Qu.: 0.70352   1st Qu.: 0.63642  
 Median : 0.86081   Median : 0.8497   Median : 0.90761   Median : 0.85817  
 Mean   : 0.91081   Mean   : 0.9483   Mean   : 0.92776   Mean   : 0.92514  
 3rd Qu.: 1.04846   3rd Qu.: 1.0628   3rd Qu.: 1.08535   3rd Qu.: 1.07670  
 Max.   :28.86213   Max.   :43.2873   Max.   :27.00389   Max.   :38.62118  
   GSM1176207         GSM1176208         GSM1176209         GSM1176210      
 Min.   : 0.07596   Min.   :  0.0100   Min.   : 0.01621   Min.   :  0.0100  
 1st Qu.: 0.67690   1st Qu.:  0.6549   1st Qu.: 0.64651   1st Qu.:  0.6410  
 Median : 0.87203   Median :  0.8835   Median : 0.90417   Median :  0.8715  
 Mean   : 0.89949   Mean   :  0.9876   Mean   : 0.99287   Mean   :  0.9651  
 3rd Qu.: 1.05520   3rd Qu.:  1.0882   3rd Qu.: 1.14876   3rd Qu.:  1.1091  
 Max.   :30.61647   Max.   :235.4788   Max.   :92.10214   Max.   :465.8624  
   GSM1176211         GSM1176212         GSM1176213         GSM1176214      
 Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.0100  
 1st Qu.:  0.6174   1st Qu.:  0.5791   1st Qu.:  0.5703   1st Qu.:  0.5521  
 Median :  0.8814   Median :  0.8444   Median :  0.8294   Median :  0.8209  
 Mean   :  0.9699   Mean   :  1.0281   Mean   :  1.0117   Mean   :  0.9307  
 3rd Qu.:  1.1114   3rd Qu.:  1.1154   3rd Qu.:  1.1374   3rd Qu.:  1.0749  
 Max.   :491.3960   Max.   :367.1830   Max.   :425.6297   Max.   :536.9973  
   GSM1176215         GSM1176216          GSM1176217         GSM1176218      
 Min.   : 0.02175   Min.   :  0.01138   Min.   :  0.0100   Min.   :  0.0106  
 1st Qu.: 0.66771   1st Qu.:  0.52326   1st Qu.:  0.5909   1st Qu.:  0.5686  
 Median : 0.89560   Median :  0.82678   Median :  0.8443   Median :  0.8340  
 Mean   : 0.91389   Mean   :  1.07050   Mean   :  0.9734   Mean   :  1.0788  
 3rd Qu.: 1.08266   3rd Qu.:  1.18225   3rd Qu.:  1.1023   3rd Qu.:  1.1458  
 Max.   :43.57256   Max.   :141.36652   Max.   :427.3170   Max.   :428.7741  
   GSM1176219         GSM1176220         GSM1176221         GSM1176222      
 Min.   :  0.0100   Min.   : 0.03657   Min.   : 0.01947   Min.   : 0.08076  
 1st Qu.:  0.5765   1st Qu.: 0.58822   1st Qu.: 0.57797   1st Qu.: 0.69001  
 Median :  0.8377   Median : 0.83778   Median : 0.83211   Median : 0.91176  
 Mean   :  1.0282   Mean   : 0.96266   Mean   : 0.95395   Mean   : 0.89629  
 3rd Qu.:  1.1492   3rd Qu.: 1.10130   3rd Qu.: 1.10391   3rd Qu.: 1.05484  
 Max.   :464.3944   Max.   :62.36919   Max.   :38.75187   Max.   :35.83370  
   GSM1176223         GSM1176224        GSM1176225         GSM1176226     
 Min.   :  0.0100   Min.   : 0.0178   Min.   : 0.08122   Min.   : 0.0232  
 1st Qu.:  0.5475   1st Qu.: 0.5820   1st Qu.: 0.67773   1st Qu.: 0.6133  
 Median :  0.8411   Median : 0.8590   Median : 0.90146   Median : 0.8383  
 Mean   :  0.9909   Mean   : 1.0658   Mean   : 0.91054   Mean   : 0.9141  
 3rd Qu.:  1.1407   3rd Qu.: 1.1839   3rd Qu.: 1.05629   3rd Qu.: 1.0543  
 Max.   :467.3872   Max.   :80.6223   Max.   :34.34052   Max.   :43.8862  
   GSM1176227         GSM1176228         GSM1176229          GSM1176230      
 Min.   :  0.0175   Min.   :  0.0100   Min.   :  0.01943   Min.   :  0.0100  
 1st Qu.:  0.5896   1st Qu.:  0.5312   1st Qu.:  0.63213   1st Qu.:  0.5994  
 Median :  0.8262   Median :  0.8244   Median :  0.86280   Median :  0.8972  
 Mean   :  0.9789   Mean   :  0.9322   Mean   :  0.93610   Mean   :  1.2724  
 3rd Qu.:  1.0739   3rd Qu.:  1.1103   3rd Qu.:  1.04129   3rd Qu.:  1.1976  
 Max.   :321.9957   Max.   :576.0618   Max.   :127.14636   Max.   :800.6404  
   GSM1176231         GSM1176232         GSM1176233         GSM1176234      
 Min.   :  0.0100   Min.   :  0.0100   Min.   : 0.01526   Min.   :  0.0100  
 1st Qu.:  0.6000   1st Qu.:  0.6643   1st Qu.: 0.69391   1st Qu.:  0.6492  
 Median :  0.8532   Median :  0.8708   Median : 0.92194   Median :  0.8830  
 Mean   :  0.9818   Mean   :  0.9309   Mean   : 0.92324   Mean   :  0.9529  
 3rd Qu.:  1.0878   3rd Qu.:  1.0400   3rd Qu.: 1.09457   3rd Qu.:  1.1099  
 Max.   :505.3052   Max.   :577.5082   Max.   :34.28196   Max.   :489.9823  
   GSM1176235         GSM1176236          GSM1176237         GSM1176238      
 Min.   : 0.01317   Min.   :  0.08173   Min.   :  0.0100   Min.   :  0.0100  
 1st Qu.: 0.52214   1st Qu.:  0.66392   1st Qu.:  0.6034   1st Qu.:  0.6131  
 Median : 0.78793   Median :  0.90919   Median :  0.8398   Median :  0.8444  
 Mean   : 1.03074   Mean   :  0.92754   Mean   :  0.9563   Mean   :  0.9740  
 3rd Qu.: 1.21507   3rd Qu.:  1.09633   3rd Qu.:  1.0821   3rd Qu.:  1.0646  
 Max.   :60.61777   Max.   :148.88857   Max.   :601.8692   Max.   :619.7251  
   GSM1176239         GSM1176240         GSM1176241         GSM1176242      
 Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.0100   Min.   : 0.01415  
 1st Qu.:  0.6668   1st Qu.:  0.6260   1st Qu.:  0.6748   1st Qu.: 0.67110  
 Median :  0.9096   Median :  0.8752   Median :  0.8886   Median : 0.89231  
 Mean   :  0.9580   Mean   :  0.9199   Mean   :  0.9207   Mean   : 0.91919  
 3rd Qu.:  1.1151   3rd Qu.:  1.0751   3rd Qu.:  1.0398   3rd Qu.: 1.06624  
 Max.   :410.8429   Max.   :465.1847   Max.   :490.5593   Max.   :47.74039  
   GSM1176243         GSM1176244         GSM1176245         GSM1176246     
 Min.   :  0.0100   Min.   :  0.0100   Min.   : 0.05118   Min.   : 0.0827  
 1st Qu.:  0.6474   1st Qu.:  0.5817   1st Qu.: 0.64099   1st Qu.: 0.6777  
 Median :  0.8699   Median :  0.8283   Median : 0.87372   Median : 0.8961  
 Mean   :  0.9376   Mean   :  0.9374   Mean   : 0.93239   Mean   : 0.9215  
 3rd Qu.:  1.0538   3rd Qu.:  1.0763   3rd Qu.: 1.09972   3rd Qu.: 1.0771  
 Max.   :474.5391   Max.   :731.5267   Max.   :21.98091   Max.   :29.4242  
   GSM1176247         GSM1176248         GSM1176249         GSM1176250      
 Min.   : 0.03961   Min.   : 0.08826   Min.   : 0.06883   Min.   : 0.05613  
 1st Qu.: 0.62443   1st Qu.: 0.68255   1st Qu.: 0.63952   1st Qu.: 0.67192  
 Median : 0.86233   Median : 0.90719   Median : 0.85815   Median : 0.89407  
 Mean   : 0.90204   Mean   : 0.97282   Mean   : 0.91778   Mean   : 0.90563  
 3rd Qu.: 1.06715   3rd Qu.: 1.10518   3rd Qu.: 1.06277   3rd Qu.: 1.06076  
 Max.   :27.49851   Max.   :32.64111   Max.   :34.07566   Max.   :31.55135  
   GSM1176251        GSM1176252         GSM1176253         GSM1176254      
 Min.   : 0.0382   Min.   : 0.02092   Min.   : 0.04858   Min.   : 0.06698  
 1st Qu.: 0.6623   1st Qu.: 0.63568   1st Qu.: 0.67792   1st Qu.: 0.66690  
 Median : 0.9045   Median : 0.85876   Median : 0.92143   Median : 0.88924  
 Mean   : 0.9284   Mean   : 0.91941   Mean   : 0.92160   Mean   : 0.91495  
 3rd Qu.: 1.0940   3rd Qu.: 1.07057   3rd Qu.: 1.09060   3rd Qu.: 1.05304  
 Max.   :52.7081   Max.   :26.39957   Max.   :97.65575   Max.   :21.05555  
   GSM1176255         GSM1176256         GSM1176257         GSM1176258      
 Min.   :  0.0100   Min.   :  0.0100   Min.   : 0.05799   Min.   :  0.0100  
 1st Qu.:  0.6242   1st Qu.:  0.7029   1st Qu.: 0.66182   1st Qu.:  0.7049  
 Median :  0.8619   Median :  0.9027   Median : 0.93037   Median :  0.9079  
 Mean   :  1.0431   Mean   :  0.9504   Mean   : 0.98504   Mean   :  0.9640  
 3rd Qu.:  1.1170   3rd Qu.:  1.0679   3rd Qu.: 1.19471   3rd Qu.:  1.0966  
 Max.   :342.8059   Max.   :469.4184   Max.   :77.15756   Max.   :502.1976  
   GSM1176259         GSM1176260         GSM1176261         GSM1176262      
 Min.   :  0.0907   Min.   :  0.0100   Min.   :  0.0151   Min.   :  0.0100  
 1st Qu.:  0.6880   1st Qu.:  0.6243   1st Qu.:  0.6207   1st Qu.:  0.6606  
 Median :  0.9299   Median :  0.8884   Median :  0.8757   Median :  0.8662  
 Mean   :  0.9608   Mean   :  1.0431   Mean   :  0.9874   Mean   :  0.9862  
 3rd Qu.:  1.1326   3rd Qu.:  1.1829   3rd Qu.:  1.1572   3rd Qu.:  1.0654  
 Max.   :189.8559   Max.   :572.8378   Max.   :427.8145   Max.   :436.9121  
   GSM1176263         GSM1176264         GSM1176265         GSM1176266      
 Min.   :  0.0100   Min.   :  0.0100   Min.   :  0.0100   Min.   : 0.01593  
 1st Qu.:  0.6964   1st Qu.:  0.6872   1st Qu.:  0.6489   1st Qu.: 0.70854  
 Median :  0.9382   Median :  0.9233   Median :  0.8706   Median : 0.91307  
 Mean   :  0.9856   Mean   :  0.9657   Mean   :  0.9921   Mean   : 0.93861  
 3rd Qu.:  1.1521   3rd Qu.:  1.1154   3rd Qu.:  1.1182   3rd Qu.: 1.08195  
 Max.   :444.6575   Max.   :316.5403   Max.   :514.1650   Max.   :42.10733  
   GSM1176267         GSM1176268         GSM1176269        GSM1176270      
 Min.   :  0.0100   Min.   :  0.0100   Min.   : 0.0894   Min.   : 0.07345  
 1st Qu.:  0.6591   1st Qu.:  0.6314   1st Qu.: 0.6385   1st Qu.: 0.65348  
 Median :  0.8756   Median :  0.8765   Median : 0.8596   Median : 0.88361  
 Mean   :  0.9628   Mean   :  0.9615   Mean   : 0.9588   Mean   : 0.99911  
 3rd Qu.:  1.0977   3rd Qu.:  1.1300   3rd Qu.: 1.1177   3rd Qu.: 1.13868  
 Max.   :450.3197   Max.   :503.5956   Max.   :33.0079   Max.   :60.47261  
   GSM1176271         GSM1176272         GSM1176273         GSM1176274       
 Min.   : 0.01561   Min.   : 0.06572   Min.   : 0.01604   Min.   :  0.08801  
 1st Qu.: 0.70836   1st Qu.: 0.65774   1st Qu.: 0.70002   1st Qu.:  0.66820  
 Median : 0.90954   Median : 0.88288   Median : 0.90282   Median :  0.90962  
 Mean   : 0.93953   Mean   : 0.99103   Mean   : 0.92900   Mean   :  0.97400  
 3rd Qu.: 1.08779   3rd Qu.: 1.15003   3rd Qu.: 1.08691   3rd Qu.:  1.13418  
 Max.   :22.64926   Max.   :79.40704   Max.   :38.79817   Max.   :136.52701  
   GSM1176275      
 Min.   : 0.09829  
 1st Qu.: 0.71525  
 Median : 0.92243  
 Mean   : 0.94740  
 3rd Qu.: 1.09188  
 Max.   :21.67946  
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">exprs</span><span class="p">(</span><span class="n">gse</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">log2</span><span class="p">(</span><span class="nf">exprs</span><span class="p">(</span><span class="n">gse</span><span class="p">))</span>
<span class="nf">boxplot</span><span class="p">(</span><span class="nf">exprs</span><span class="p">(</span><span class="n">gse</span><span class="p">),</span><span class="n">outline</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/Module1-Data_Processing_15_0.png" src="_images/Module1-Data_Processing_15_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># log2 transformation</span>
<span class="n">ex</span> <span class="o">&lt;-</span> <span class="nf">exprs</span><span class="p">(</span><span class="n">gset</span><span class="p">)</span>
<span class="n">qx</span> <span class="o">&lt;-</span> <span class="nf">as.numeric</span><span class="p">(</span><span class="nf">quantile</span><span class="p">(</span><span class="n">ex</span><span class="p">,</span> <span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="n">.</span><span class="p">,</span> <span class="m">0.25</span><span class="p">,</span> <span class="m">0.5</span><span class="p">,</span> <span class="m">0.75</span><span class="p">,</span> <span class="m">0.99</span><span class="p">,</span> <span class="m">1.0</span><span class="p">),</span> <span class="n">na.rm</span><span class="o">=</span><span class="bp">T</span><span class="p">))</span>
<span class="n">LogC</span> <span class="o">&lt;-</span> <span class="p">(</span><span class="n">qx</span><span class="p">[</span><span class="m">5</span><span class="p">]</span> <span class="o">&gt;</span> <span class="m">100</span><span class="p">)</span> <span class="o">||</span>
  <span class="p">(</span><span class="n">qx</span><span class="p">[</span><span class="m">6</span><span class="p">]</span><span class="o">-</span><span class="n">qx</span><span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="m">50</span> <span class="o">&amp;&amp;</span> <span class="n">qx</span><span class="p">[</span><span class="m">2</span><span class="p">]</span> <span class="o">&gt;</span> <span class="m">0</span><span class="p">)</span>
<span class="nf">if </span><span class="p">(</span><span class="n">LogC</span><span class="p">)</span> <span class="p">{</span> <span class="n">ex</span><span class="p">[</span><span class="nf">which</span><span class="p">(</span><span class="n">ex</span> <span class="o">&lt;=</span> <span class="m">0</span><span class="p">)]</span> <span class="o">&lt;-</span> <span class="kc">NaN</span>
<span class="nf">exprs</span><span class="p">(</span><span class="n">gset</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">log2</span><span class="p">(</span><span class="n">ex</span><span class="p">)</span> <span class="p">}</span>

<span class="c1"># assign samples to groups and set up design matrix</span>
<span class="n">gs</span> <span class="o">&lt;-</span> <span class="nf">factor</span><span class="p">(</span><span class="n">sml</span><span class="p">)</span>
<span class="n">groups</span> <span class="o">&lt;-</span> <span class="nf">make.names</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;c&quot;</span><span class="p">,</span><span class="s">&quot;d&quot;</span><span class="p">))</span>
<span class="nf">levels</span><span class="p">(</span><span class="n">gs</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">groups</span>
<span class="n">gset</span><span class="o">$</span><span class="n">group</span> <span class="o">&lt;-</span> <span class="n">gs</span>
<span class="n">design</span> <span class="o">&lt;-</span> <span class="nf">model.matrix</span><span class="p">(</span><span class="o">~</span><span class="n">group</span> <span class="o">+</span> <span class="m">0</span><span class="p">,</span> <span class="n">gset</span><span class="p">)</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">design</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">levels</span><span class="p">(</span><span class="n">gs</span><span class="p">)</span>

<span class="n">fit</span> <span class="o">&lt;-</span> <span class="nf">lmFit</span><span class="p">(</span><span class="n">gset</span><span class="p">,</span> <span class="n">design</span><span class="p">)</span>  <span class="c1"># fit linear model</span>

<span class="c1"># set up contrasts of interest and recalculate model coefficients</span>
<span class="n">cts</span> <span class="o">&lt;-</span> <span class="nf">paste</span><span class="p">(</span><span class="n">groups</span><span class="p">[</span><span class="m">1</span><span class="p">],</span> <span class="n">groups</span><span class="p">[</span><span class="m">2</span><span class="p">],</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;-&quot;</span><span class="p">)</span>
<span class="n">cont.matrix</span> <span class="o">&lt;-</span> <span class="nf">makeContrasts</span><span class="p">(</span><span class="n">contrasts</span><span class="o">=</span><span class="n">cts</span><span class="p">,</span> <span class="n">levels</span><span class="o">=</span><span class="n">design</span><span class="p">)</span>
<span class="n">fit2</span> <span class="o">&lt;-</span> <span class="nf">contrasts.fit</span><span class="p">(</span><span class="n">fit</span><span class="p">,</span> <span class="n">cont.matrix</span><span class="p">)</span>

<span class="c1"># compute statistics and table of top significant genes</span>
<span class="n">fit2</span> <span class="o">&lt;-</span> <span class="nf">eBayes</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span> <span class="m">0.01</span><span class="p">)</span>
<span class="n">tT</span> <span class="o">&lt;-</span> <span class="nf">topTable</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span> <span class="n">adjust</span><span class="o">=</span><span class="s">&quot;fdr&quot;</span><span class="p">,</span> <span class="n">sort.by</span><span class="o">=</span><span class="s">&quot;B&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="m">250</span><span class="p">)</span>

<span class="n">tT</span> <span class="o">&lt;-</span> <span class="nf">subset</span><span class="p">(</span><span class="n">tT</span><span class="p">,</span> <span class="n">select</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ID&quot;</span><span class="p">,</span><span class="s">&quot;adj.P.Val&quot;</span><span class="p">,</span><span class="s">&quot;P.Value&quot;</span><span class="p">,</span><span class="s">&quot;t&quot;</span><span class="p">,</span><span class="s">&quot;B&quot;</span><span class="p">,</span><span class="s">&quot;logFC&quot;</span><span class="p">,</span><span class="s">&quot;Gene.symbol&quot;</span><span class="p">,</span><span class="s">&quot;Gene.title&quot;</span><span class="p">))</span>
<span class="nf">write.table</span><span class="p">(</span><span class="n">tT</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="nf">stdout</span><span class="p">(),</span> <span class="n">row.names</span><span class="o">=</span><span class="bp">F</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">summary</span><span class="p">(</span><span class="nf">exprs</span><span class="p">(</span><span class="n">gse</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize and quality control test results.</span>
<span class="c1"># Build histogram of P-values for all genes. Normal test</span>
<span class="c1"># assumption is that most genes are not differentially expressed.</span>
<span class="n">tT2</span> <span class="o">&lt;-</span> <span class="nf">topTable</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span> <span class="n">adjust</span><span class="o">=</span><span class="s">&quot;fdr&quot;</span><span class="p">,</span> <span class="n">sort.by</span><span class="o">=</span><span class="s">&quot;B&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="kc">Inf</span><span class="p">)</span>
<span class="nf">hist</span><span class="p">(</span><span class="n">tT2</span><span class="o">$</span><span class="n">adj.P.Val</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="s">&quot;grey&quot;</span><span class="p">,</span> <span class="n">border</span> <span class="o">=</span> <span class="s">&quot;white&quot;</span><span class="p">,</span> <span class="n">xlab</span> <span class="o">=</span> <span class="s">&quot;P-adj&quot;</span><span class="p">,</span>
     <span class="n">ylab</span> <span class="o">=</span> <span class="s">&quot;Number of genes&quot;</span><span class="p">,</span> <span class="n">main</span> <span class="o">=</span> <span class="s">&quot;P-adj value distribution&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/Module1-Data_Processing_18_0.png" src="_images/Module1-Data_Processing_18_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># summarize test results as &quot;up&quot;, &quot;down&quot; or &quot;not expressed&quot;</span>
<span class="n">dT</span> <span class="o">&lt;-</span> <span class="nf">decideTests</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span> <span class="n">adjust.method</span><span class="o">=</span><span class="s">&quot;fdr&quot;</span><span class="p">,</span> <span class="n">p.value</span><span class="o">=</span><span class="m">0.05</span><span class="p">)</span>
<span class="c1"># Venn diagram of results</span>
<span class="nf">vennDiagram</span><span class="p">(</span><span class="n">dT</span><span class="p">,</span> <span class="n">circle.col</span><span class="o">=</span><span class="nf">palette</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/Module1-Data_Processing_19_0.png" src="_images/Module1-Data_Processing_19_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># create Q-Q plot for t-statistic</span>
<span class="n">t.good</span> <span class="o">&lt;-</span> <span class="nf">which</span><span class="p">(</span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">fit2</span><span class="o">$</span><span class="bp">F</span><span class="p">))</span> <span class="c1"># filter out bad probes</span>
<span class="nf">qqt</span><span class="p">(</span><span class="n">fit2</span><span class="o">$</span><span class="n">t</span><span class="p">[</span><span class="n">t.good</span><span class="p">],</span> <span class="n">fit2</span><span class="o">$</span><span class="n">df.total</span><span class="p">[</span><span class="n">t.good</span><span class="p">],</span> <span class="n">main</span><span class="o">=</span><span class="s">&quot;Moderated t statistic&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/Module1-Data_Processing_20_0.png" src="_images/Module1-Data_Processing_20_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># volcano plot (log P-value vs log fold change)</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">fit2</span><span class="p">)</span> <span class="c1"># list contrast names</span>
<span class="n">ct</span> <span class="o">&lt;-</span> <span class="m">1</span>        <span class="c1"># choose contrast of interest</span>
<span class="nf">volcanoplot</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span> <span class="n">coef</span><span class="o">=</span><span class="n">ct</span><span class="p">,</span> <span class="n">main</span><span class="o">=</span><span class="nf">colnames</span><span class="p">(</span><span class="n">fit2</span><span class="p">)[</span><span class="n">ct</span><span class="p">],</span> <span class="n">pch</span><span class="o">=</span><span class="m">20</span><span class="p">,</span>
            <span class="n">highlight</span><span class="o">=</span><span class="nf">length</span><span class="p">(</span><span class="nf">which</span><span class="p">(</span><span class="n">dT</span><span class="p">[,</span><span class="n">ct</span><span class="p">]</span><span class="o">!=</span><span class="m">0</span><span class="p">)),</span> <span class="n">names</span><span class="o">=</span><span class="nf">rep</span><span class="p">(</span><span class="s">&#39;+&#39;</span><span class="p">,</span> <span class="nf">nrow</span><span class="p">(</span><span class="n">fit2</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">'c-d'</div><img alt="_images/Module1-Data_Processing_21_1.png" src="_images/Module1-Data_Processing_21_1.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># MD plot (log fold change vs mean log expression)</span>
<span class="c1"># highlight statistically significant (p-adj &lt; 0.05) probes</span>
<span class="nf">plotMD</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span> <span class="n">column</span><span class="o">=</span><span class="n">ct</span><span class="p">,</span> <span class="n">status</span><span class="o">=</span><span class="n">dT</span><span class="p">[,</span><span class="n">ct</span><span class="p">],</span> <span class="n">legend</span><span class="o">=</span><span class="bp">F</span><span class="p">,</span> <span class="n">pch</span><span class="o">=</span><span class="m">20</span><span class="p">,</span> <span class="n">cex</span><span class="o">=</span><span class="m">1</span><span class="p">)</span>
<span class="nf">abline</span><span class="p">(</span><span class="n">h</span><span class="o">=</span><span class="m">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/Module1-Data_Processing_22_0.png" src="_images/Module1-Data_Processing_22_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># General expression data analysis</span>
<span class="n">ex</span> <span class="o">&lt;-</span> <span class="nf">exprs</span><span class="p">(</span><span class="n">gset</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">head</span><span class="p">(</span><span class="n">ex</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A matrix: 6 × 54 of type dbl</caption>
<thead>
	<tr><th></th><th scope=col>GSM300173</th><th scope=col>GSM300177</th><th scope=col>GSM300181</th><th scope=col>GSM300186</th><th scope=col>GSM300189</th><th scope=col>GSM300192</th><th scope=col>GSM300196</th><th scope=col>GSM300204</th><th scope=col>GSM300208</th><th scope=col>GSM300214</th><th scope=col>⋯</th><th scope=col>GSM1176201</th><th scope=col>GSM1176202</th><th scope=col>GSM1176203</th><th scope=col>GSM1176204</th><th scope=col>GSM1176205</th><th scope=col>GSM1176206</th><th scope=col>GSM1176207</th><th scope=col>GSM1176208</th><th scope=col>GSM1176209</th><th scope=col>GSM1176210</th></tr>
</thead>
<tbody>
	<tr><th scope=row>1007_s_at</th><td>-0.9304071</td><td>-0.31429476</td><td> 0.302564486</td><td> 1.706508981</td><td> 0.44324721</td><td>-0.35662310</td><td>-0.42414478</td><td>-0.68015019</td><td> 0.70561100</td><td> 0.23055657</td><td>⋯</td><td> 0.29735707</td><td> 1.0451786</td><td> 0.35818266</td><td> 0.86631591</td><td>-0.86054531</td><td> 0.73855074</td><td> 0.43443811</td><td> 0.548127436</td><td>-0.2959882</td><td> 0.54792328</td></tr>
	<tr><th scope=row>1053_at</th><td>-0.1899698</td><td>-0.13598038</td><td>-0.303115126</td><td> 0.003071112</td><td>-0.59434151</td><td>-0.30161908</td><td> 0.04182821</td><td>-0.07260928</td><td>-0.29828630</td><td>-0.29855911</td><td>⋯</td><td>-0.24029072</td><td>-0.1255642</td><td> 0.07094486</td><td>-0.15935824</td><td> 0.07998875</td><td>-0.38857263</td><td>-0.09313018</td><td>-0.029009885</td><td> 0.3229603</td><td>-0.01513872</td></tr>
	<tr><th scope=row>117_at</th><td>-0.2784191</td><td>-0.34321870</td><td>-0.008780405</td><td>-0.040806964</td><td> 3.44181211</td><td> 0.02899129</td><td> 0.32742570</td><td> 0.04146729</td><td>-0.39414786</td><td>-0.30109524</td><td>⋯</td><td>-0.10819355</td><td> 0.6173384</td><td>-0.09974673</td><td> 0.59459326</td><td> 0.13922468</td><td>-0.04004724</td><td>-0.03073124</td><td> 0.805958790</td><td> 0.3936756</td><td>-0.20228573</td></tr>
	<tr><th scope=row>121_at</th><td> 0.2035331</td><td> 0.08515227</td><td> 0.027424018</td><td> 0.180686002</td><td> 0.07046586</td><td>-0.11509540</td><td> 0.22434064</td><td> 0.18612576</td><td> 0.01974077</td><td> 0.10861143</td><td>⋯</td><td> 0.01109913</td><td>-0.2074838</td><td>-0.10970414</td><td> 0.41964341</td><td> 0.10523085</td><td>-0.13914736</td><td> 0.02295550</td><td> 0.009695871</td><td> 0.2261897</td><td> 0.14256048</td></tr>
	<tr><th scope=row>1255_g_at</th><td> 0.5083751</td><td>-0.28938746</td><td> 0.081224535</td><td> 0.063953076</td><td>-0.31641582</td><td> 0.60411337</td><td> 0.25961200</td><td> 0.03355860</td><td>-0.19762676</td><td>-0.28731400</td><td>⋯</td><td> 0.57820991</td><td>-0.2487116</td><td>-0.18000081</td><td> 0.04598304</td><td> 0.07277371</td><td>-0.17741782</td><td> 0.16908992</td><td>-0.080829885</td><td> 3.2641306</td><td> 0.15702094</td></tr>
	<tr><th scope=row>1294_at</th><td>-0.3542675</td><td> 0.01620918</td><td> 0.065301945</td><td> 0.933474441</td><td>-0.26148660</td><td> 0.01263875</td><td>-0.62108366</td><td>-0.09775381</td><td>-0.40135476</td><td> 0.04252303</td><td>⋯</td><td> 0.45727226</td><td> 1.3568780</td><td>-0.26581161</td><td> 0.71747749</td><td>-0.18435227</td><td>-0.08382117</td><td> 0.23467213</td><td> 0.351081562</td><td> 0.1501239</td><td> 0.51403397</td></tr>
</tbody>
</table>
</div></div>
</div>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            kernelName: "ir",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="VertexAI.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Setting up Google Cloud environment</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="Module1-Data_Processing2.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Data Processing</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Bang Tran, Hung Nguyen, and Tin Nguyen<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>