
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Differential Expression Analysis &#8212; Consensus Pathway Analysis using Google Cloud Infrastructure</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=62ba249389abaaa9ffc34bf36a076bdc1d65ee18" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=f31d14ad54b65d19161ba51d4ffff3a77ae00456"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Gene mapping" href="Module01-Gene_Mapping.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/Cover.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Consensus Pathway Analysis using Google Cloud Infrastructure</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Overview
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting Started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="VertexAI.html">
   Setting up Google Cloud environment
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 01
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Module01-Data_Processing.html">
   GEO Data processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Module01-Gene_Mapping.html">
   Gene mapping
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Module 02
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Differential Expression Analysis
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://mybinder.org/v2/gh/executablebooks/jupyter-book/master?urlpath=tree/docs/Module02-DE_Analysis.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Launch on Binder"
>
  

<span class="headerbtn__icon-container">
  
    <img src="_static/images/logo_binder.svg">
  </span>
<span class="headerbtn__text-container">Binder</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/executablebooks/jupyter-book"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2FModule02-DE_Analysis.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="_sources/Module02-DE_Analysis.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#learning-objectives">
   Learning Objectives:
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#grouping-samples-and-setting-up-the-design-matrix">
   Grouping samples and setting up the design matrix.
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Differential Expression Analysis</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#learning-objectives">
   Learning Objectives:
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#grouping-samples-and-setting-up-the-design-matrix">
   Grouping samples and setting up the design matrix.
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="differential-expression-analysis">
<h1>Differential Expression Analysis<a class="headerlink" href="#differential-expression-analysis" title="Permalink to this headline">#</a></h1>
<p><img alt="" src="_images/DE_Analysis.png" /></p>
<section id="learning-objectives">
<h2>Learning Objectives:<a class="headerlink" href="#learning-objectives" title="Permalink to this headline">#</a></h2>
<ol class="simple">
<li><p>Assigning samples to groups and set up adesign matrix.</p></li>
<li><p>Performing differential expression (DE) analysis using <code class="docutils literal notranslate"><span class="pre">limma</span></code> R package.</p></li>
<li><p>Further processing and visualisation of DE results.</p></li>
<li><p>Filtering and exporting the results table.</p></li>
<li><p>Further visualisation</p></li>
</ol>
</section>
<section id="grouping-samples-and-setting-up-the-design-matrix">
<h2>Grouping samples and setting up the design matrix.<a class="headerlink" href="#grouping-samples-and-setting-up-the-design-matrix" title="Permalink to this headline">#</a></h2>
<p>In this section we will continue to analyze the <code class="docutils literal notranslate"><span class="pre">GSE48350</span></code> dataset that we have used in the previous chapter. Recall from the <span class="xref myst"><strong>GEO Data processing</strong> </span> section,<code class="docutils literal notranslate"><span class="pre">GSE48350</span></code> contains sequenced data of Human Alzheimer’s
Disease using microarray sequencing technology collected from 4 brain regions. Here we will create create two groups: <em>“condition”</em> and <em>“control”</em> from the Entorhinal Cortex region.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="s">&quot;GEOquery&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="s">&quot;lima&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Loading required package: Biobase
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Loading required package: BiocGenerics
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Attaching package: &#39;BiocGenerics&#39;
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>The following objects are masked from &#39;package:stats&#39;:

    IQR, mad, sd, var, xtabs
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>The following objects are masked from &#39;package:base&#39;:

    anyDuplicated, append, as.data.frame, basename, cbind, colnames,
    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
    union, unique, unsplit, which.max, which.min
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    &#39;browseVignettes()&#39;. To cite Bioconductor, see
    &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Setting options(&#39;download.file.method.GEOquery&#39;=&#39;auto&#39;)
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Setting options(&#39;GEOquery.inmemory.gpl&#39;=FALSE)
</pre></div>
</div>
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="n">Error</span> <span class="ow">in</span> <span class="n">library</span><span class="p">(</span><span class="s2">&quot;lima&quot;</span><span class="p">):</span> <span class="n">there</span> <span class="ow">is</span> <span class="n">no</span> <span class="n">package</span> <span class="n">called</span> <span class="s1">&#39;lima&#39;</span>
<span class="ne">Traceback</span>:

<span class="mi">1</span><span class="o">.</span> <span class="n">library</span><span class="p">(</span><span class="s2">&quot;lima&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># load series and platform data from GEO</span>

<span class="n">gset</span> <span class="o">&lt;-</span> <span class="nf">getGEO</span><span class="p">(</span><span class="s">&quot;GSE48350&quot;</span><span class="p">,</span> <span class="n">GSEMatrix</span> <span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span> <span class="n">AnnotGPL</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
<span class="nf">if </span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">gset</span><span class="p">)</span> <span class="o">&gt;</span> <span class="m">1</span><span class="p">)</span> <span class="n">idx</span> <span class="o">&lt;-</span> <span class="nf">grep</span><span class="p">(</span><span class="s">&quot;GPL570&quot;</span><span class="p">,</span> <span class="nf">attr</span><span class="p">(</span><span class="n">gset</span><span class="p">,</span> <span class="s">&quot;names&quot;</span><span class="p">))</span> <span class="n">else</span> <span class="n">idx</span> <span class="o">&lt;-</span> <span class="m">1</span>
<span class="n">gset</span> <span class="o">&lt;-</span> <span class="n">gset</span><span class="p">[[</span><span class="n">idx</span><span class="p">]]</span>

<span class="c1"># make proper column names to match toptable </span>
<span class="nf">fvarLabels</span><span class="p">(</span><span class="n">gset</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">make.names</span><span class="p">(</span><span class="nf">fvarLabels</span><span class="p">(</span><span class="n">gset</span><span class="p">))</span>

<span class="c1"># group membership for all samples</span>
<span class="n">gsms</span> <span class="o">&lt;-</span> <span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;XXXXXXX0XXX0XXX0XXX0XX0XX0XXX0XXXXXXX0XXX0XXXXX0XX&quot;</span><span class="p">,</span>
               <span class="s">&quot;X0XXX0XXXXX0X0XXX0XXX0XXX0XXXXX0XXX0XXXX0XX0XXXXX0&quot;</span><span class="p">,</span>
               <span class="s">&quot;XXX0XXX0XXX0XXXXX0XXX0XXX0XXX0XX0XXX0XXX0XXX0XXX0X&quot;</span><span class="p">,</span>
               <span class="s">&quot;XX0XXX0XXX0XXX0XX0XXXXX111111111111111XXXXXXXXXXXX&quot;</span><span class="p">,</span>
               <span class="s">&quot;XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX&quot;</span><span class="p">,</span>
               <span class="s">&quot;XXX&quot;</span><span class="p">)</span>
<span class="n">sml</span> <span class="o">&lt;-</span> <span class="nf">strsplit</span><span class="p">(</span><span class="n">gsms</span><span class="p">,</span> <span class="n">split</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">)[[</span><span class="m">1</span><span class="p">]]</span>

<span class="c1"># filter out excluded samples (marked as &quot;X&quot;)</span>
<span class="n">sel</span> <span class="o">&lt;-</span> <span class="nf">which</span><span class="p">(</span><span class="n">sml</span> <span class="o">!=</span> <span class="s">&quot;X&quot;</span><span class="p">)</span>
<span class="n">sml</span> <span class="o">&lt;-</span> <span class="n">sml</span><span class="p">[</span><span class="n">sel</span><span class="p">]</span>
<span class="n">gset</span> <span class="o">&lt;-</span> <span class="n">gset</span><span class="p">[</span> <span class="p">,</span><span class="n">sel</span><span class="p">]</span>

<span class="c1"># log2 transformation</span>
<span class="n">ex</span> <span class="o">&lt;-</span> <span class="nf">exprs</span><span class="p">(</span><span class="n">gset</span><span class="p">)</span>
<span class="n">qx</span> <span class="o">&lt;-</span> <span class="nf">as.numeric</span><span class="p">(</span><span class="nf">quantile</span><span class="p">(</span><span class="n">ex</span><span class="p">,</span> <span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="n">.</span><span class="p">,</span> <span class="m">0.25</span><span class="p">,</span> <span class="m">0.5</span><span class="p">,</span> <span class="m">0.75</span><span class="p">,</span> <span class="m">0.99</span><span class="p">,</span> <span class="m">1.0</span><span class="p">),</span> <span class="n">na.rm</span><span class="o">=</span><span class="bp">T</span><span class="p">))</span>
<span class="n">LogC</span> <span class="o">&lt;-</span> <span class="p">(</span><span class="n">qx</span><span class="p">[</span><span class="m">5</span><span class="p">]</span> <span class="o">&gt;</span> <span class="m">100</span><span class="p">)</span> <span class="o">||</span>
  <span class="p">(</span><span class="n">qx</span><span class="p">[</span><span class="m">6</span><span class="p">]</span><span class="o">-</span><span class="n">qx</span><span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="m">50</span> <span class="o">&amp;&amp;</span> <span class="n">qx</span><span class="p">[</span><span class="m">2</span><span class="p">]</span> <span class="o">&gt;</span> <span class="m">0</span><span class="p">)</span>
<span class="nf">if </span><span class="p">(</span><span class="n">LogC</span><span class="p">)</span> <span class="p">{</span> <span class="n">ex</span><span class="p">[</span><span class="nf">which</span><span class="p">(</span><span class="n">ex</span> <span class="o">&lt;=</span> <span class="m">0</span><span class="p">)]</span> <span class="o">&lt;-</span> <span class="kc">NaN</span>
<span class="nf">exprs</span><span class="p">(</span><span class="n">gset</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">log2</span><span class="p">(</span><span class="n">ex</span><span class="p">)</span> <span class="p">}</span>

<span class="c1"># assign samples to groups and set up design matrix</span>
<span class="n">gs</span> <span class="o">&lt;-</span> <span class="nf">factor</span><span class="p">(</span><span class="n">sml</span><span class="p">)</span>
<span class="n">groups</span> <span class="o">&lt;-</span> <span class="nf">make.names</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;c&quot;</span><span class="p">,</span><span class="s">&quot;d&quot;</span><span class="p">))</span>
<span class="nf">levels</span><span class="p">(</span><span class="n">gs</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">groups</span>
<span class="n">gset</span><span class="o">$</span><span class="n">group</span> <span class="o">&lt;-</span> <span class="n">gs</span>
<span class="n">design</span> <span class="o">&lt;-</span> <span class="nf">model.matrix</span><span class="p">(</span><span class="o">~</span><span class="n">group</span> <span class="o">+</span> <span class="m">0</span><span class="p">,</span> <span class="n">gset</span><span class="p">)</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">design</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">levels</span><span class="p">(</span><span class="n">gs</span><span class="p">)</span>

<span class="n">fit</span> <span class="o">&lt;-</span> <span class="nf">lmFit</span><span class="p">(</span><span class="n">gset</span><span class="p">,</span> <span class="n">design</span><span class="p">)</span>  <span class="c1"># fit linear model</span>

<span class="c1"># set up contrasts of interest and recalculate model coefficients</span>
<span class="n">cts</span> <span class="o">&lt;-</span> <span class="nf">paste</span><span class="p">(</span><span class="n">groups</span><span class="p">[</span><span class="m">1</span><span class="p">],</span> <span class="n">groups</span><span class="p">[</span><span class="m">2</span><span class="p">],</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;-&quot;</span><span class="p">)</span>
<span class="n">cont.matrix</span> <span class="o">&lt;-</span> <span class="nf">makeContrasts</span><span class="p">(</span><span class="n">contrasts</span><span class="o">=</span><span class="n">cts</span><span class="p">,</span> <span class="n">levels</span><span class="o">=</span><span class="n">design</span><span class="p">)</span>
<span class="n">fit2</span> <span class="o">&lt;-</span> <span class="nf">contrasts.fit</span><span class="p">(</span><span class="n">fit</span><span class="p">,</span> <span class="n">cont.matrix</span><span class="p">)</span>

<span class="c1"># compute statistics and table of top significant genes</span>
<span class="n">fit2</span> <span class="o">&lt;-</span> <span class="nf">eBayes</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span> <span class="m">0.01</span><span class="p">)</span>
<span class="n">tT</span> <span class="o">&lt;-</span> <span class="nf">topTable</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span> <span class="n">adjust</span><span class="o">=</span><span class="s">&quot;fdr&quot;</span><span class="p">,</span> <span class="n">sort.by</span><span class="o">=</span><span class="s">&quot;B&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="m">250</span><span class="p">)</span>

<span class="n">tT</span> <span class="o">&lt;-</span> <span class="nf">subset</span><span class="p">(</span><span class="n">tT</span><span class="p">,</span> <span class="n">select</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ID&quot;</span><span class="p">,</span><span class="s">&quot;adj.P.Val&quot;</span><span class="p">,</span><span class="s">&quot;P.Value&quot;</span><span class="p">,</span><span class="s">&quot;t&quot;</span><span class="p">,</span><span class="s">&quot;B&quot;</span><span class="p">,</span><span class="s">&quot;logFC&quot;</span><span class="p">,</span><span class="s">&quot;Gene.symbol&quot;</span><span class="p">,</span><span class="s">&quot;Gene.title&quot;</span><span class="p">))</span>
<span class="nf">write.table</span><span class="p">(</span><span class="n">tT</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="nf">stdout</span><span class="p">(),</span> <span class="n">row.names</span><span class="o">=</span><span class="bp">F</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Found 1 file(s)

GSE48350_series_matrix.txt.gz

Using locally cached version: C:\Users\bangt\AppData\Local\Temp\RtmpOQuY8S/GSE48350_series_matrix.txt.gz
</pre></div>
</div>
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="nn">Error</span> in <span class="ni">writeBin(bfr, con = out, size = 1L): &#39;Calloc&#39; could not allocate memory </span><span class="nt">(10000000 of 1 bytes)</span>
<span class="ne">Traceback</span>:

<span class="mi">1</span><span class="o">.</span> <span class="n">getGEO</span><span class="p">(</span><span class="s2">&quot;GSE48350&quot;</span><span class="p">,</span> <span class="n">GSEMatrix</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">AnnotGPL</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
<span class="mi">2</span><span class="o">.</span> <span class="n">getAndParseGSEMatrices</span><span class="p">(</span><span class="n">GEO</span><span class="p">,</span> <span class="n">destdir</span><span class="p">,</span> <span class="n">AnnotGPL</span> <span class="o">=</span> <span class="n">AnnotGPL</span><span class="p">,</span> <span class="n">getGPL</span> <span class="o">=</span> <span class="n">getGPL</span><span class="p">,</span> 
 <span class="o">.</span>     <span class="n">parseCharacteristics</span> <span class="o">=</span> <span class="n">parseCharacteristics</span><span class="p">)</span>
<span class="mi">3</span><span class="o">.</span> <span class="n">parseGSEMatrix</span><span class="p">(</span><span class="n">destfile</span><span class="p">,</span> <span class="n">destdir</span> <span class="o">=</span> <span class="n">destdir</span><span class="p">,</span> <span class="n">AnnotGPL</span> <span class="o">=</span> <span class="n">AnnotGPL</span><span class="p">,</span> 
 <span class="o">.</span>     <span class="n">getGPL</span> <span class="o">=</span> <span class="n">getGPL</span><span class="p">)</span>
<span class="mi">4</span><span class="o">.</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">::</span><span class="n">fread</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="mi">5</span><span class="o">.</span> <span class="n">R</span><span class="o">.</span><span class="n">utils</span><span class="p">::</span><span class="n">decompressFile</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">decompFile</span> <span class="o">&lt;-</span> <span class="n">tempfile</span><span class="p">(</span><span class="n">tmpdir</span> <span class="o">=</span> <span class="n">tmpdir</span><span class="p">),</span> 
 <span class="o">.</span>     <span class="n">ext</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">,</span> <span class="n">FUN</span> <span class="o">=</span> <span class="n">FUN</span><span class="p">,</span> <span class="n">remove</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
<span class="mi">6</span><span class="o">.</span> <span class="n">decompressFile</span><span class="o">.</span><span class="n">default</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">decompFile</span> <span class="o">&lt;-</span> <span class="n">tempfile</span><span class="p">(</span><span class="n">tmpdir</span> <span class="o">=</span> <span class="n">tmpdir</span><span class="p">),</span> 
 <span class="o">.</span>     <span class="n">ext</span> <span class="o">=</span> <span class="n">NULL</span><span class="p">,</span> <span class="n">FUN</span> <span class="o">=</span> <span class="n">FUN</span><span class="p">,</span> <span class="n">remove</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
<span class="mi">7</span><span class="o">.</span> <span class="n">writeBin</span><span class="p">(</span><span class="n">bfr</span><span class="p">,</span> <span class="n">con</span> <span class="o">=</span> <span class="n">out</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="mi">1</span><span class="n">L</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">write.table</span><span class="p">(</span><span class="n">tT</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="nf">stdout</span><span class="p">(),</span> <span class="n">row.names</span><span class="o">=</span><span class="bp">F</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="n">Error</span> <span class="ow">in</span> <span class="ow">is</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">x</span><span class="p">):</span> <span class="nb">object</span> <span class="s1">&#39;tT&#39;</span> <span class="ow">not</span> <span class="n">found</span>
<span class="ne">Traceback</span>:

<span class="mi">1</span><span class="o">.</span> <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">tT</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">stdout</span><span class="p">(),</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="mi">2</span><span class="o">.</span> <span class="ow">is</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize and quality control test results.</span>
<span class="c1"># Build histogram of P-values for all genes. Normal test</span>
<span class="c1"># assumption is that most genes are not differentially expressed.</span>
<span class="n">tT2</span> <span class="o">&lt;-</span> <span class="nf">topTable</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span> <span class="n">adjust</span><span class="o">=</span><span class="s">&quot;fdr&quot;</span><span class="p">,</span> <span class="n">sort.by</span><span class="o">=</span><span class="s">&quot;B&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="kc">Inf</span><span class="p">)</span>
<span class="nf">hist</span><span class="p">(</span><span class="n">tT2</span><span class="o">$</span><span class="n">adj.P.Val</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="s">&quot;grey&quot;</span><span class="p">,</span> <span class="n">border</span> <span class="o">=</span> <span class="s">&quot;white&quot;</span><span class="p">,</span> <span class="n">xlab</span> <span class="o">=</span> <span class="s">&quot;P-adj&quot;</span><span class="p">,</span>
     <span class="n">ylab</span> <span class="o">=</span> <span class="s">&quot;Number of genes&quot;</span><span class="p">,</span> <span class="n">main</span> <span class="o">=</span> <span class="s">&quot;P-adj value distribution&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/Module02-DE_Analysis_5_0.png" src="_images/Module02-DE_Analysis_5_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># summarize test results as &quot;up&quot;, &quot;down&quot; or &quot;not expressed&quot;</span>
<span class="n">dT</span> <span class="o">&lt;-</span> <span class="nf">decideTests</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span> <span class="n">adjust.method</span><span class="o">=</span><span class="s">&quot;fdr&quot;</span><span class="p">,</span> <span class="n">p.value</span><span class="o">=</span><span class="m">0.05</span><span class="p">)</span>
<span class="c1"># Venn diagram of results</span>
<span class="nf">vennDiagram</span><span class="p">(</span><span class="n">dT</span><span class="p">,</span> <span class="n">circle.col</span><span class="o">=</span><span class="nf">palette</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/Module02-DE_Analysis_6_0.png" src="_images/Module02-DE_Analysis_6_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># create Q-Q plot for t-statistic</span>
<span class="n">t.good</span> <span class="o">&lt;-</span> <span class="nf">which</span><span class="p">(</span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">fit2</span><span class="o">$</span><span class="bp">F</span><span class="p">))</span> <span class="c1"># filter out bad probes</span>
<span class="nf">qqt</span><span class="p">(</span><span class="n">fit2</span><span class="o">$</span><span class="n">t</span><span class="p">[</span><span class="n">t.good</span><span class="p">],</span> <span class="n">fit2</span><span class="o">$</span><span class="n">df.total</span><span class="p">[</span><span class="n">t.good</span><span class="p">],</span> <span class="n">main</span><span class="o">=</span><span class="s">&quot;Moderated t statistic&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/Module02-DE_Analysis_7_0.png" src="_images/Module02-DE_Analysis_7_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># volcano plot (log P-value vs log fold change)</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">fit2</span><span class="p">)</span> <span class="c1"># list contrast names</span>
<span class="n">ct</span> <span class="o">&lt;-</span> <span class="m">1</span>        <span class="c1"># choose contrast of interest</span>
<span class="nf">volcanoplot</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span> <span class="n">coef</span><span class="o">=</span><span class="n">ct</span><span class="p">,</span> <span class="n">main</span><span class="o">=</span><span class="nf">colnames</span><span class="p">(</span><span class="n">fit2</span><span class="p">)[</span><span class="n">ct</span><span class="p">],</span> <span class="n">pch</span><span class="o">=</span><span class="m">20</span><span class="p">,</span>
            <span class="n">highlight</span><span class="o">=</span><span class="nf">length</span><span class="p">(</span><span class="nf">which</span><span class="p">(</span><span class="n">dT</span><span class="p">[,</span><span class="n">ct</span><span class="p">]</span><span class="o">!=</span><span class="m">0</span><span class="p">)),</span> <span class="n">names</span><span class="o">=</span><span class="nf">rep</span><span class="p">(</span><span class="s">&#39;+&#39;</span><span class="p">,</span> <span class="nf">nrow</span><span class="p">(</span><span class="n">fit2</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">'c-d'</div><img alt="_images/Module02-DE_Analysis_8_1.png" src="_images/Module02-DE_Analysis_8_1.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># MD plot (log fold change vs mean log expression)</span>
<span class="c1"># highlight statistically significant (p-adj &lt; 0.05) probes</span>
<span class="nf">plotMD</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span> <span class="n">column</span><span class="o">=</span><span class="n">ct</span><span class="p">,</span> <span class="n">status</span><span class="o">=</span><span class="n">dT</span><span class="p">[,</span><span class="n">ct</span><span class="p">],</span> <span class="n">legend</span><span class="o">=</span><span class="bp">F</span><span class="p">,</span> <span class="n">pch</span><span class="o">=</span><span class="m">20</span><span class="p">,</span> <span class="n">cex</span><span class="o">=</span><span class="m">1</span><span class="p">)</span>
<span class="nf">abline</span><span class="p">(</span><span class="n">h</span><span class="o">=</span><span class="m">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/Module02-DE_Analysis_9_0.png" src="_images/Module02-DE_Analysis_9_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># General expression data analysis</span>
<span class="n">ex</span> <span class="o">&lt;-</span> <span class="nf">exprs</span><span class="p">(</span><span class="n">gset</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">head</span><span class="p">(</span><span class="n">ex</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A matrix: 6 × 54 of type dbl</caption>
<thead>
	<tr><th></th><th scope=col>GSM300173</th><th scope=col>GSM300177</th><th scope=col>GSM300181</th><th scope=col>GSM300186</th><th scope=col>GSM300189</th><th scope=col>GSM300192</th><th scope=col>GSM300196</th><th scope=col>GSM300204</th><th scope=col>GSM300208</th><th scope=col>GSM300214</th><th scope=col>⋯</th><th scope=col>GSM1176201</th><th scope=col>GSM1176202</th><th scope=col>GSM1176203</th><th scope=col>GSM1176204</th><th scope=col>GSM1176205</th><th scope=col>GSM1176206</th><th scope=col>GSM1176207</th><th scope=col>GSM1176208</th><th scope=col>GSM1176209</th><th scope=col>GSM1176210</th></tr>
</thead>
<tbody>
	<tr><th scope=row>1007_s_at</th><td>-0.9304071</td><td>-0.31429476</td><td> 0.302564486</td><td> 1.706508981</td><td> 0.44324721</td><td>-0.35662310</td><td>-0.42414478</td><td>-0.68015019</td><td> 0.70561100</td><td> 0.23055657</td><td>⋯</td><td> 0.29735707</td><td> 1.0451786</td><td> 0.35818266</td><td> 0.86631591</td><td>-0.86054531</td><td> 0.73855074</td><td> 0.43443811</td><td> 0.548127436</td><td>-0.2959882</td><td> 0.54792328</td></tr>
	<tr><th scope=row>1053_at</th><td>-0.1899698</td><td>-0.13598038</td><td>-0.303115126</td><td> 0.003071112</td><td>-0.59434151</td><td>-0.30161908</td><td> 0.04182821</td><td>-0.07260928</td><td>-0.29828630</td><td>-0.29855911</td><td>⋯</td><td>-0.24029072</td><td>-0.1255642</td><td> 0.07094486</td><td>-0.15935824</td><td> 0.07998875</td><td>-0.38857263</td><td>-0.09313018</td><td>-0.029009885</td><td> 0.3229603</td><td>-0.01513872</td></tr>
	<tr><th scope=row>117_at</th><td>-0.2784191</td><td>-0.34321870</td><td>-0.008780405</td><td>-0.040806964</td><td> 3.44181211</td><td> 0.02899129</td><td> 0.32742570</td><td> 0.04146729</td><td>-0.39414786</td><td>-0.30109524</td><td>⋯</td><td>-0.10819355</td><td> 0.6173384</td><td>-0.09974673</td><td> 0.59459326</td><td> 0.13922468</td><td>-0.04004724</td><td>-0.03073124</td><td> 0.805958790</td><td> 0.3936756</td><td>-0.20228573</td></tr>
	<tr><th scope=row>121_at</th><td> 0.2035331</td><td> 0.08515227</td><td> 0.027424018</td><td> 0.180686002</td><td> 0.07046586</td><td>-0.11509540</td><td> 0.22434064</td><td> 0.18612576</td><td> 0.01974077</td><td> 0.10861143</td><td>⋯</td><td> 0.01109913</td><td>-0.2074838</td><td>-0.10970414</td><td> 0.41964341</td><td> 0.10523085</td><td>-0.13914736</td><td> 0.02295550</td><td> 0.009695871</td><td> 0.2261897</td><td> 0.14256048</td></tr>
	<tr><th scope=row>1255_g_at</th><td> 0.5083751</td><td>-0.28938746</td><td> 0.081224535</td><td> 0.063953076</td><td>-0.31641582</td><td> 0.60411337</td><td> 0.25961200</td><td> 0.03355860</td><td>-0.19762676</td><td>-0.28731400</td><td>⋯</td><td> 0.57820991</td><td>-0.2487116</td><td>-0.18000081</td><td> 0.04598304</td><td> 0.07277371</td><td>-0.17741782</td><td> 0.16908992</td><td>-0.080829885</td><td> 3.2641306</td><td> 0.15702094</td></tr>
	<tr><th scope=row>1294_at</th><td>-0.3542675</td><td> 0.01620918</td><td> 0.065301945</td><td> 0.933474441</td><td>-0.26148660</td><td> 0.01263875</td><td>-0.62108366</td><td>-0.09775381</td><td>-0.40135476</td><td> 0.04252303</td><td>⋯</td><td> 0.45727226</td><td> 1.3568780</td><td>-0.26581161</td><td> 0.71747749</td><td>-0.18435227</td><td>-0.08382117</td><td> 0.23467213</td><td> 0.351081562</td><td> 0.1501239</td><td> 0.51403397</td></tr>
</tbody>
</table>
</div></div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            kernelName: "ir",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="Module01-Gene_Mapping.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Gene mapping</p>
        </div>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Bang Tran, Hung Nguyen, and Tin Nguyen<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>